---
title: "P471-1 Single cell RNAseq of Treg from IL2 mutein-treated mice"
author: "Matt Lawrance"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    df_print: paged
editor_options: 
  chunk_output_type: console
---

<style type="text/css">
body{ /* Normal  */
      font-size: 14px;
  }
h1 { /* Header 1 */
  font-size: 28px;
}
h2 { /* Header 2 */
    font-size: 24px;
}
h3 { /* Header 3 */
  font-size: 20px;
}
h4 { /* Header 4 */
  font-size: 16px;
}
</style>

# Project Summary

This project contains 10x single cell sequencing of three groups of mice -- niave, WT Il-2 treated and IL-2 Mutein treated. T-cells have been captured and TCRs + GEX + FB recorded.

```{r setup, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
library(knitr)
library(tidyverse)

library(ggthemes)
library(edgeR)
library(RColorBrewer)
library(kableExtra)
library(viridis)
library(ggbeeswarm)
library(gplots)
library(GGally)
library(plotscale) # allows outputting sized plots

library(ComplexHeatmap)
library(Seurat)

# BCR analysis tools
library(clonality) # devtools::install_github("moodymudskipper/safejoin"); devtools::install_github("victoraLab/clonality")
library(immunarch) # install.packages("immunarch")
library(alakazam)
library(BrepPhylo) # devtools::install_github("Fraternalilab/BrepPhylo")
library(scRepertoire) # devtools::install_github("ncborcherding/scRepertoire")

library(bRi) # install instructions at github.com/BenaroyaResearch/bRi
library(miscHelpers) # devtools::install_github("BenaroyaResearch/miscHelpers")
library(RNAseQC) # devtools::install_github("BenaroyaResearch/RNAseQC")
library(countSubsetNorm) # devtools::install_github("BenaroyaResearch/countSubsetNorm")

library(magrittr)
opts_chunk$set(
  fig.width=6, fig.height=4.25, 
  cache = TRUE, # caching does not work with Seurat objects
  echo = FALSE, warning=FALSE, message=FALSE)
options(stringsAsFactors = FALSE)

theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill = NA, size = 1),
          axis.text = element_text(colour="black"),
          axis.ticks = element_line(colour="black"),
          axis.text.x = element_text(angle=0)))
update_geom_defaults("point", list(shape=16))
grDevices::pdf.options(useDingbats = FALSE)

# test_plot <- ggplot2::ggplot(mapping = ggplot2::aes(x = 1:10, y = 1:10)) + ggplot2::geom_point()
# ggplot(mapping = aes(x = 1:10, y = 1:10)) + geom_point()
```

```{r setFilenamesAndGlobalVariables}
dirBoxBase <- 
  file.path("~", "Library", "CloudStorage", "Box-Box")
dirRoot <-
  file.path(
    dirBoxBase, "Projects", # Box version
    # "~", "Documents", "Projects_local", # local version
    "P471-1_10x_analysis")
dataDate <- "2022-04-13"

filenameSuffix <- paste0("P471-1_10x_analysis.", dataDate)

projectNumberGenomicsCore <- "P389-4"
tenxSummaryDir <- "220413-P471-1allFlowcells"
tenxSampleName <- "Pool_4"
```

```{r setDirectory, cache=FALSE}
opts_knit$set(root.dir = dirRoot)
setwd(dirRoot)

# save/load temporary versions of data objects
# save(list = ls(), file = file.path("data_saved", "workspaceTmp.RData"))
# load(file.path("data_saved", "workspaceTmp.RData"))

# save/load temporary version of data10x
# saveRDS(data10x, file = file.path("data_saved", "data10xTmp.RDS"))
# readRDS(file.path("data_saved", "data10xTmp.RDS"))
```

# Load project data

## Load project info and libraries

```{r getProjectInfo, dependson="setFilenamesAndGlobalVariables"}
filenameProjectInfo <-
  file.path("data_saved", paste("projectInfo", filenameSuffix, "RDS", sep="."))
  
# check for local version of project info
if (file.exists(filenameProjectInfo)) {
  projectInfo <- readRDS(filenameProjectInfo)
} else {
  # read in projectInfo from ResearchDB
  projectInfo <- apird::getGcqProjectInfo(projectNumberGenomicsCore)
  
  # save local copy
  saveRDS(projectInfo, file = filenameProjectInfo)
}

projectInfo <-
  projectInfo %>%
  dplyr::rename(project = projName) %>%
  # dplyr::filter(
  #   str_detect(project, file_name_input_regex),
  #   !str_detect(project, file_name_input_regex_exclude)) %>% # filter sub-projects
  mutate(
    project =
      factor(project,
             levels = str_sort(unique(project), numeric=TRUE))) %>%
  dplyr::arrange(project)
```

```{r getProjectLibs, dependson="getProjectInfo", eval=FALSE}
filenameProjectLibs <-
  file.path("data_saved", paste("projectLibs", filenameSuffix, "RDS", sep="."))
  
# check for local version of project info
if (file.exists(filenameProjectLibs)) {
  projectLibs <- readRDS(filenameProjectLibs)
} else {
  # read in projectLibs from ResearchDB and convert to data frame
  if (length(projectInfo$project) == 1) {
    projectLibs <-
      apird::getProjectLibs(projectInfo$project)
    projectLibs <-
      data.frame(
        project =
          rep(projectInfo$project, times = length(projectLibs)),
        libid = projectLibs)
  } else if (length(projectInfo$project) > 1) {
    projectLibs <-
      sapply(as.character(projectInfo$project), apird::getProjectLibs)
    projectLibs <-
      data.frame(
        project = rep(names(projectLibs), times = sapply(projectLibs, length)),
        libid = unlist(projectLibs))
  }
  
  projectLibs <-
    projectLibs %>%
    mutate(
      project =
        factor(project,
               levels = str_sort(unique(project), numeric=TRUE))) %>%
    dplyr::arrange(project, libid)
  
  # save local copy
  saveRDS(projectLibs, file = filenameProjectLibs)
}
```

## Load metrics

```{r loadMetricsFromFiles, dependson="getProjectLibs"}
filenameMetrics <-
  file.path("data_saved", paste("metrics", filenameSuffix, "RDS", sep="."))

# check for local version of metrics
if (file.exists(filenameMetrics)) {
  metrics <- readRDS(filenameMetrics)
} else {
  fileDirMetrics <-
    file.path("/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
              tenxSummaryDir, "multi_per_sample")
  
  if (!dir.exists(fileDirMetrics)) 
    stop(paste("Cannot access directory", fileDirMetrics, "containing metrics files."))
  
  filenameListMetrics <-
    dir(fileDirMetrics, pattern="metrics_summary.csv")
  
  if (exists("metrics")) stop("Object `metrics` already exists")
  for (filename.tmp in filenameListMetrics) {
    if (exists("metrics")) {
      metrics <-
        bind_rows(
          metrics,
          read.csv(file.path(fileDirMetrics, filename.tmp)) %>%
            mutate(sampleId = str_replace(filename.tmp, "_metrics_summary.csv", "")))
    } else {
      metrics <-
        read.csv(file.path(fileDirMetrics, filename.tmp)) %>%
        mutate(sampleId = str_replace(filename.tmp, "_metrics_summary.csv", ""))
    }
  }
  
  metrics <-
    metrics %>%
    standardize_dimnames() %>%
    dplyr::select(sampleId = sampleid, everything())
  
  rm_tmp(ask=FALSE)
  
  # save local copy
  saveRDS(metrics, file = filenameMetrics)
}
```

```{r loadMetrics, dependson="getProjectLibs", eval=FALSE}
filenameMetrics <-
  file.path("data_saved", paste("metrics", filenameSuffix, "RDS", sep="."))

# check for local version of metrics
if (file.exists(filenameMetrics)) {
  metrics <- readRDS(filenameMetrics)
} else {
  # read in metrics from ResearchDB
  metrics <-
    apird::getMetrics(projectLibs$libid) %>%
    dplyr::arrange(libid_fcid)
  
  # save local copy
  saveRDS(metrics, file = filenameMetrics)
}

metrics <-
  metrics %>%
  mutate(
    libid = str_extract(libid_fcid, "^lib[0-9]+")) %>%
  merge(projectLibs, by = "libid") %>%
  dplyr::arrange(project, libid)
```

## Load sample annotation

```{r loadRnaseqAnnotationFromFile, results="hide", dependson="setFilenamesAndGlobalVariables"}
### read in library-level annotation from file, so I can play with it before the database is updated
filenameRnaseqAnnotationFromFile <-
  file.path(
    "annotation",
    paste("P389-4_Final Annotation.xlsx"))

rnaseqAnnotation <-
  readxl::read_xlsx(filenameRnaseqAnnotationFromFile) %>%
  remove_all_NA_rowcols() %>%
  dplyr::rename(
    project = `Proj-Sub...1`,
    libid = `library Sample ID`) %>%
  coalesce_with_checks(columns = c("project", "Proj-Sub...22")) %>%
  mutate( # make project a factor, sorted properly
    project =
      project %>%
      factor(levels = str_sort(unique(project), numeric=TRUE))) %>%
  dplyr::arrange(project, libid)
```

```{r cleanRnaseqAnnotationFromFile, results="hide", dependson="loadRnaseqAnnotationFromFile"}
## this set of commands cleans up the messy rnaseqAnnotation data
rnaseqAnnotation <-
  rnaseqAnnotation %>%
  coerce_NA_text() %>%
  remove_all_NA_rowcols() %>%
  dplyr::rename(
    sampleId = `Sample Name`,
    contentType = `Content Type`,
    studyGroup = `Study Group`,
    donorId = `Donor ID`,
    cellType = `Cell Type`,
    cellNumber = `Cell Number`,
    reportedSex = Sex,
    collectedDate = `Date collected`,
    receivedDate = `Date received`,
    sample_id = `Sample ID`,
    # sampleRepositoryID = sample_repository_id,
    cellCountSorter = `Initial Cell Count- Sorter`,
    cellTarget = `Targeted Cell Recovery`,
    cdnaSampleId = `cDNA Sample ID`,
    cdnaCycleNumber = `cDNA Cycle #`,
    cdnaType = `cDNA Type`,
    pelletCdnaNg = `Pellet cDNA ng`,
    pelletCdnaNgPerUl = `Pellet cDNA ng/ul`,
    bcrCdnaNgPerUl = `BCR cDNA ng/ul`,
    bcrCdnaAvgSizeBp = `BCR cDNA Avg size (bp)`,
    cdnaProtocolId = protocolId,
    libraryType = `library type`,
    index7 = I7_Index_ID,
    index7Sequence = index,
    index5 = I5_Index_ID,
    index5Sequence = index2
  ) %>%
  mutate(
    sampleId = sampleId %>%
      factor(levels = str_sort(unique(sampleId))),
    collectedDate = standardize_dates(collectedDate),
    reportedSex =
      reportedSex %>%
      str_to_upper() %>%
      str_extract("^[FM]") %>%
      factor(levels=c("F", "M")),
    `BCR 1 PCR` = `BCR 1 PCR` %>%
      str_extract("^[0-9]+") %>%
      as.numeric()
  ) %>%
  coalesce_with_checks(columns = c("cdnaCycleNumber", "BCR 1 PCR")) %>%
  dplyr::select(-`0`)
```

```{r loadRnaseqAnnotation, results="hide", dependson="getProjectLibs", eval=FALSE}
### read in library-level annotation
filenameRnaseqAnnotation <-
  file.path(
    "data_saved",
    paste("rnaseqAnnotation", filenameSuffix, "RDS", sep="."))

# check for local version of rnaseqAnnotation
if (file.exists(filenameRnaseqAnnotation)) {
  rnaseqAnnotation <- readRDS(filenameRnaseqAnnotation)
} else {
  # read in rnaseqAnnotation from ResearchDB
  rnaseqAnnotation <-
    apird::getAnno(projectLibs$libid) %>%
    mutate( # make project a factor, sorted properly
      project =
        project %>%
        factor(levels = str_sort(unique(project), numeric=TRUE))) %>%
    dplyr::arrange(project, libid)
  
  # save local copy
  saveRDS(rnaseqAnnotation, file=filenameRnaseqAnnotation)
}
```

```{r combineAnnotationMetrics, results="hide", dependson=c("loadMetricsFromFile", "loadRnaseqAnnotation"), eval=FALSE}
## combine library annotation, sample annotation, metrics into single object
design <-
  rnaseqAnnotation %>%
  dplyr::select(
    project, sampleId, libid, libraryType,
    everything()) %>%
  left_join(metrics) %>%
  droplevels() %>%
  dplyr::arrange(project, libid)
```

```{r setHtoSampleId}
# generate data frame with hashtag-mouse links; easier to update the naming this way
nSamples.tmp <- 3

sampleBarcodes <- paste0("HTO-M", 1:nSamples.tmp)
sampleNames <- paste0("mouse", 1:nSamples.tmp)

linkSampleBarcodeToSampleName <-
  data.frame(
    sampleBarcode = 
      c(sampleBarcodes, "multiplePositive", "unassigned") %>%
      factor(levels = str_sort(., numeric = TRUE)),
    sampleName =
      c(sampleNames, "multiplePositive", "unassigned") %>%
      factor(levels = str_sort(., numeric = TRUE)))

rm_tmp(ask=FALSE)
```

```{r setHeatmapColors}
heatmapColors <-
  colorRampPalette(rev(RColorBrewer::brewer.pal(9, "RdBu")))(101) # blue - white - red
```

## Generate color palettes for downstream use

```{r setupPalettes, dependson=c("combineAnnotationMetrics", "loadPalettesAndSettings")}
# set plotOutputFormat
plotOutputFormat <- "pdf"
outputPlots <- function(x, fileDir = "plots", fileStem, format = c("pdf", "png"), width, height, pngRes = 150) {
  if ("pdf" %in% format) {
    pdf(file.path(fileDir, paste0(fileStem, ".pdf")),
        w = width, h = height)
    print(plot.tmp)
    invisible(dev.off())
  }
  if ("png" %in% format) {
    png(file.path(fileDir, paste0(fileStem, ".png")),
        w = width, h = height, units = "in", res = pngRes)
    print(plot.tmp)
    invisible(dev.off())
  }
}

# set default resolution for raster layers
rasterResolutionDpi <- 300

# cell barcode with colors for HTO_M1 through HTO_M10, black for unassigned, red for multiplePositive
pal.sampleBarcode <-
  big_colorblind_pal(
    n_colors = 4,
    drop_yellow = TRUE, drop_black = FALSE) %>%
  # B0251 = Hashtag1, B0260 = Hashtag2
  setNames(c("unassigned", paste0("HTO-M", 1:3))) %>%
  c("multiplePositive" = "red")

pal.clusters <-
  big_colorblind_pal(
    n_colors = 7,
    drop_yellow = TRUE, drop_black = FALSE) %>%
  # B0251 = Hashtag1, B0260 = Hashtag2
  setNames(0:6)
# palette for sample names
pal.sampleName <-
  pal.sampleBarcode[str_detect(names(pal.sampleBarcode), "HTO")] %>%
  setNames(str_replace(names(.), "HTO-M", "mouse"))

pal.treatmentName <- pal.sampleName
pal.treatmentName <- sub("mouse1", "Untreated WT", pal.treatmentName)
pal.treatmentName <- sub("mouse2", "IL-2", pal.treatmentName)
pal.treatmentName <- sub("mouse3", "IL-2 Mutein", pal.treatmentName)

# palette for isotypes
pal.isotype <-
  big_colorblind_pal(2) %>%
  setNames(c("TRbc2", "TRbc1")) %>%
  c("undetermined" = "grey50")

pal.cycle <-
      big_colorblind_pal(3) %>%
  setNames(c("G1", "G2M", "S")) 

# color palette for plotting gene counts, e.g. on tSNE and UMAP plots
cols.geneCounts <- c("gray90", "darkred")
cols.geneCountsDiverging <- c("blue", "gray90", "red")
```

## Save data loaded upstream of 10x data matrices

The 10x data to be loaded immediately after this step include very large data matrices. It's easiest to save the data loaded up to this point, so that they can be easily loaded for later use.

```{r saveProjectDataEx10x, dependson="setupPalettes"}
filenameDataEx10x <-
  file.path("data_saved", paste0("dataEx10x.", filenameSuffix, ".RData"))

save(list = ls(), file = filenameDataEx10x)
# load(file = filenameDataEx10x)
```

## Load 10x data matrix

```{r loadData10x, dependson="getProjectInfo"}
filePool_410xUnfiltered <-
  file.path(dataDir, 
            paste0("Pool410x", 
                   filenameSuffix, ".RDS"))

if (file.exists(filenameData10xUnfiltered)) {
  data10xUnfiltered <- readRDS(filenameData10xUnfiltered)
} else {
  # set counts filename
  filePool_410xUnfiltered <-
    file.path(
      "/Volumes", "Bioinformatics", "pipeline", "tenxSummaries",
      tenxSummaryDir, "multi_per_sample",
      paste0("Pool_4", "_sample_feature_bc_matrix.h5"))
  
  inputData10x <-
    Seurat::Read10X_h5("/Volumes/Bioinformatics/pipeline/tenxSummaries/220413-P471-1allFlowcells/multi_per_sample/Pool_4_sample_feature_bc_matrix.h5")
  
  # filter gene expression data to include only protein-coding and [TB]CR genes
  biotypeRegex.tmp <-
    "(protein_coding)|(IG_[CVDJ]_gene)|(TR_[CVDJ]_gene)"
  `Pool_410x.P471-1_10x_analysis.2022-04-13`                          $`Gene Expression` <-
    inputData10x$`Gene Expression`[
      rownames(inputData10x$`Gene Expression`) %in% 
        (annotables::grcm38 %>%
           dplyr::filter(str_detect(biotype, biotypeRegex.tmp)) %>%
           dplyr::pull(symbol)),]
  
  # create Seurat object with RNA and antibody-derived tag data
  Pool_410xUnfiltered <-
    CreateSeuratObject(counts = inputData10x$`Gene Expression`)
  Pool_410xUnfiltered[["HTO"]] <- CreateAssayObject(counts = inputData10x$`Antibody Capture`)
  
  # also store the cellBarcode information, in case it gets lost from the rownames of the meta.data slot
  Pool_410xUnfiltered <- Pool_410xUnfiltered %>%
    AddMetaData(
      metadata = colnames(Pool_410xUnfiltered),
      col.name = "cellBarcode")

  saveRDS(`Pool_410x.P471-1_10x_analysis.2022-04-13`, file = filePool_410xUnfiltered)
  
  rm(inputData10x)
}

rm_tmp(ask=FALSE)
```

```{r loadData10xVdj, dependson="loadData10x"}
# load contigs and rearrangements
filenamePool1VdjUnfiltered <-
  file.path(dirRoot, 
            paste0("Pool1VdjUnfiltered.", 
                   filenameSuffix, ".RDS"))

if (file.exists(filenamePool1VdjUnfiltered)) {
  data10xVdjUnfiltered <- readRDS(filenamePool110xVdjUnfiltered)
} else {
  # set input filename
  filenameInputData10xVdj <-
    file.path(
      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
      tenxSummaryDir, "multi_per_sample",
      paste0("Pool_1", "_filtered_contig_annotations.csv"))
  Pool1VdjUnfiltered <-
    read.csv(filenameInputData10xVdj) %>%
    # store cell barcode
    dplyr::rename(cellBarcode = barcode,
                  cellBarcodeContig = contig_id) %>%
    # store full variable region nucleotide sequence
    dplyr::mutate(
      varRegionSequence = paste0(fwr1_nt, cdr1_nt, fwr2_nt, cdr2_nt, fwr3_nt, cdr3_nt, fwr4_nt),
      is_cell = is_cell == "true",
      full_length = full_length == "true",
      productive = productive == "true",
      contigNumPerCell = str_extract(cellBarcodeContig, "contig_[0-9]+"))
  
  # previously pulled contig sequences from filtered_contig.fasta, but those are a subset of the information in the AIRR rearrangements file read in below
  
#   pull in AIRR rearrangements, which includes a bunch of useful stuff
#  filenameInputData10xVdjRearrangements <-
#    file.path(
#      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
#      tenxSummaryDir, "multi_per_sample",
#      paste0("Pool_1", "_filtered_contig_annotations.csv"))
#  Pool1VdjRearrangements.tmp <-
#    readr::read_delim(
#      filenameInputData10xVdjRearrangements) %>%
#    dplyr::rename(
#      cellBarcode = barcode,
#      raw_clonotype_id = clone_id,
#      cellBarcodeContig = contig_id,
#      reads = consensus_count,
#      umis = duplicate_count
#    )
  
  Pool1VdjUnfiltered$high_confidence <- as.logical(toupper(Pool1VdjUnfiltered$high_confidence))
  
  # merge contig annotations and AIRR rearrangements
#  Pool1VdjUnfiltered <-
#    Pool1VdjUnfiltered %>%
#    full_join(Pool1VdjRearrangements.tmp)
  
  saveRDS(Pool1VdjUnfiltered, file = filenamePool1VdjUnfiltered)
}

# load clonotype assignment data
filenamePool1VdjClonotypes <-
  file.path(dirRoot, 
            paste0("Pool1VdjClonotypes.", 
                   filenameSuffix, ".RDS"))

if (file.exists(filenamePool1VdjClonotypes)) {
  Pool1VdjClonotypes <- readRDS(filenamePool1VdjClonotypes)
} else {
  # set input filename
  filenameInputData10xVdjClonotypes <-
    file.path(
      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
      tenxSummaryDir, "multi_per_sample",
      paste0("Pool_1", "_clonotypes.csv"))

  # pull in clonotype assignments
  Pool1VdjClonotypes <-
    readr::read_csv(filenameInputData10xVdjClonotypes)
  saveRDS(Pool1VdjClonotypes, file = filenamePool1VdjClonotypes)
} 
  
rm_tmp(ask=FALSE)

# load contigs and rearrangements
filenamePool2VdjUnfiltered <-
  file.path(dirRoot, 
            paste0("Pool2VdjUnfiltered.", 
                   filenameSuffix, ".RDS"))

if (file.exists(filenamePool2VdjUnfiltered)) {
  data10xVdjUnfiltered <- readRDS(filenamePool2VdjUnfiltered)
} else {
  # set input filename
  filenameInputData10xVdj <-
    file.path(
      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
      tenxSummaryDir, "multi_per_sample",
      paste0("Pool_2", "_filtered_contig_annotations.csv"))
  Pool2VdjUnfiltered <-
    read.csv(filenameInputData10xVdj) %>%
    # store cell barcode
    dplyr::rename(cellBarcode = barcode,
                  cellBarcodeContig = contig_id) %>%
    # store full variable region nucleotide sequence
    dplyr::mutate(
      varRegionSequence = paste0(fwr1_nt, cdr1_nt, fwr2_nt, cdr2_nt, fwr3_nt, cdr3_nt, fwr4_nt),
      is_cell = is_cell == "true",
      full_length = full_length == "true",
      productive = productive == "true",
      contigNumPerCell = str_extract(cellBarcodeContig, "contig_[0-9]+"))
  
  # previously pulled contig sequences from filtered_contig.fasta, but those are a subset of the information in the AIRR rearrangements file read in below
  
  # pull in AIRR rearrangements, which includes a bunch of useful stuff
#  filenameInputData10xVdjRearrangements <-
#    file.path(
#      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
#      tenxSummaryDir, "multi_per_sample",
#      paste0("Pool_2", "_filtered_contig_annotations.csv"))
#  Pool2VdjRearrangements.tmp <-
#    readr::read_delim(
#      filenameInputData10xVdjRearrangements) %>%
#    dplyr::rename(
#      cellBarcode = barcode,
#      #raw_clonotype_id = clone_id,
#      cellBarcodeContig = contig_id,
#      #reads = consensus_count,
#      #umis = duplicate_count
#    )
  
  Pool2VdjUnfiltered$high_confidence <- as.logical(toupper(Pool2VdjUnfiltered$high_confidence))
  
#  # merge contig annotations and AIRR rearrangements
#  Pool2VdjUnfiltered <-
#    Pool2VdjUnfiltered %>%
#    full_join(Pool2VdjRearrangements.tmp)
  
  saveRDS(Pool2VdjUnfiltered, file = filenamePool2VdjUnfiltered)
}

# load clonotype assignment data
filenamePool2VdjClonotypes <-
  file.path(dirRoot, 
            paste0("Pool2VdjClonotypes.", 
                   filenameSuffix, ".RDS"))

if (file.exists(filenamePool2VdjClonotypes)) {
  Pool2VdjClonotypes <- readRDS(filenamePool2VdjClonotypes)
} else {
  # set input filename
  filenameInputData10xVdjClonotypes <-
    file.path(
      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
      tenxSummaryDir, "multi_per_sample",
      paste0("Pool_2", "_clonotypes.csv"))

  # pull in clonotype assignments
  Pool2VdjClonotypes <-
    readr::read_csv(filenameInputData10xVdjClonotypes)
  saveRDS(Pool2VdjClonotypes, file = filenamePool2VdjClonotypes)
}
  
rm_tmp(ask=FALSE)

filenamePool3VdjUnfiltered <-
  file.path(dirRoot, 
            paste0("Pool3VdjUnfiltered.", 
                   filenameSuffix, ".RDS"))

if (file.exists(filenamePool3VdjUnfiltered)) {
  data10xVdjUnfiltered <- readRDS(filenamePool3VdjUnfiltered)
} else {
  # set input filename
  filenameInputData10xVdj <-
    file.path(
      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
      tenxSummaryDir, "multi_per_sample",
      paste0("Pool_3", "_filtered_contig_annotations.csv"))
  Pool3VdjUnfiltered <-
    read.csv(filenameInputData10xVdj) %>%
    # store cell barcode
    dplyr::rename(cellBarcode = barcode,
                  cellBarcodeContig = contig_id) %>%
    # store full variable region nucleotide sequence
    dplyr::mutate(
      varRegionSequence = paste0(fwr1_nt, cdr1_nt, fwr2_nt, cdr2_nt, fwr3_nt, cdr3_nt, fwr4_nt),
      is_cell = is_cell == "true",
      full_length = full_length == "true",
      productive = productive == "true",
      contigNumPerCell = str_extract(cellBarcodeContig, "contig_[0-9]+"))
  
  # previously pulled contig sequences from filtered_contig.fasta, but those are a subset of the information in the AIRR rearrangements file read in below
  
#  # pull in AIRR rearrangements, which includes a bunch of useful stuff
#  filenameInputData10xVdjRearrangements <-
#    file.path(
#      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
#      tenxSummaryDir, "multi_per_sample",
#      paste0("Pool_3", "_filtered_contig_annotations.csv"))
#  Pool3VdjRearrangements.tmp <-
#    readr::read_delim(
#      filenameInputData10xVdjRearrangements) %>%
#    dplyr::rename(
#      cellBarcode = barcode,
#      #raw_clonotype_id = clone_id,
#      cellBarcodeContig = contig_id,
#      #reads = consensus_count,
#      #umis = duplicate_count
#   )
  
  Pool3VdjUnfiltered$high_confidence <- toupper(Pool3VdjUnfiltered$high_confidence) %>% as.logical()
  
  # merge contig annotations and AIRR rearrangements
#  Pool3VdjUnfiltered <-
#    Pool3VdjUnfiltered %>%
#    full_join(Pool3VdjRearrangements.tmp)
  
  saveRDS(Pool3VdjUnfiltered, file = filenamePool3VdjUnfiltered)
}

# load clonotype assignment data
filenamePool3VdjClonotypes <-
  file.path(dirRoot, 
            paste0("Pool3VdjClonotypes.", 
                   filenameSuffix, ".RDS"))

if (file.exists(filenamePool2VdjClonotypes)) {
  Pool3VdjClonotypes <- readRDS(filenamePool3VdjClonotypes)
} else {
  # set input filename
  filenameInputData10xVdjClonotypes <-
    file.path(
      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
      tenxSummaryDir, "multi_per_sample",
      paste0("Pool_3", "_clonotypes.csv"))

  # pull in clonotype assignments
  Pool3VdjClonotypes <-
    readr::read_csv(filenameInputData10xVdjClonotypes)
  saveRDS(Pool3VdjClonotypes, file = filenamePool3VdjClonotypes)
}
  
rm_tmp(ask=FALSE)

filenamePool4VdjUnfiltered <-
  file.path(dirRoot, 
            paste0("Pool4VdjUnfiltered.", 
                   filenameSuffix, ".RDS"))

if (file.exists(filenamePool4VdjUnfiltered)) {
  data10xVdjUnfiltered <- readRDS(filenamePool4VdjUnfiltered)
} else {
  # set input filename
  filenameInputData10xVdj <-
    file.path(
      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
      tenxSummaryDir, "multi_per_sample",
      paste0("Pool_4", "_filtered_contig_annotations.csv"))
  Pool4VdjUnfiltered <-
    read.csv(filenameInputData10xVdj) %>%
    # store cell barcode
    dplyr::rename(cellBarcode = barcode,
                  cellBarcodeContig = contig_id) %>%
    # store full variable region nucleotide sequence
    dplyr::mutate(
      varRegionSequence = paste0(fwr1_nt, cdr1_nt, fwr2_nt, cdr2_nt, fwr3_nt, cdr3_nt, fwr4_nt),
      is_cell = is_cell == "true",
      full_length = full_length == "true",
      productive = productive == "true",
      contigNumPerCell = str_extract(cellBarcodeContig, "contig_[0-9]+"))
  
  # previously pulled contig sequences from filtered_contig.fasta, but those are a subset of the information in the AIRR rearrangements file read in below
  
  # pull in AIRR rearrangements, which includes a bunch of useful stuff
  filenameInputData10xVdjRearrangements <-
    file.path(
      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
      tenxSummaryDir, "multi_per_sample",
      paste0("Pool_4", "_filtered_contig_annotations.csv"))
  Pool4VdjRearrangements.tmp <-
    readr::read_delim(
      filenameInputData10xVdjRearrangements) %>%
    dplyr::rename(
      cellBarcode = barcode,
      #raw_clonotype_id = clone_id,
      cellBarcodeContig = contig_id,
      #reads = consensus_count,
      #umis = duplicate_count
    )
  
  Pool4VdjUnfiltered$high_confidence <- toupper(Pool4VdjUnfiltered$high_confidence) %>% as.logical()
  
  # merge contig annotations and AIRR rearrangements
  Pool4VdjUnfiltered <-
    Pool4VdjUnfiltered %>%
    full_join(Pool4VdjRearrangements.tmp)
  
  saveRDS(Pool4VdjUnfiltered, file = filenamePool4VdjUnfiltered)
}

# load clonotype assignment data
filenamePool4VdjClonotypes <-
  file.path(dirRoot, 
            paste0("Pool4VdjClonotypes.", 
                   filenameSuffix, ".RDS"))

if (file.exists(filenamePool4VdjClonotypes)) {
  Pool4VdjClonotypes <- readRDS(filenamePool4VdjClonotypes)
} else {
  # set input filename
  filenameInputData10xVdjClonotypes <-
    file.path(
      "/Volumes", "bioinformatics", "pipeline", "tenxSummaries",
      tenxSummaryDir, "multi_per_sample",
      paste0("Pool_4", "_clonotypes.csv"))

  # pull in clonotype assignments
  Pool4VdjClonotypes <-
    readr::read_csv(filenameInputData10xVdjClonotypes)
  saveRDS(Pool4VdjClonotypes, file = filenamePool4VdjClonotypes)
}
  
rm_tmp(ask=FALSE)
```

```{r setSubsetDataDivisor}
subsetDataDivisor <- 100
```

```{r subsetDataForCodePrototyping, dependson="loadData10X", eval=FALSE}
# Need to remove this or set eval=FALSE for final runs

cellsIncludeSubsetData <-
  colnames(data10xUnfiltered)[
    seq(from = 1, to = ncol(data10xUnfiltered), by = subsetDataDivisor)]
data10xUnfiltered <-
  subset(data10xUnfiltered, cells = cellsIncludeSubsetData)
```

Some 10x datasets are large enough that it is difficult to do exploratory analyses, as small steps take a large amount of time. We may work with a subsetted dataset while piloting the analysis, then run later on the full dataset once we have some of the methods more fully worked out. We use `r 1/subsetDataDivisor` of the data for these test runs.

## Load gene sets

We may want to use gene sets to summarize expression related to specific pathways. Here we load the MSigDB Hallmark gene sets. We may pull in additional sets in the future.

```{r loadGeneSets, dependson="setFilenamesAndGlobalVariables"}
```

# Quality control of cells using RNAseq data

## Quality metrics

We use several metrics to identify high-quality cells:
1) Total UMI counts in each cell
2) Percent mitochondrial reads for each cell
3) Percent ribosomal reads for each cell

```{r calculateQcMetrics, dependson="loadData10x"}
`Pool_110x.P471-1_10x_analysis.2022-04-13` <- 
  PercentageFeatureSet(`Pool_110x.P471-1_10x_analysis.2022-04-13`, "^mt-", col.name = "percent_mito") # percent mitochondrial reads
`Pool_110x.P471-1_10x_analysis.2022-04-13` <-
  PercentageFeatureSet(`Pool_110x.P471-1_10x_analysis.2022-04-13`, "^Rp[sl]", col.name = "percent_ribo") # percent ribosomal protein reads
`Pool_110x.P471-1_10x_analysis.2022-04-13` <-
  PercentageFeatureSet(`Pool_110x.P471-1_10x_analysis.2022-04-13`, "^Hb[^(p)]", col.name = "percent_hb") # percent hemoglobin reads

`Pool_210x.P471-1_10x_analysis.2022-04-13` <- 
  PercentageFeatureSet(`Pool_210x.P471-1_10x_analysis.2022-04-13`, "^mt-", col.name = "percent_mito") # percent mitochondrial reads
`Pool_210x.P471-1_10x_analysis.2022-04-13` <-
  PercentageFeatureSet(`Pool_210x.P471-1_10x_analysis.2022-04-13`, "^Rp[sl]", col.name = "percent_ribo") # percent ribosomal protein reads
`Pool_210x.P471-1_10x_analysis.2022-04-13` <-
  PercentageFeatureSet(`Pool_210x.P471-1_10x_analysis.2022-04-13`, "^Hb[^(p)]", col.name = "percent_hb") # percent hemoglobin reads

`Pool_310x.P471-1_10x_analysis.2022-04-13` <- 
  PercentageFeatureSet(`Pool_310x.P471-1_10x_analysis.2022-04-13`, "^mt-", col.name = "percent_mito") # percent mitochondrial reads
`Pool_310x.P471-1_10x_analysis.2022-04-13` <-
  PercentageFeatureSet(`Pool_310x.P471-1_10x_analysis.2022-04-13`, "^Rp[sl]", col.name = "percent_ribo") # percent ribosomal protein reads
`Pool_310x.P471-1_10x_analysis.2022-04-13` <-
  PercentageFeatureSet(`Pool_310x.P471-1_10x_analysis.2022-04-13`, "^Hb[^(p)]", col.name = "percent_hb") # percent hemoglobin reads

`Pool_410x.P471-1_10x_analysis.2022-04-13` <- 
  PercentageFeatureSet(`Pool_410x.P471-1_10x_analysis.2022-04-13`, "^mt-", col.name = "percent_mito") # percent mitochondrial reads
`Pool_410x.P471-1_10x_analysis.2022-04-13` <-
  PercentageFeatureSet(`Pool_410x.P471-1_10x_analysis.2022-04-13`, "^Rp[sl]", col.name = "percent_ribo") # percent ribosomal protein reads
`Pool_410x.P471-1_10x_analysis.2022-04-13` <-
  PercentageFeatureSet(`Pool_410x.P471-1_10x_analysis.2022-04-13`, "^Hb[^(p)]", col.name = "percent_hb") # percent hemoglobin reads
```

```{r setQcMetricsThresholds}
qcMetricsThresholds <-
  c(
    "min_nFeature_RNA" = 250,
    "max_nFeature_RNA" = 6000,
    "min_nCount_RNA" = 450,
    "max_nCount_RNA" = 35000,
    # "nFeature_ADT" = 50,
    # "nCount_ADT" = 500,
    "percent_mito" = 50,
    "percent_ribo" = 50,
    "percent_hb" = 10
  )

qcMetricsDfForPlot <-
  data.frame(
    metric = str_replace(names(qcMetricsThresholds), "^(min|max)_", ""),
    threshold = unname(qcMetricsThresholds)
  )

## plots to determine thresholds
# # nFeature_RNA
# ggplot(data10xUnfiltered@meta.data, mapping = aes(x = nFeature_RNA)) +
#   # geom_density(mapping = aes(color = sampleId), size = 0.5) +
#   geom_histogram(bins=200) +
#   scale_x_log10() +
#   geom_vline(xintercept = qcMetricsThresholds["min_nFeature_RNA"],
#              color = "red", linetype = "dashed") +
#   geom_vline(xintercept = qcMetricsThresholds["max_nFeature_RNA"],
#              color = "red", linetype = "dashed")
# 
# # nCount_RNA
# ggplot(data10xUnfiltered@meta.data, mapping = aes(x = nCount_RNA)) +
#   # geom_density(mapping = aes(color = sampleId), size = 0.5) +
#   geom_histogram(bins=200) +
#   scale_x_log10() +
#   geom_vline(xintercept = qcMetricsThresholds["min_nCount_RNA"],
#              color = "red", linetype = "dashed") +
#   geom_vline(xintercept = qcMetricsThresholds["max_nCount_RNA"],
#              color = "red", linetype = "dashed")
# 
# # nFeature_ADT
# ggplot(data10xUnfiltered@meta.data, mapping = aes(x = nFeature_ADT)) +
#   geom_density(mapping = aes(color = sampleId), size = 0.5) +
#   # geom_histogram(bins=200) +
#   # scale_x_log10() +
#   geom_vline(xintercept = qcMetricsThresholds["nFeature_ADT"],
#              color = "red", linetype = "dashed")
#
# # nCount_ADT
# ggplot(data10xUnfiltered@meta.data, mapping = aes(x = nCount_ADT)) +
#   geom_density(mapping = aes(color = sampleId), size = 0.5) +
#   # geom_histogram(bins=200) +
#   scale_x_log10() +
#   geom_vline(xintercept = qcMetricsThresholds["nCount_ADT"],
#              color = "red", linetype = "dashed")
```

```{r plotQcMetricsTall, dependson="calculateQcMetrics", fig.width=8, fig.height=12}
qcMetrics.tmp <-
  c("nFeature_RNA", "nCount_RNA",
    # "nFeature_ADT", "nCount_ADT",
    "percent_mito", "percent_ribo", "percent_hb")

# Seurat version; the manual version below allows more control
# plot.tmp <-
#   VlnPlot(data10xUnfiltered, features = qcMetrics.tmp,
#           pt.size = 0.01, ncol = 2) +
#   NoLegend()
# color and modifying themes don't work very well with these ggplot2 patchwork objects

# alternate manual approach
plot.tmp <-
  `Pool_110x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(
    cols = one_of(qcMetrics.tmp),
    names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin() +
  ggrastr::rasterise(
    geom_jitter(size = 0.5), dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = paste0("plotQcMetricsTallPool1"),
  format = plotOutputFormat,
  w = 8, h = 10)


plot.tmp <-
  `Pool_210x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(
    cols = one_of(qcMetrics.tmp),
    names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin() +
  ggrastr::rasterise(
    geom_jitter(size = 0.5), dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = paste0("plotQcMetricsTallPool2"),
  format = plotOutputFormat,
  w = 8, h = 10)

plot.tmp <-
  `Pool_310x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(
    cols = one_of(qcMetrics.tmp),
    names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin() +
  ggrastr::rasterise(
    geom_jitter(size = 0.5), dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = paste0("plotQcMetricsTallPool3"),
  format = plotOutputFormat,
  w = 8, h = 10)

rm_tmp(ask=FALSE)

plot.tmp <-
  `Pool_410x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(
    cols = one_of(qcMetrics.tmp),
    names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin() +
  ggrastr::rasterise(
    geom_jitter(size = 0.5), dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = paste0("plotQcMetricsTallPool4"),
  format = plotOutputFormat,
  w = 8, h = 10)

rm_tmp(ask=FALSE)
```

```{r plotQcMetrics, dependson="calculateQcMetrics", fig.width=12, fig.height=7}
qcMetrics.tmp <-
  c("nFeature_RNA", "nCount_RNA",
    # "nFeature_ADT", "nCount_ADT",
    "percent_mito", "percent_ribo", "percent_hb")

# # Seurat version
# plot.tmp <-
#   VlnPlot(data10xUnfiltered, group.by = "sampleId", features = qcMetrics.tmp, 
#           pt.size = 0.01, ncol = 4) +
#   NoLegend()
# # color and modifying themes don't work very well with these ggplot2 patchwork objects

# alternate manual approach
plot.tmp <-
  `Pool_110x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(cols = one_of(qcMetrics.tmp), names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin(col = "blue") +
  ggrastr::rasterise(
    geom_jitter(size = 0.5),
    dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 3, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = paste0("plotQcMetricsPool1.pdf"),
  format = plotOutputFormat,
  w = 12, h = 7)


plot.tmp <-
  `Pool_210x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(cols = one_of(qcMetrics.tmp), names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin(col = "blue") +
  ggrastr::rasterise(
    geom_jitter(size = 0.5),
    dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 3, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = paste0("plotQcMetricsPool2.pdf"),
  format = plotOutputFormat,
  w = 12, h = 7)

plot.tmp <-
  `Pool_310x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(cols = one_of(qcMetrics.tmp), names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin(col = "blue") +
  ggrastr::rasterise(
    geom_jitter(size = 0.5),
    dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 3, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = paste0("plotQcMetricsPool3.pdf"),
  format = plotOutputFormat,
  w = 12, h = 7)

plot.tmp <-
  `Pool_410x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(cols = one_of(qcMetrics.tmp), names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin(col = "blue") +
  ggrastr::rasterise(
    geom_jitter(size = 0.5),
    dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 3, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = paste0("plotQcMetricsPool4.pdf"),
  format = plotOutputFormat,
  w = 12, h = 7)

rm_tmp(ask=FALSE)
```

```{r setNonBCellGeneThresholds}
nonBCellGeneThresholds <-
  c(
    "Cd3d" = 3,
    "Cd3e" = 3,
    "Vil1" = 20,
    "Reg3g" = 5
  )

nonBCellGeneThresholdsDfForPlot <-
  data.frame(
    gene = names(nonBCellGeneThresholds),
    threshold = unname(nonBCellGeneThresholds)
  )
```

```{r plotExpressionNonBCellGenes, dependson=c("setNonBCellGeneThresholds", "calculateQcMetrics")}
genesToPlot.tmp <-
  c("Cd3d", "Cd3e", # T cell genes
    "Reg3g", "Vil1" # non-lymphoid cell genes
  )

# alternate manual approach
plot.tmp <-
  `Pool_110x.P471-1_10x_analysis.2022-04-13`@assays$RNA@counts[genesToPlot.tmp,] %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  pivot_longer(
    cols = everything(), names_to = "gene", values_to = "count") %>%
  mutate(gene = factor(gene, levels = genesToPlot.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = count)) +
  geom_violin(col = "blue") +
  ggrastr::rasterise(
    geom_jitter(size = 0.5),
    dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(gene), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      nonBCellGeneThresholdsDfForPlot %>%
      dplyr::filter(gene %in% genesToPlot.tmp) %>%
      mutate(gene = gene %>% factor(levels = genesToPlot.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = "plotExpressionTcellPool1",
  format = plotOutputFormat,
  w = 8.5, h = 7)

plot.tmp <-
  `Pool_210x.P471-1_10x_analysis.2022-04-13`@assays$RNA@counts[genesToPlot.tmp,] %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  pivot_longer(
    cols = everything(), names_to = "gene", values_to = "count") %>%
  mutate(gene = factor(gene, levels = genesToPlot.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = count)) +
  geom_violin(col = "blue") +
  ggrastr::rasterise(
    geom_jitter(size = 0.5),
    dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(gene), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      nonBCellGeneThresholdsDfForPlot %>%
      dplyr::filter(gene %in% genesToPlot.tmp) %>%
      mutate(gene = gene %>% factor(levels = genesToPlot.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = "plotExpressionTcellPool2",
  format = plotOutputFormat,
  w = 8.5, h = 7)

plot.tmp <-
  `Pool_310x.P471-1_10x_analysis.2022-04-13`@assays$RNA@counts[genesToPlot.tmp,] %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  pivot_longer(
    cols = everything(), names_to = "gene", values_to = "count") %>%
  mutate(gene = factor(gene, levels = genesToPlot.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = count)) +
  geom_violin(col = "blue") +
  ggrastr::rasterise(
    geom_jitter(size = 0.5),
    dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(gene), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      nonBCellGeneThresholdsDfForPlot %>%
      dplyr::filter(gene %in% genesToPlot.tmp) %>%
      mutate(gene = gene %>% factor(levels = genesToPlot.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = "plotExpressionTcellPool3",
  format = plotOutputFormat,
  w = 8.5, h = 7)

plot.tmp <-
  `Pool_410x.P471-1_10x_analysis.2022-04-13`@assays$RNA@counts[genesToPlot.tmp,] %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  pivot_longer(
    cols = everything(), names_to = "gene", values_to = "count") %>%
  mutate(gene = factor(gene, levels = genesToPlot.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = count)) +
  geom_violin(col = "blue") +
  ggrastr::rasterise(
    geom_jitter(size = 0.5),
    dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(gene), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      nonBCellGeneThresholdsDfForPlot %>%
      dplyr::filter(gene %in% genesToPlot.tmp) %>%
      mutate(gene = gene %>% factor(levels = genesToPlot.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = "plotExpressionTcellPool4",
  format = plotOutputFormat,
  w = 8.5, h = 7)

rm_tmp(ask=FALSE)
```

```{r filterCellsQcMetrics, dependson=c("setQcMetricsThresholds", "calculateQcMetrics")}
cellsKeepPool1 <-
  WhichCells(
    `Pool_110x.P471-1_10x_analysis.2022-04-13`,
    expression = nFeature_RNA >= qcMetricsThresholds["min_nFeature_RNA"]) %>% # keeps 2841 of 2984
  intersect(
    WhichCells(
      `Pool_110x.P471-1_10x_analysis.2022-04-13`,
      expression = nFeature_RNA <= qcMetricsThresholds["max_nFeature_RNA"])) %>% # keeps 2771 of 2841
  intersect(
    WhichCells(
      `Pool_110x.P471-1_10x_analysis.2022-04-13`,
      expression = nCount_RNA >= qcMetricsThresholds["min_nCount_RNA"])) %>% # keeps 2761 of 2771
  intersect(
    WhichCells(
      `Pool_110x.P471-1_10x_analysis.2022-04-13`,
      expression = nCount_RNA <= qcMetricsThresholds["max_nCount_RNA"])) %>% # keeps 2722 of 2761
  # intersect(
  #   WhichCells(
  #     data10xUnfiltered, expression = nFeature_ADT >= qcMetricsThresholds["nFeature_ADT"])) %>% # keeps ? of ?
  # intersect(
  #   WhichCells(
  #     data10xUnfiltered, expression = nCount_ADT >= qcMetricsThresholds["nCount_ADT"])) %>% # keeps ? of ?
  intersect(
    WhichCells(
      `Pool_110x.P471-1_10x_analysis.2022-04-13`, expression = percent_mito <= qcMetricsThresholds["percent_mito"])) %>% # keeps 2701 of 2722
  intersect(
    WhichCells(
      `Pool_110x.P471-1_10x_analysis.2022-04-13`, expression = percent_ribo <= qcMetricsThresholds["percent_ribo"])) %>% # keeps 2701 of 2701
  intersect(
    WhichCells(
      `Pool_110x.P471-1_10x_analysis.2022-04-13`, expression = percent_hb <= qcMetricsThresholds["percent_hb"])) %>% # keeps 2701 of 2701
  #intersect(
    # drop T cells based on Cd3d
    #WhichCells(
      #data10xUnfiltered,
      #expression = Cd3d < nonBCellGeneThresholds[["Cd3d"]])) %>% # keeps 2697 of 2701
   #intersect(
    # drop T cells based on Cd3e
    #WhichCells(
      #`Pool_110x.P471-1_10x_analysis.2022-04-13`,
      #expression = Cd3e  nonBCellGeneThresholds[["Cd3e"]])) %>% # keeps 2697 of 2697
  intersect(
     #drop other non-B cells based on Vil1
    WhichCells(
      `Pool_110x.P471-1_10x_analysis.2022-04-13`,
      expression = Vil1 < nonBCellGeneThresholds[["Vil1"]])) %>% # keeps 2695 of 2697
 intersect(
    # drop other non-B cells based on Vil1
      WhichCells(
      `Pool_110x.P471-1_10x_analysis.2022-04-13`,
      expression = Reg3g < nonBCellGeneThresholds[["Reg3g"]])) # keeps 2691 of 2695

cellsDropPool1 <-
  setdiff(colnames(`Pool_110x.P471-1_10x_analysis.2022-04-13`), cellsKeepPool1)

Pool1InclMultiplets <-
  subset(`Pool_110x.P471-1_10x_analysis.2022-04-13`, cells = cellsKeepPool1)

cellsKeepPool2 <-
  WhichCells(
    `Pool_210x.P471-1_10x_analysis.2022-04-13`,
    expression = nFeature_RNA >= qcMetricsThresholds["min_nFeature_RNA"]) %>% # keeps 2841 of 2984
  intersect(
    WhichCells(
      `Pool_210x.P471-1_10x_analysis.2022-04-13`,
      expression = nFeature_RNA <= qcMetricsThresholds["max_nFeature_RNA"])) %>% # keeps 2771 of 2841
  intersect(
    WhichCells(
      `Pool_210x.P471-1_10x_analysis.2022-04-13`,
      expression = nCount_RNA >= qcMetricsThresholds["min_nCount_RNA"])) %>% # keeps 2761 of 2771
  intersect(
    WhichCells(
      `Pool_210x.P471-1_10x_analysis.2022-04-13`,
      expression = nCount_RNA <= qcMetricsThresholds["max_nCount_RNA"])) %>% # keeps 2722 of 2761
  # intersect(
  #   WhichCells(
  #     data10xUnfiltered, expression = nFeature_ADT >= qcMetricsThresholds["nFeature_ADT"])) %>% # keeps ? of ?
  # intersect(
  #   WhichCells(
  #     data10xUnfiltered, expression = nCount_ADT >= qcMetricsThresholds["nCount_ADT"])) %>% # keeps ? of ?
  intersect(
    WhichCells(
      `Pool_210x.P471-1_10x_analysis.2022-04-13`, expression = percent_mito <= qcMetricsThresholds["percent_mito"])) %>% # keeps 2701 of 2722
  intersect(
    WhichCells(
      `Pool_210x.P471-1_10x_analysis.2022-04-13`, expression = percent_ribo <= qcMetricsThresholds["percent_ribo"])) %>% # keeps 2701 of 2701
  intersect(
    WhichCells(
      `Pool_210x.P471-1_10x_analysis.2022-04-13`, expression = percent_hb <= qcMetricsThresholds["percent_hb"])) %>% # keeps 2701 of 2701
  #intersect(
    # drop T cells based on Cd3d
    #WhichCells(
      #data10xUnfiltered,
      #expression = Cd3d < nonBCellGeneThresholds[["Cd3d"]])) %>% # keeps 2697 of 2701
   #intersect(
    # drop T cells based on Cd3e
    #WhichCells(
      #`Pool_110x.P471-1_10x_analysis.2022-04-13`,
      #expression = Cd3e  nonBCellGeneThresholds[["Cd3e"]])) %>% # keeps 2697 of 2697
  intersect(
     #drop other non-B cells based on Vil1
    WhichCells(
      `Pool_210x.P471-1_10x_analysis.2022-04-13`,
      expression = Vil1 < nonBCellGeneThresholds[["Vil1"]])) %>% # keeps 2695 of 2697
 intersect(
    # drop other non-B cells based on Vil1
      WhichCells(
      `Pool_210x.P471-1_10x_analysis.2022-04-13`,
      expression = Reg3g < nonBCellGeneThresholds[["Reg3g"]])) # keeps 2691 of 2695

cellsDropPool2 <-
  setdiff(colnames(`Pool_210x.P471-1_10x_analysis.2022-04-13`), cellsKeepPool2)

Pool2InclMultiplets <-
  subset(`Pool_210x.P471-1_10x_analysis.2022-04-13`, cells = cellsKeepPool2)

cellsKeepPool3 <-
  WhichCells(
    `Pool_310x.P471-1_10x_analysis.2022-04-13`,
    expression = nFeature_RNA >= qcMetricsThresholds["min_nFeature_RNA"]) %>% # keeps 2841 of 2984
  intersect(
    WhichCells(
      `Pool_310x.P471-1_10x_analysis.2022-04-13`,
      expression = nFeature_RNA <= qcMetricsThresholds["max_nFeature_RNA"])) %>% # keeps 2771 of 2841
  intersect(
    WhichCells(
      `Pool_310x.P471-1_10x_analysis.2022-04-13`,
      expression = nCount_RNA >= qcMetricsThresholds["min_nCount_RNA"])) %>% # keeps 2761 of 2771
  intersect(
    WhichCells(
      `Pool_310x.P471-1_10x_analysis.2022-04-13`,
      expression = nCount_RNA <= qcMetricsThresholds["max_nCount_RNA"])) %>% # keeps 2722 of 2761
  # intersect(
  #   WhichCells(
  #     data10xUnfiltered, expression = nFeature_ADT >= qcMetricsThresholds["nFeature_ADT"])) %>% # keeps ? of ?
  # intersect(
  #   WhichCells(
  #     data10xUnfiltered, expression = nCount_ADT >= qcMetricsThresholds["nCount_ADT"])) %>% # keeps ? of ?
  intersect(
    WhichCells(
      `Pool_310x.P471-1_10x_analysis.2022-04-13`, expression = percent_mito <= qcMetricsThresholds["percent_mito"])) %>% # keeps 2701 of 2722
  intersect(
    WhichCells(
      `Pool_310x.P471-1_10x_analysis.2022-04-13`, expression = percent_ribo <= qcMetricsThresholds["percent_ribo"])) %>% # keeps 2701 of 2701
  intersect(
    WhichCells(
      `Pool_310x.P471-1_10x_analysis.2022-04-13`, expression = percent_hb <= qcMetricsThresholds["percent_hb"])) %>% # keeps 2701 of 2701
  #intersect(
    # drop T cells based on Cd3d
    #WhichCells(
      #data10xUnfiltered,
      #expression = Cd3d < nonBCellGeneThresholds[["Cd3d"]])) %>% # keeps 2697 of 2701
   #intersect(
    # drop T cells based on Cd3e
    #WhichCells(
      #`Pool_110x.P471-1_10x_analysis.2022-04-13`,
      #expression = Cd3e  nonBCellGeneThresholds[["Cd3e"]])) %>% # keeps 2697 of 2697
  intersect(
     #drop other non-B cells based on Vil1
    WhichCells(
      `Pool_310x.P471-1_10x_analysis.2022-04-13`,
      expression = Vil1 < nonBCellGeneThresholds[["Vil1"]])) %>% # keeps 2695 of 2697
 intersect(
    # drop other non-B cells based on Vil1
      WhichCells(
      `Pool_310x.P471-1_10x_analysis.2022-04-13`,
      expression = Reg3g < nonBCellGeneThresholds[["Reg3g"]])) # keeps 2691 of 2695

cellsDropPool3 <-
  setdiff(colnames(`Pool_310x.P471-1_10x_analysis.2022-04-13`), cellsKeepPool3)

Pool3InclMultiplets <-
  subset(`Pool_310x.P471-1_10x_analysis.2022-04-13`, cells = cellsKeepPool3)

cellsKeepPool4 <-
  WhichCells(
    `Pool_410x.P471-1_10x_analysis.2022-04-13`,
    expression = nFeature_RNA >= qcMetricsThresholds["min_nFeature_RNA"]) %>% # keeps 2841 of 2984
  intersect(
    WhichCells(
      `Pool_410x.P471-1_10x_analysis.2022-04-13`,
      expression = nFeature_RNA <= qcMetricsThresholds["max_nFeature_RNA"])) %>% # keeps 2771 of 2841
  intersect(
    WhichCells(
      `Pool_410x.P471-1_10x_analysis.2022-04-13`,
      expression = nCount_RNA >= qcMetricsThresholds["min_nCount_RNA"])) %>% # keeps 2761 of 2771
  intersect(
    WhichCells(
      `Pool_410x.P471-1_10x_analysis.2022-04-13`,
      expression = nCount_RNA <= qcMetricsThresholds["max_nCount_RNA"])) %>% # keeps 2722 of 2761
  # intersect(
  #   WhichCells(
  #     data10xUnfiltered, expression = nFeature_ADT >= qcMetricsThresholds["nFeature_ADT"])) %>% # keeps ? of ?
  # intersect(
  #   WhichCells(
  #     data10xUnfiltered, expression = nCount_ADT >= qcMetricsThresholds["nCount_ADT"])) %>% # keeps ? of ?
  intersect(
    WhichCells(
      `Pool_410x.P471-1_10x_analysis.2022-04-13`, expression = percent_mito <= qcMetricsThresholds["percent_mito"])) %>% # keeps 2701 of 2722
  intersect(
    WhichCells(
      `Pool_410x.P471-1_10x_analysis.2022-04-13`, expression = percent_ribo <= qcMetricsThresholds["percent_ribo"])) %>% # keeps 2701 of 2701
  intersect(
    WhichCells(
      `Pool_410x.P471-1_10x_analysis.2022-04-13`, expression = percent_hb <= qcMetricsThresholds["percent_hb"])) %>% # keeps 2701 of 2701
  #intersect(
    # drop T cells based on Cd3d
    #WhichCells(
      #data10xUnfiltered,
      #expression = Cd3d < nonBCellGeneThresholds[["Cd3d"]])) %>% # keeps 2697 of 2701
   #intersect(
    # drop T cells based on Cd3e
    #WhichCells(
      #`Pool_110x.P471-1_10x_analysis.2022-04-13`,
      #expression = Cd3e  nonBCellGeneThresholds[["Cd3e"]])) %>% # keeps 2697 of 2697
  intersect(
     #drop other non-B cells based on Vil1
    WhichCells(
      `Pool_410x.P471-1_10x_analysis.2022-04-13`,
      expression = Vil1 < nonBCellGeneThresholds[["Vil1"]])) %>% # keeps 2695 of 2697
 intersect(
    # drop other non-B cells based on Vil1
      WhichCells(
      `Pool_410x.P471-1_10x_analysis.2022-04-13`,
      expression = Reg3g < nonBCellGeneThresholds[["Reg3g"]])) # keeps 2691 of 2695


cellsdrop1 <-   setdiff(colnames(Seurat1_SlICC37_BAL_10xUnfiltered), `1_cellsKeepPool1`)
cellsdrop2 <-   setdiff(colnames(Seurat2_SLICC38_BAL_10xUnfiltered), `2_cellsKeepPool1`)
cellsdrop3 <-   setdiff(colnames(Seurat3_HIPC1_11_11_22_Sput_10xUnfiltered), `3_cellsKeepPool1`)
cellsdrop4 <-   setdiff(colnames(Seurat4_HIPC1_12_13_22_Sput_10xUnfiltered), `4_cellsKeepPool1`)


Seurat1_SLICC37_BAL <-
  subset(Seurat1_SlICC37_BAL_10xUnfiltered, cells = `1_cellsKeepPool1`)

Seurat2_SLICC38_BAL <-
  subset(Seurat2_SLICC38_BAL_10xUnfiltered, cells = `2_cellsKeepPool1`)

Seurat3_HIPC1_11_11_22_Sput <-
  subset(Seurat3_HIPC1_11_11_22_Sput_10xUnfiltered, cells = `3_cellsKeepPool1`)

Seurat4_HIPC1_12_13_22_Sput <-
  subset(Seurat4_HIPC1_12_13_22_Sput_10xUnfiltered, cells = `4_cellsKeepPool1`)
```

## Infer sex of donors based on X-Y chromosome reads

```{r calculateLogXYratio, dependson="filterCellsQcMetrics"}
Pool1InclMultiplets <- Pool1InclMultiplets %>%
  AddMetaData(
    metadata = logXYratio(Pool1InclMultiplets[["RNA"]]@counts),
    col.name = "logXYratio")

Pool2InclMultiplets <- Pool2InclMultiplets %>%
  AddMetaData(
    metadata = logXYratio(Pool1InclMultiplets[["RNA"]]@counts),
    col.name = "logXYratio")

Pool3InclMultiplets <- Pool3InclMultiplets %>%
  AddMetaData(
    metadata = logXYratio(Pool1InclMultiplets[["RNA"]]@counts),
    col.name = "logXYratio")

Pool4InclMultiplets <- Pool4InclMultiplets %>%
  AddMetaData(
    metadata = logXYratio(Pool1InclMultiplets[["RNA"]]@counts),
    col.name = "logXYratio")
```

# Hashtag demultiplexing of samples

## Calculate variations on hashtag counts and store in metadata

```{r calculateHashtagCounts, dependson="calculateLogXYratio"}
# add "clrMd" for each
Pool1InclMultiplets <- Pool1InclMultiplets %>%
  AddMetaData(
    metadata =
      data.frame(
        apply(
          Pool1InclMultiplets[["HTO"]]@counts, MARGIN = 1,
          FUN = function(x) {
            log(x) / apply(Pool1InclMultiplets[["HTO"]]@counts,
                           MARGIN=2, FUN = function(x) geomean(x+1))
          }),
        row.names = colnames(Pool1InclMultiplets)) %>%
      set_colnames(paste0("clrMd", colnames(.)))
  )
for (col.tmp in grep("clrMd", colnames(P487SeuratObjFiltered@meta.data), value = TRUE)) {
  P487SeuratObjFiltered@meta.data[P487SeuratObjFiltered@meta.data[[col.tmp]] == -Inf, col.tmp] <- 0
}

Pool2InclMultiplets <- Pool2InclMultiplets %>%
  AddMetaData(
    metadata =
      data.frame(
        apply(
          Pool2InclMultiplets[["HTO"]]@counts, MARGIN = 1,
          FUN = function(x) {
            log(x) / apply(Pool2InclMultiplets[["HTO"]]@counts,
                           MARGIN=2, FUN = function(x) geomean(x+1))
          }),
        row.names = colnames(Pool2InclMultiplets)) %>%
      set_colnames(paste0("clrMd", colnames(.)))
  )
for (col.tmp in grep("clrMd", colnames(Pool2InclMultiplets@meta.data), value = TRUE)) {
  Pool2InclMultiplets@meta.data[Pool2InclMultiplets@meta.data[[col.tmp]] == -Inf, col.tmp] <- 0
}

Pool3InclMultiplets <- Pool3InclMultiplets %>%
  AddMetaData(
    metadata =
      data.frame(
        apply(
          Pool3InclMultiplets[["HTO"]]@counts, MARGIN = 1,
          FUN = function(x) {
            log(x) / apply(Pool3InclMultiplets[["HTO"]]@counts,
                           MARGIN=2, FUN = function(x) geomean(x+1))
          }),
        row.names = colnames(Pool3InclMultiplets)) %>%
      set_colnames(paste0("clrMd", colnames(.)))
  )
for (col.tmp in grep("clrMd", colnames(Pool3InclMultiplets@meta.data), value = TRUE)) {
  Pool3InclMultiplets@meta.data[Pool3InclMultiplets@meta.data[[col.tmp]] == -Inf, col.tmp] <- 0
}

Pool4InclMultiplets <- Pool4InclMultiplets %>%
  AddMetaData(
    metadata =
      data.frame(
        apply(
          Pool4InclMultiplets[["HTO"]]@counts, MARGIN = 1,
          FUN = function(x) {
            log(x) / apply(Pool4InclMultiplets[["HTO"]]@counts,
                           MARGIN=2, FUN = function(x) geomean(x+1))
          }),
        row.names = colnames(Pool4InclMultiplets)) %>%
      set_colnames(paste0("clrMd", colnames(.)))
  )
for (col.tmp in grep("clrMd", colnames(Pool4InclMultiplets@meta.data), value = TRUE)) {
  Pool4InclMultiplets@meta.data[Pool4InclMultiplets@meta.data[[col.tmp]] == -Inf, col.tmp] <- 0
}

# add log counts for each
offset.tmp <- 1
Pool1InclMultiplets <- Pool1InclMultiplets %>%
  AddMetaData(
    metadata =
      log(Pool1InclMultiplets[["HTO"]]@counts + offset.tmp) %>%
      as.matrix() %>%
      t() %>%
      as.data.frame() %>%
      set_colnames(make.names(paste0("log", colnames(.))))
  )

offset.tmp <- 1
Pool2InclMultiplets <- Pool2InclMultiplets %>%
  AddMetaData(
    metadata =
      log(Pool2InclMultiplets[["HTO"]]@counts + offset.tmp) %>%
      as.matrix() %>%
      t() %>%
      as.data.frame() %>%
      set_colnames(make.names(paste0("log", colnames(.))))
  )

Pool3InclMultiplets <- Pool3InclMultiplets %>%
  AddMetaData(
    metadata =
      log(Pool3InclMultiplets[["HTO"]]@counts + offset.tmp) %>%
      as.matrix() %>%
      t() %>%
      as.data.frame() %>%
      set_colnames(make.names(paste0("log", colnames(.))))
  )

Pool4InclMultiplets <- Pool4InclMultiplets %>%
  AddMetaData(
    metadata =
      log(Pool4InclMultiplets[["HTO"]]@counts + offset.tmp) %>%
      as.matrix() %>%
      t() %>%
      as.data.frame() %>%
      set_colnames(make.names(paste0("log", colnames(.))))
  )

# add logCpm for each
offset.tmp <- 1
Pool1InclMultiplets <- Pool1InclMultiplets %>%
  AddMetaData(
    metadata =
      log(t(as.matrix(Pool1InclMultiplets[["HTO"]]@counts + offset.tmp) ) /
            Pool1InclMultiplets$nCount_HTO * 1e6) %>%
      as.data.frame() %>%
      set_colnames(make.names(paste0("logCpm", colnames(.))))
  )

Pool2InclMultiplets <- Pool2InclMultiplets %>%
  AddMetaData(
    metadata =
      log(t(as.matrix(Pool2InclMultiplets[["HTO"]]@counts + offset.tmp) ) /
            Pool2InclMultiplets$nCount_HTO * 1e6) %>%
      as.data.frame() %>%
      set_colnames(make.names(paste0("logCpm", colnames(.))))
  )

Pool3InclMultiplets <- Pool3InclMultiplets %>%
  AddMetaData(
    metadata =
      log(t(as.matrix(Pool3InclMultiplets[["HTO"]]@counts + offset.tmp) ) /
            Pool3InclMultiplets$nCount_HTO * 1e6) %>%
      as.data.frame() %>%
      set_colnames(make.names(paste0("logCpm", colnames(.))))
  )

Pool4InclMultiplets <- Pool4InclMultiplets %>%
  AddMetaData(
    metadata =
      log(t(as.matrix(Pool4InclMultiplets[["HTO"]]@counts + offset.tmp) ) /
            Pool4InclMultiplets$nCount_HTO * 1e6) %>%
      as.data.frame() %>%
      set_colnames(make.names(paste0("logCpm", colnames(.))))
  )
rm_tmp(ask=FALSE)
```

```{r split HTO assay from ADT assay}
Pool1InclMultiplets[["Hashes"]] <- CreateAssayObject(Pool1InclMultiplets@assays$HTO[c("HTO-M1", "HTO-M2", "HTO-M3")])
Pool1InclMultiplets[["ADT"]] <- CreateAssayObject(Pool1InclMultiplets@assays$HTO[c("ICOS", "GITR", "PD-1", "CD25", "CD44", "CD62L")])
Pool1InclMultiplets[["HTO"]] <- NULL

Pool2InclMultiplets[["Hashes"]] <- CreateAssayObject(Pool2InclMultiplets@assays$HTO[c("HTO-M1", "HTO-M2", "HTO-M3")])
Pool2InclMultiplets[["ADT"]] <- CreateAssayObject(Pool2InclMultiplets@assays$HTO[c("ICOS", "GITR", "PD-1", "CD25", "CD44", "CD62L")])
Pool2InclMultiplets[["HTO"]] <- NULL

Pool3InclMultiplets[["Hashes"]] <- CreateAssayObject(Pool3InclMultiplets@assays$HTO[c("HTO-M1", "HTO-M2", "HTO-M3")])
Pool3InclMultiplets[["ADT"]] <- CreateAssayObject(Pool3InclMultiplets@assays$HTO[c("ICOS", "GITR", "PD-1", "CD25", "CD44", "CD62L")])
Pool3InclMultiplets[["HTO"]] <- NULL

Pool4InclMultiplets[["Hashes"]] <- CreateAssayObject(Pool4InclMultiplets@assays$HTO[c("HTO-M1", "HTO-M2", "HTO-M3")])
Pool4InclMultiplets[["ADT"]] <- CreateAssayObject(Pool4InclMultiplets@assays$HTO[c("ICOS", "GITR", "PD-1", "CD25", "CD44", "CD62L")])
Pool4InclMultiplets[["HTO"]] <- NULL
```
## Hashtag demultiplexing using HTODemux

Our next approach for assigning cells to hashtags is to use the HTODemux algorithm implemented in Seurat. This uses a combination of thresholds and clustering to identify single-positive populations, negative populations, and multiple-positive populations.

```{r hashtagDemultiplexingHTODemux, dependson="filterCellsQcMetrics"}
Pool1InclMultiplets <-
  NormalizeData(Pool1InclMultiplets, assay = "Hashes", normalization.method = "CLR")

Pool2InclMultiplets <-
  NormalizeData(Pool2InclMultiplets, assay = "Hashes", normalization.method = "CLR")

Pool3InclMultiplets <-
  NormalizeData(Pool3InclMultiplets, assay = "Hashes", normalization.method = "CLR")

Pool4InclMultiplets <-
  NormalizeData(Pool4InclMultiplets, assay = "Hashes", normalization.method = "CLR")

# store CLR-transformed hashtag counts in metadata
Pool1InclMultiplets <-
  Pool1InclMultiplets %>%
  AddMetaData(
    metadata =
      data.frame(
        Pool1InclMultiplets@assays$Hashes@data %>%
          t() %>%
          set_colnames(paste0("clrSeurat", colnames(.)))))

Pool2InclMultiplets <-
  Pool2InclMultiplets %>%
  AddMetaData(
    metadata =
      data.frame(
        Pool2InclMultiplets@assays$Hashes@data %>%
          t() %>%
          set_colnames(paste0("clrSeurat", colnames(.)))))

Pool3InclMultiplets <-
  Pool3InclMultiplets %>%
  AddMetaData(
    metadata =
      data.frame(
        Pool3InclMultiplets@assays$Hashes@data %>%
          t() %>%
          set_colnames(paste0("clrSeurat", colnames(.)))))

Pool4InclMultiplets <-
  Pool4InclMultiplets %>%
  AddMetaData(
    metadata =
      data.frame(
        Pool4InclMultiplets@assays$Hashes@data %>%
          t() %>%
          set_colnames(paste0("clrSeurat", colnames(.)))))

Pool1InclMultiplets <-
  HTODemux(
    Pool1InclMultiplets,
    assay = "Hashes",
    init = 2,
    positive.quantile = 0.99)

Pool2InclMultiplets <-
  HTODemux(
    Pool2InclMultiplets,
    assay = "Hashes",
    init = 2,
    positive.quantile = 0.99)

Pool3InclMultiplets <-
  HTODemux(
    Pool3InclMultiplets,
    assay = "Hashes",
    init = 2,
    positive.quantile = 0.99)

Pool4InclMultiplets <-
  HTODemux(
    Pool4InclMultiplets,
    assay = "Hashes",
    init = 2,
    positive.quantile = 0.99)

# no matter how much I tune this, I get way more doublets than I should

# table(data10xIncMultiplets$HTO_classification.global)

# Make a variable that tallies HTO assignment, negative or multiplet
sampleBarcode.tmp <-
  case_when(
    Pool1InclMultiplets$Hashes_classification.global == "Negative" ~ "unassigned",
    Pool1InclMultiplets$Hashes_classification.global == "Doublet" ~ "multiplePositive",
    Pool1InclMultiplets$Hashes_classification.global == "Singlet" ~
      Pool1InclMultiplets$Hashes_maxID) %>%
  factor(levels = str_sort(unique(.), numeric=TRUE))

# add to Seurat object metadata
Pool1InclMultiplets <- Pool1InclMultiplets %>%
  AddMetaData(
    metadata = factor(sampleBarcode.tmp, levels = str_sort(unique(sampleBarcode.tmp), numeric=TRUE)),
    col.name = "sampleBarcodeHTODemux")

rm_tmp(ask=FALSE)

sampleBarcode.tmp <-
  case_when(
    Pool2InclMultiplets$Hashes_classification.global == "Negative" ~ "unassigned",
    Pool2InclMultiplets$Hashes_classification.global == "Doublet" ~ "multiplePositive",
    Pool2InclMultiplets$Hashes_classification.global == "Singlet" ~
      Pool2InclMultiplets$Hashes_maxID) %>%
  factor(levels = str_sort(unique(.), numeric=TRUE))

# add to Seurat object metadata
Pool2InclMultiplets <- Pool2InclMultiplets %>%
  AddMetaData(
    metadata = factor(sampleBarcode.tmp, levels = str_sort(unique(sampleBarcode.tmp), numeric=TRUE)),
    col.name = "sampleBarcodeHTODemux")

rm_tmp(ask=FALSE)

sampleBarcode.tmp <-
  case_when(
    Pool3InclMultiplets$Hashes_classification.global == "Negative" ~ "unassigned",
    Pool3InclMultiplets$Hashes_classification.global == "Doublet" ~ "multiplePositive",
    Pool3InclMultiplets$Hashes_classification.global == "Singlet" ~
      Pool3InclMultiplets$Hashes_maxID) %>%
  factor(levels = str_sort(unique(.), numeric=TRUE))

# add to Seurat object metadata
Pool3InclMultiplets <- Pool3InclMultiplets %>%
  AddMetaData(
    metadata = factor(sampleBarcode.tmp, levels = str_sort(unique(sampleBarcode.tmp), numeric=TRUE)),
    col.name = "sampleBarcodeHTODemux")

rm_tmp(ask=FALSE)

sampleBarcode.tmp <-
  case_when(
    Pool4InclMultiplets$Hashes_classification.global == "Negative" ~ "unassigned",
    Pool4InclMultiplets$Hashes_classification.global == "Doublet" ~ "multiplePositive",
    Pool4InclMultiplets$Hashes_classification.global == "Singlet" ~
      Pool4InclMultiplets$Hashes_maxID) %>%
  factor(levels = str_sort(unique(.), numeric=TRUE))

# add to Seurat object metadata
Pool4InclMultiplets <- Pool4InclMultiplets %>%
  AddMetaData(
    metadata = factor(sampleBarcode.tmp, levels = str_sort(unique(sampleBarcode.tmp), numeric=TRUE)),
    col.name = "sampleBarcodeHTODemux")

rm_tmp(ask=FALSE)

# data10xIncMultiplets@meta.data <-
#   data10xIncMultiplets@meta.data %>%
#   left_join(linkSampleBarcodeToSampleName)
```

```{r countCellsBySampleBarcodeHTODemux, dependson="hashtagDemultiplexingHTODemux"}
Pool1InclMultiplets@meta.data %>%
  dplyr::group_by(sampleBarcodeHTODemux) %>%
  dplyr::summarise(nCells = n()) %>%
  kable(row.names = F,
        caption = "Hashtag demultiplexing by HTODemux") %>%
  kable_styling(full_width = F, 
                position = "left")

Pool2InclMultiplets@meta.data %>%
  dplyr::group_by(sampleBarcodeHTODemux) %>%
  dplyr::summarise(nCells = n()) %>%
  kable(row.names = F,
        caption = "Hashtag demultiplexing by HTODemux") %>%
  kable_styling(full_width = F, 
                position = "left")

Pool3InclMultiplets@meta.data %>%
  dplyr::group_by(sampleBarcodeHTODemux) %>%
  dplyr::summarise(nCells = n()) %>%
  kable(row.names = F,
        caption = "Hashtag demultiplexing by HTODemux") %>%
  kable_styling(full_width = F, 
                position = "left")

Pool4InclMultiplets@meta.data %>%
  dplyr::group_by(sampleBarcodeHTODemux) %>%
  dplyr::summarise(nCells = n()) %>%
  kable(row.names = F,
        caption = "Hashtag demultiplexing by HTODemux") %>%
  kable_styling(full_width = F, 
                position = "left")
```

```{r plotBiaxialAndRidgeHtoLogCountBySampleBarcodeHtoDemux, dependson="hashtagDemultiplexingHTODemux"}
plot.tmp <-
  ggpairs(
    Pool1InclMultiplets@meta.data,
    columns = make.names(paste0("log", sampleBarcodes)),
    upper = list(continuous = "blank"),
    aes(color = sampleBarcodeHTODemux,
        alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.5)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCountBySampleBarcodeHTODemuxPool1.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggpairs(
    Pool2InclMultiplets@meta.data,
    columns = make.names(paste0("log", sampleBarcodes)),
    upper = list(continuous = "blank"),
    aes(color = sampleBarcodeHTODemux,
        alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.5)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCountBySampleBarcodeHTODemuxPool2.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggpairs(
    Pool3InclMultiplets@meta.data,
    columns = make.names(paste0("log", sampleBarcodes)),
    upper = list(continuous = "blank"),
    aes(color = sampleBarcodeHTODemux,
        alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.5)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCountBySampleBarcodeHTODemuxPool3.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggpairs(
    Pool4InclMultiplets@meta.data,
    columns = make.names(paste0("log", sampleBarcodes)),
    upper = list(continuous = "blank"),
    aes(color = sampleBarcodeHTODemux,
        alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.5)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCountBySampleBarcodeHTODemuxPool4.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  RidgePlot(
    Pool1InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("log", sampleBarcodes)), 
    group.by = "sampleBarcodeHTODemux",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCountBySampleBarcodeHTODemuxPool1.pdf")),
    height = 8, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  RidgePlot(
    Pool2InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("log", sampleBarcodes)), 
    group.by = "sampleBarcodeHTODemux",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCountBySampleBarcodeHTODemuxPool2.pdf")),
    height = 8, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  RidgePlot(
    Pool3InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("log", sampleBarcodes)), 
    group.by = "sampleBarcodeHTODemux",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCountBySampleBarcodeHTODemuxPool3.pdf")),
    height = 8, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  RidgePlot(
    Pool4InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("log", sampleBarcodes)), 
    group.by = "sampleBarcodeHTODemux",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCountBySampleBarcodeHTODemuxPool4.pdf")),
    height = 8, width = 20)
print(plot.tmp)
invisible(dev.off())

rm_tmp(ask=FALSE)
```

```{r plotBiaxialAndRidgeHtoLogCpmBySampleBarcodeHtoDemux, dependson="hashtagDemultiplexingHTODemux"}
plot.tmp <-
  ggpairs(
    Pool1InclMultiplets@meta.data,
    columns = make.names(paste0("logCpm", sampleBarcodes)),
    upper = list(continuous = "blank"),
    aes(color = sampleBarcodeHTODemux,
        alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.5)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCpmBySampleBarcodeHTODemuxPool1.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  RidgePlot(
    Pool1InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("logCpm", sampleBarcodes)), 
    group.by = "sampleBarcodeHTODemux",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCpmBySampleBarcodeHTODemuxPool1.pdf")),
    height = 8, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggpairs(
    Pool2InclMultiplets@meta.data,
    columns = make.names(paste0("logCpm", sampleBarcodes)),
    upper = list(continuous = "blank"),
    aes(color = sampleBarcodeHTODemux,
        alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.5)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCpmBySampleBarcodeHTODemuxPool2.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  RidgePlot(
    Pool2InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("logCpm", sampleBarcodes)), 
    group.by = "sampleBarcodeHTODemux",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCpmBySampleBarcodeHTODemuxPool2.pdf")),
    height = 8, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggpairs(
    Pool3InclMultiplets@meta.data,
    columns = make.names(paste0("logCpm", sampleBarcodes)),
    upper = list(continuous = "blank"),
    aes(color = sampleBarcodeHTODemux,
        alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.5)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCpmBySampleBarcodeHTODemuxPool3.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  RidgePlot(
    Pool3InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("logCpm", sampleBarcodes)), 
    group.by = "sampleBarcodeHTODemux",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCpmBySampleBarcodeHTODemuxPool3.pdf")),
    height = 8, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggpairs(
    Pool4InclMultiplets@meta.data,
    columns = make.names(paste0("logCpm", sampleBarcodes)),
    upper = list(continuous = "blank"),
    aes(color = sampleBarcodeHTODemux,
        alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.5)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCpmBySampleBarcodeHTODemuxPool4.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  RidgePlot(
    Pool4InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("logCpm", sampleBarcodes)), 
    group.by = "sampleBarcodeHTODemux",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCpmBySampleBarcodeHTODemuxPool4.pdf")),
    height = 8, width = 20)
print(plot.tmp)
invisible(dev.off())
rm_tmp(ask=FALSE)
```

```{r save files before dmultiplexing}
saveRDS(Pool1InclMultiplets, file.path(dirRoot, 
            paste0("Pool1InclMultiplets.RDS")))

saveRDS(Pool2InclMultiplets, file.path(dirRoot, 
            paste0("Pool2InclMultiplets.RDS")))

saveRDS(Pool3InclMultiplets, file.path(dirRoot, 
            paste0("Pool3InclMultiplets.RDS")))

saveRDS(Pool4InclMultiplets, file.path(dirRoot, 
            paste0("Pool4InclMultiplets.RDS")))

```

```{r reload}

Pool1InclMultiplets <- readRDS(file.path(dirRoot, paste0("Pool1InclMultiplets.RDS")))
Pool2InclMultiplets <- readRDS(file.path(dirRoot, paste0("Pool2InclMultiplets.RDS")))
Pool3InclMultiplets <- readRDS(file.path(dirRoot, paste0("Pool3InclMultiplets.RDS")))
Pool4InclMultiplets <- readRDS(file.path(dirRoot, paste0("Pool4InclMultiplets.RDS")))
```
## Hashtag demultiplexing using distributions of hashtag counts

Our next approach for assigning cells to hashtags is to use a simple threshold for each hashtag. Based on the plots above, we can use the same threshold for each hashtag, and the logCpm appears to be the most effect.

We identify each mouse based on the hashtag used to label cells from that mouse. So mouse1 = HTO-M1, mouse2 = HTO-M2, and so on.

```{r setThresholdsHto}
thresholdsHtoPool1 <-
  list(logCounts = c("posAbove" = 5.5, "negBelow" = 4.5),
       logCpm = c("posAbove" = 11.5, "negBelow" = 11, "minGap" = 1.5))

thresholdsHtoPool2 <-
  list(logCounts = c("posAbove" = 5.5, "negBelow" = 4.5),
       logCpm = c("posAbove" = 11.5, "negBelow" = 10.5, "minGap" = 1.5))

thresholdsHtoPool3 <-
  list(logCounts = c("posAbove" = 5.5, "negBelow" = 4.5),
       logCpm = c("posAbove" = 11.75, "negBelow" = 10.75, "minGap" = 1.5))

thresholdsHtoPool4 <-
  list(logCounts = c("posAbove" = 5.5, "negBelow" = 4.5),
       logCpm = c("posAbove" = 11.5, "negBelow" = 11, "minGap" = 1.5))

```

### Plot barcode distributions

To determine the appropriate thresholds for determining which cells are positive and negative for a given hashtag, we use the plots below.

#### Plot barcode distributions, log counts

```{r plotDensityLogCountsHashtag, dependson=c("calculateHashtagCounts", "setThresholdsHTO", "setupPalettes"), fig.width=10, fig.height=6}
Pool1InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("logHTO"),
    names_to = "hashtag", names_prefix = "log",
    values_to = "count") %>%
  mutate(hashtag = factor(hashtag, levels = str_sort(unique(hashtag), numeric=TRUE))) %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 2) +
  labs(x = "Hashtag counts (log)") +
  geom_vline(
    xintercept = thresholdsHtoPool1[["logCounts"]]["negBelow"],
    linetype = "dashed", color = "red") +
  geom_vline(
    xintercept = thresholdsHtoPool1[["logCounts"]]["posAbove"],
    linetype = "dashed", color = "blue")

Pool2InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("logHTO"),
    names_to = "hashtag", names_prefix = "log",
    values_to = "count") %>%
  mutate(hashtag = factor(hashtag, levels = str_sort(unique(hashtag), numeric=TRUE))) %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 2) +
  labs(x = "Hashtag counts (log)") +
  geom_vline(
    xintercept = thresholdsHtoPool2[["logCounts"]]["negBelow"],
    linetype = "dashed", color = "red") +
  geom_vline(
    xintercept = thresholdsHtoPool2[["logCounts"]]["posAbove"],
    linetype = "dashed", color = "blue")

Pool3InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("logHTO"),
    names_to = "hashtag", names_prefix = "log",
    values_to = "count") %>%
  mutate(hashtag = factor(hashtag, levels = str_sort(unique(hashtag), numeric=TRUE))) %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 2) +
  labs(x = "Hashtag counts (log)") +
  geom_vline(
    xintercept = thresholdsHtoPool3[["logCounts"]]["negBelow"],
    linetype = "dashed", color = "red") +
  geom_vline(
    xintercept = thresholdsHtoPool3[["logCounts"]]["posAbove"],
    linetype = "dashed", color = "blue")

Pool4InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("logHTO"),
    names_to = "hashtag", names_prefix = "log",
    values_to = "count") %>%
  mutate(hashtag = factor(hashtag, levels = str_sort(unique(hashtag), numeric=TRUE))) %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 2) +
  labs(x = "Hashtag counts (log)") +
  geom_vline(
    xintercept = thresholdsHtoPool4[["logCounts"]]["negBelow"],
    linetype = "dashed", color = "red") +
  geom_vline(
    xintercept = thresholdsHtoPool4[["logCounts"]]["posAbove"],
    linetype = "dashed", color = "blue")
```

#### Plot barcode distributions, log CPM

```{r plotBarcodelogCpmDistributionAllSamples, dependson=c("calculateHashtagCounts", "setupPalettes"), fig.width=10, fig.height=6}
Pool1InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("logCpmHTO"),
    names_to = "hashtag", names_prefix = "clrMd",
    values_to = "count") %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 2) +
  labs(x = "Hashtag counts (log CPM)") +
  geom_vline(
    xintercept = thresholdsHtoPool1[["logCpm"]]["negBelow"],
    linetype = "dashed", color = "red") +
  geom_vline(
    xintercept = thresholdsHtoPool1[["logCpm"]]["posAbove"],
    linetype = "dashed", color = "blue")

Pool2InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("logCpmHTO"),
    names_to = "hashtag", names_prefix = "clrMd",
    values_to = "count") %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 2) +
  labs(x = "Hashtag counts (log CPM)") +
  geom_vline(
    xintercept = thresholdsHtoPool2[["logCpm"]]["negBelow"],
    linetype = "dashed", color = "red") +
  geom_vline(
    xintercept = thresholdsHtoPool2[["logCpm"]]["posAbove"],
    linetype = "dashed", color = "blue")

Pool3InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("logCpmHTO"),
    names_to = "hashtag", names_prefix = "clrMd",
    values_to = "count") %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 2) +
  labs(x = "Hashtag counts (log CPM)") +
  geom_vline(
    xintercept = thresholdsHtoPool3[["logCpm"]]["negBelow"],
    linetype = "dashed", color = "red") +
  geom_vline(
    xintercept = thresholdsHtoPool3[["logCpm"]]["posAbove"],
    linetype = "dashed", color = "blue")

Pool4InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("logCpmHTO"),
    names_to = "hashtag", names_prefix = "clrMd",
    values_to = "count") %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 2) +
  labs(x = "Hashtag counts (log CPM)") +
  geom_vline(
    xintercept = thresholdsHtoPool4[["logCpm"]]["negBelow"],
    linetype = "dashed", color = "red") +
  geom_vline(
    xintercept = thresholdsHtoPool4[["logCpm"]]["posAbove"],
    linetype = "dashed", color = "blue")
```

#### Plot barcode distributions, CLR calculated by MD

```{r plotDensityClrMdHashtag, dependson=c("calculateHashtagCounts", "setupPalettes"), fig.width=10, fig.height=6}
Pool1InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("clrMdHTO"),
    names_to = "hashtag", names_prefix = "clrMd",
    values_to = "count") %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 2) +
  labs(x = "Hashtag counts (CLR)")

Pool2InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("clrMdHTO"),
    names_to = "hashtag", names_prefix = "clrMd",
    values_to = "count") %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 3) +
  labs(x = "Hashtag counts (CLR)")

Pool3InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("clrMdHTO"),
    names_to = "hashtag", names_prefix = "clrMd",
    values_to = "count") %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 3) +
  labs(x = "Hashtag counts (CLR)")

Pool4InclMultiplets@meta.data %>%
  pivot_longer(
    cols = matches("clrMdHTO"),
    names_to = "hashtag", names_prefix = "clrMd",
    values_to = "count") %>%
  ggplot(mapping = aes(x = count)) +
  geom_density(size=1) +
  facet_wrap(vars(hashtag), nrow = 3) +
  labs(x = "Hashtag counts (CLR)")
```


### Assign cells to hashtag using barcode count thresholds

```{r hastagDemultiplexingThresholdLogCpm for Pool1, dependson=c("calculateHashtagCounts", "setThresholdsHTO")}
logCpmHTO.tmp <-
  Pool1InclMultiplets@meta.data %>%
  dplyr::select(matches("logCpmHTO"))

thresholdLogCpmHTOStatus.tmp <-
  matrix(
    "ambiguous",
    nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
    dimnames = dimnames(logCpmHTO.tmp))

thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHtoPool1[["logCpm"]]["posAbove"]] <- "positive"
thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp < thresholdsHtoPool1[["logCpm"]]["negBelow"]] <- "negative"

# ## assign cells based on posAbove and negBelow
# # positive if > posAbove
# # negative if < negBelow
# # ambiguous if in between
# # multiplePositive if more than one is not negative (positive or ambiguous)
# # single positive if one is positive and all others are negative
# # unassigned if neither of the cases above applies (either all are negative/ambiguous, or one is positive and one or more is ambiguous)
# thresholdLogCpmHTOStatus.tmp <-
#   matrix(
#     "ambiguous",
#     nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
#     dimnames = dimnames(logCpmHTO.tmp))
# thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHto[["logCpm"]]["posAbove"]] <- "positive"
# thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp < thresholdsHto[["logCpm"]]["negBelow"]] <- "negative"
# 
# sampleBarcode.tmp <- rep("unassigned", ncol(data10xIncMultiplets))
# 
# # identify multiplePositive
# cells.tmp <-
#   which(apply(thresholdLogCpmHTOStatus.tmp, 
#         MARGIN = 1, FUN = function(x) {sum(x %in% c("ambiguous", "positive")) > 1}))
# sampleBarcode.tmp[cells.tmp] <- "multiplePositive"
# 
# # identify  single positives
# cells.tmp <-
#   which(apply(thresholdLogCpmHTOStatus.tmp,
#         MARGIN = 1, 
#         FUN = function(x) {(sum(x %in% "positive") == 1) & (sum(x %in% "ambiguous") == 0)}))
# sampleBarcode.tmp[cells.tmp] <-
#   colnames(thresholdLogCpmHTOStatus.tmp)[
#     apply(thresholdLogCpmHTOStatus.tmp[cells.tmp,],
#           MARGIN = 1, function(x) which(x %in% "positive"))] %>%
#   str_replace_all("logCpm", "")


## assign cells based on posAbove and negBelow
# positive if x > posAbove
# negative if (max(x) - x) > minGap
# ambiguous if neither
# multiplePositive if more than one is not negative (positive or ambiguous)
# single positive if one is positive and all others are negative
# unassigned if neither of the cases above applies (either all are negative/ambiguous, or one is positive and one or more is ambiguous)
thresholdLogCpmHTOStatus.tmp <-
  matrix(
    "ambiguous",
    nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
    dimnames = dimnames(logCpmHTO.tmp))
thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHtoPool1[["logCpm"]]["posAbove"]] <- "positive"
thresholdLogCpmHTOStatus.tmp[
  t(apply(logCpmHTO.tmp, MARGIN = 1, function(x) {max(x) - x})) >
    thresholdsHtoPool1[["logCpm"]]["minGap"]] <-
    "negative"

sampleBarcode.tmp <- rep("unassigned", ncol(Pool1InclMultiplets))

# identify multiplePositive
cells.tmp <-
  which(apply(thresholdLogCpmHTOStatus.tmp, 
        MARGIN = 1, FUN = function(x) {sum(x %in% c("ambiguous", "positive")) > 1}))
sampleBarcode.tmp[cells.tmp] <- "multiplePositive"

# identify  single positives
cells.tmp <-
  which(apply(thresholdLogCpmHTOStatus.tmp,
        MARGIN = 1, 
        FUN = function(x) {(sum(x %in% "positive") == 1) & (sum(x %in% "ambiguous") == 0)}))
sampleBarcode.tmp[cells.tmp] <-
  colnames(thresholdLogCpmHTOStatus.tmp)[
    apply(thresholdLogCpmHTOStatus.tmp[cells.tmp,],
          MARGIN = 1, function(x) which(x %in% "positive"))] %>%
  str_replace_all("logCpmHTO.", "HTO-")

sampleBarcode.tmp <-
  sampleBarcode.tmp %>%
  factor(levels = str_sort(unique(.), numeric=TRUE))

# add to Seurat object metadata
Pool1InclMultiplets <- Pool1InclMultiplets %>%
  AddMetaData(
    metadata = sampleBarcode.tmp,
    col.name = "sampleBarcodeThresholdLogCpm")

# diagnostic stuff
# for (i in levels(data10xIncMultiplets@meta.data$sampleBarcodeThresholdLogCpm)) {
#   cat("\n\n", i, "\n")
#   data10xIncMultiplets@meta.data %>% 
#     dplyr::filter(sampleBarcodeThresholdLogCpm == i) %>%
#     dplyr::select(matches("logCpmHTO")) %>%
#     print()
# }
# 
# data10xIncMultiplets@meta.data %>% 
#   dplyr::filter(sampleBarcodeThresholdLogCpm == "multiplePositive") %>%
#   dplyr::select(matches("logCpmHTO")) %>%
#   apply(MARGIN = 1, sort)

rm_tmp(ask=FALSE)
```

```{r hastagDemultiplexingThresholdLogCpm for Pool2, dependson=c("calculateHashtagCounts", "setThresholdsHTO")}
logCpmHTO.tmp <-
  Pool2InclMultiplets@meta.data %>%
  dplyr::select(matches("logCpmHTO"))

thresholdLogCpmHTOStatus.tmp <-
  matrix(
    "ambiguous",
    nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
    dimnames = dimnames(logCpmHTO.tmp))

thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHtoPool2[["logCpm"]]["posAbove"]] <- "positive"
thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp < thresholdsHtoPool2[["logCpm"]]["negBelow"]] <- "negative"

# ## assign cells based on posAbove and negBelow
# # positive if > posAbove
# # negative if < negBelow
# # ambiguous if in between
# # multiplePositive if more than one is not negative (positive or ambiguous)
# # single positive if one is positive and all others are negative
# # unassigned if neither of the cases above applies (either all are negative/ambiguous, or one is positive and one or more is ambiguous)
# thresholdLogCpmHTOStatus.tmp <-
#   matrix(
#     "ambiguous",
#     nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
#     dimnames = dimnames(logCpmHTO.tmp))
# thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHto[["logCpm"]]["posAbove"]] <- "positive"
# thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp < thresholdsHto[["logCpm"]]["negBelow"]] <- "negative"
# 
# sampleBarcode.tmp <- rep("unassigned", ncol(data10xIncMultiplets))
# 
# # identify multiplePositive
# cells.tmp <-
#   which(apply(thresholdLogCpmHTOStatus.tmp, 
#         MARGIN = 1, FUN = function(x) {sum(x %in% c("ambiguous", "positive")) > 1}))
# sampleBarcode.tmp[cells.tmp] <- "multiplePositive"
# 
# # identify  single positives
# cells.tmp <-
#   which(apply(thresholdLogCpmHTOStatus.tmp,
#         MARGIN = 1, 
#         FUN = function(x) {(sum(x %in% "positive") == 1) & (sum(x %in% "ambiguous") == 0)}))
# sampleBarcode.tmp[cells.tmp] <-
#   colnames(thresholdLogCpmHTOStatus.tmp)[
#     apply(thresholdLogCpmHTOStatus.tmp[cells.tmp,],
#           MARGIN = 1, function(x) which(x %in% "positive"))] %>%
#   str_replace_all("logCpm", "")


## assign cells based on posAbove and negBelow
# positive if x > posAbove
# negative if (max(x) - x) > minGap
# ambiguous if neither
# multiplePositive if more than one is not negative (positive or ambiguous)
# single positive if one is positive and all others are negative
# unassigned if neither of the cases above applies (either all are negative/ambiguous, or one is positive and one or more is ambiguous)
thresholdLogCpmHTOStatus.tmp <-
  matrix(
    "ambiguous",
    nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
    dimnames = dimnames(logCpmHTO.tmp))
thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHtoPool2[["logCpm"]]["posAbove"]] <- "positive"
thresholdLogCpmHTOStatus.tmp[
  t(apply(logCpmHTO.tmp, MARGIN = 1, function(x) {max(x) - x})) >
    thresholdsHtoPool2[["logCpm"]]["minGap"]] <-
    "negative"

sampleBarcode.tmp <- rep("unassigned", ncol(Pool2InclMultiplets))

# identify multiplePositive
cells.tmp <-
  which(apply(thresholdLogCpmHTOStatus.tmp, 
        MARGIN = 1, FUN = function(x) {sum(x %in% c("ambiguous", "positive")) > 1}))
sampleBarcode.tmp[cells.tmp] <- "multiplePositive"

# identify  single positives
cells.tmp <-
  which(apply(thresholdLogCpmHTOStatus.tmp,
        MARGIN = 1, 
        FUN = function(x) {(sum(x %in% "positive") == 1) & (sum(x %in% "ambiguous") == 0)}))
sampleBarcode.tmp[cells.tmp] <-
  colnames(thresholdLogCpmHTOStatus.tmp)[
    apply(thresholdLogCpmHTOStatus.tmp[cells.tmp,],
          MARGIN = 1, function(x) which(x %in% "positive"))] %>%
  str_replace_all("logCpmHTO.", "HTO-")

sampleBarcode.tmp <-
  sampleBarcode.tmp %>%
  factor(levels = str_sort(unique(.), numeric=TRUE))

# add to Seurat object metadata
Pool2InclMultiplets <- Pool2InclMultiplets %>%
  AddMetaData(
    metadata = sampleBarcode.tmp,
    col.name = "sampleBarcodeThresholdLogCpm")

# diagnostic stuff
# for (i in levels(data10xIncMultiplets@meta.data$sampleBarcodeThresholdLogCpm)) {
#   cat("\n\n", i, "\n")
#   data10xIncMultiplets@meta.data %>% 
#     dplyr::filter(sampleBarcodeThresholdLogCpm == i) %>%
#     dplyr::select(matches("logCpmHTO")) %>%
#     print()
# }
# 
# data10xIncMultiplets@meta.data %>% 
#   dplyr::filter(sampleBarcodeThresholdLogCpm == "multiplePositive") %>%
#   dplyr::select(matches("logCpmHTO")) %>%
#   apply(MARGIN = 1, sort)

rm_tmp(ask=FALSE)
```

```{r hastagDemultiplexingThresholdLogCpm for Pool3, dependson=c("calculateHashtagCounts", "setThresholdsHTO")}
logCpmHTO.tmp <-
  Pool3InclMultiplets@meta.data %>%
  dplyr::select(matches("logCpmHTO"))

thresholdLogCpmHTOStatus.tmp <-
  matrix(
    "ambiguous",
    nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
    dimnames = dimnames(logCpmHTO.tmp))

thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHtoPool3[["logCpm"]]["posAbove"]] <- "positive"
thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp < thresholdsHtoPool3[["logCpm"]]["negBelow"]] <- "negative"

# ## assign cells based on posAbove and negBelow
# # positive if > posAbove
# # negative if < negBelow
# # ambiguous if in between
# # multiplePositive if more than one is not negative (positive or ambiguous)
# # single positive if one is positive and all others are negative
# # unassigned if neither of the cases above applies (either all are negative/ambiguous, or one is positive and one or more is ambiguous)
# thresholdLogCpmHTOStatus.tmp <-
#   matrix(
#     "ambiguous",
#     nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
#     dimnames = dimnames(logCpmHTO.tmp))
# thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHto[["logCpm"]]["posAbove"]] <- "positive"
# thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp < thresholdsHto[["logCpm"]]["negBelow"]] <- "negative"
# 
# sampleBarcode.tmp <- rep("unassigned", ncol(data10xIncMultiplets))
# 
# # identify multiplePositive
# cells.tmp <-
#   which(apply(thresholdLogCpmHTOStatus.tmp, 
#         MARGIN = 1, FUN = function(x) {sum(x %in% c("ambiguous", "positive")) > 1}))
# sampleBarcode.tmp[cells.tmp] <- "multiplePositive"
# 
# # identify  single positives
# cells.tmp <-
#   which(apply(thresholdLogCpmHTOStatus.tmp,
#         MARGIN = 1, 
#         FUN = function(x) {(sum(x %in% "positive") == 1) & (sum(x %in% "ambiguous") == 0)}))
# sampleBarcode.tmp[cells.tmp] <-
#   colnames(thresholdLogCpmHTOStatus.tmp)[
#     apply(thresholdLogCpmHTOStatus.tmp[cells.tmp,],
#           MARGIN = 1, function(x) which(x %in% "positive"))] %>%
#   str_replace_all("logCpm", "")


## assign cells based on posAbove and negBelow
# positive if x > posAbove
# negative if (max(x) - x) > minGap
# ambiguous if neither
# multiplePositive if more than one is not negative (positive or ambiguous)
# single positive if one is positive and all others are negative
# unassigned if neither of the cases above applies (either all are negative/ambiguous, or one is positive and one or more is ambiguous)
thresholdLogCpmHTOStatus.tmp <-
  matrix(
    "ambiguous",
    nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
    dimnames = dimnames(logCpmHTO.tmp))
thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHtoPool3[["logCpm"]]["posAbove"]] <- "positive"
thresholdLogCpmHTOStatus.tmp[
  t(apply(logCpmHTO.tmp, MARGIN = 1, function(x) {max(x) - x})) >
    thresholdsHtoPool3[["logCpm"]]["minGap"]] <-
    "negative"

sampleBarcode.tmp <- rep("unassigned", ncol(Pool3InclMultiplets))

# identify multiplePositive
cells.tmp <-
  which(apply(thresholdLogCpmHTOStatus.tmp, 
        MARGIN = 1, FUN = function(x) {sum(x %in% c("ambiguous", "positive")) > 1}))
sampleBarcode.tmp[cells.tmp] <- "multiplePositive"

# identify  single positives
cells.tmp <-
  which(apply(thresholdLogCpmHTOStatus.tmp,
        MARGIN = 1, 
        FUN = function(x) {(sum(x %in% "positive") == 1) & (sum(x %in% "ambiguous") == 0)}))
sampleBarcode.tmp[cells.tmp] <-
  colnames(thresholdLogCpmHTOStatus.tmp)[
    apply(thresholdLogCpmHTOStatus.tmp[cells.tmp,],
          MARGIN = 1, function(x) which(x %in% "positive"))] %>%
  str_replace_all("logCpmHTO.", "HTO-")

sampleBarcode.tmp <-
  sampleBarcode.tmp %>%
  factor(levels = str_sort(unique(.), numeric=TRUE))

# add to Seurat object metadata
Pool3InclMultiplets <- Pool3InclMultiplets %>%
  AddMetaData(
    metadata = sampleBarcode.tmp,
    col.name = "sampleBarcodeThresholdLogCpm")

# diagnostic stuff
# for (i in levels(data10xIncMultiplets@meta.data$sampleBarcodeThresholdLogCpm)) {
#   cat("\n\n", i, "\n")
#   data10xIncMultiplets@meta.data %>% 
#     dplyr::filter(sampleBarcodeThresholdLogCpm == i) %>%
#     dplyr::select(matches("logCpmHTO")) %>%
#     print()
# }
# 
# data10xIncMultiplets@meta.data %>% 
#   dplyr::filter(sampleBarcodeThresholdLogCpm == "multiplePositive") %>%
#   dplyr::select(matches("logCpmHTO")) %>%
#   apply(MARGIN = 1, sort)

rm_tmp(ask=FALSE)
```

```{r hastagDemultiplexingThresholdLogCpm for Pool1, dependson=c("calculateHashtagCounts", "setThresholdsHTO")}
logCpmHTO.tmp <-
  Pool4InclMultiplets@meta.data %>%
  dplyr::select(matches("logCpmHTO"))

thresholdLogCpmHTOStatus.tmp <-
  matrix(
    "ambiguous",
    nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
    dimnames = dimnames(logCpmHTO.tmp))

thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHtoPool4[["logCpm"]]["posAbove"]] <- "positive"
thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp < thresholdsHtoPool4[["logCpm"]]["negBelow"]] <- "negative"

# ## assign cells based on posAbove and negBelow
# # positive if > posAbove
# # negative if < negBelow
# # ambiguous if in between
# # multiplePositive if more than one is not negative (positive or ambiguous)
# # single positive if one is positive and all others are negative
# # unassigned if neither of the cases above applies (either all are negative/ambiguous, or one is positive and one or more is ambiguous)
# thresholdLogCpmHTOStatus.tmp <-
#   matrix(
#     "ambiguous",
#     nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
#     dimnames = dimnames(logCpmHTO.tmp))
# thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHto[["logCpm"]]["posAbove"]] <- "positive"
# thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp < thresholdsHto[["logCpm"]]["negBelow"]] <- "negative"
# 
# sampleBarcode.tmp <- rep("unassigned", ncol(data10xIncMultiplets))
# 
# # identify multiplePositive
# cells.tmp <-
#   which(apply(thresholdLogCpmHTOStatus.tmp, 
#         MARGIN = 1, FUN = function(x) {sum(x %in% c("ambiguous", "positive")) > 1}))
# sampleBarcode.tmp[cells.tmp] <- "multiplePositive"
# 
# # identify  single positives
# cells.tmp <-
#   which(apply(thresholdLogCpmHTOStatus.tmp,
#         MARGIN = 1, 
#         FUN = function(x) {(sum(x %in% "positive") == 1) & (sum(x %in% "ambiguous") == 0)}))
# sampleBarcode.tmp[cells.tmp] <-
#   colnames(thresholdLogCpmHTOStatus.tmp)[
#     apply(thresholdLogCpmHTOStatus.tmp[cells.tmp,],
#           MARGIN = 1, function(x) which(x %in% "positive"))] %>%
#   str_replace_all("logCpm", "")


## assign cells based on posAbove and negBelow
# positive if x > posAbove
# negative if (max(x) - x) > minGap
# ambiguous if neither
# multiplePositive if more than one is not negative (positive or ambiguous)
# single positive if one is positive and all others are negative
# unassigned if neither of the cases above applies (either all are negative/ambiguous, or one is positive and one or more is ambiguous)
thresholdLogCpmHTOStatus.tmp <-
  matrix(
    "ambiguous",
    nrow = nrow(logCpmHTO.tmp), ncol = ncol(logCpmHTO.tmp),
    dimnames = dimnames(logCpmHTO.tmp))
thresholdLogCpmHTOStatus.tmp[logCpmHTO.tmp > thresholdsHtoPool4[["logCpm"]]["posAbove"]] <- "positive"
thresholdLogCpmHTOStatus.tmp[
  t(apply(logCpmHTO.tmp, MARGIN = 1, function(x) {max(x) - x})) >
    thresholdsHtoPool4[["logCpm"]]["minGap"]] <-
    "negative"

sampleBarcode.tmp <- rep("unassigned", ncol(Pool4InclMultiplets))

# identify multiplePositive
cells.tmp <-
  which(apply(thresholdLogCpmHTOStatus.tmp, 
        MARGIN = 1, FUN = function(x) {sum(x %in% c("ambiguous", "positive")) > 1}))
sampleBarcode.tmp[cells.tmp] <- "multiplePositive"

# identify  single positives
cells.tmp <-
  which(apply(thresholdLogCpmHTOStatus.tmp,
        MARGIN = 1, 
        FUN = function(x) {(sum(x %in% "positive") == 1) & (sum(x %in% "ambiguous") == 0)}))
sampleBarcode.tmp[cells.tmp] <-
  colnames(thresholdLogCpmHTOStatus.tmp)[
    apply(thresholdLogCpmHTOStatus.tmp[cells.tmp,],
          MARGIN = 1, function(x) which(x %in% "positive"))] %>%
  str_replace_all("logCpmHTO.", "HTO-")

sampleBarcode.tmp <-
  sampleBarcode.tmp %>%
  factor(levels = str_sort(unique(.), numeric=TRUE))

# add to Seurat object metadata
Pool4InclMultiplets <- Pool4InclMultiplets %>%
  AddMetaData(
    metadata = sampleBarcode.tmp,
    col.name = "sampleBarcodeThresholdLogCpm")

# diagnostic stuff
# for (i in levels(data10xIncMultiplets@meta.data$sampleBarcodeThresholdLogCpm)) {
#   cat("\n\n", i, "\n")
#   data10xIncMultiplets@meta.data %>% 
#     dplyr::filter(sampleBarcodeThresholdLogCpm == i) %>%
#     dplyr::select(matches("logCpmHTO")) %>%
#     print()
# }
# 
# data10xIncMultiplets@meta.data %>% 
#   dplyr::filter(sampleBarcodeThresholdLogCpm == "multiplePositive") %>%
#   dplyr::select(matches("logCpmHTO")) %>%
#   apply(MARGIN = 1, sort)

rm_tmp(ask=FALSE)
```

```{r countCellsBySampleBarcodeThresholdLogCpm, dependson="hastagDemultiplexingThresholdLogCpm"}
Pool1InclMultiplets@meta.data %>%
  dplyr::group_by(sampleBarcodeThresholdLogCpm) %>%
  dplyr::summarise(nCells = n()) %>%
  kable(row.names = F,
        caption = "Hashtag demultiplexing by ThresholdLogCpm") %>%
  kable_styling(full_width = F, 
                position = "left")

Pool2InclMultiplets@meta.data %>%
  dplyr::group_by(sampleBarcodeThresholdLogCpm) %>%
  dplyr::summarise(nCells = n()) %>%
  kable(row.names = F,
        caption = "Hashtag demultiplexing by ThresholdLogCpm") %>%
  kable_styling(full_width = F, 
                position = "left")

Pool3InclMultiplets@meta.data %>%
  dplyr::group_by(sampleBarcodeThresholdLogCpm) %>%
  dplyr::summarise(nCells = n()) %>%
  kable(row.names = F,
        caption = "Hashtag demultiplexing by ThresholdLogCpm") %>%
  kable_styling(full_width = F, 
                position = "left")

Pool4InclMultiplets@meta.data %>%
  dplyr::group_by(sampleBarcodeThresholdLogCpm) %>%
  dplyr::summarise(nCells = n()) %>%
  kable(row.names = F,
        caption = "Hashtag demultiplexing by ThresholdLogCpm") %>%
  kable_styling(full_width = F, 
                position = "left")
```



```{r plotBiaxialAndRidgeHtoLogCountBySampleBarcodeThresholdLogCpm, dependson="hashtagDemultiplexingThresholdLogCpm"}
plot.tmp <-
  ggpairs(
  Pool1InclMultiplets@meta.data,
  columns = make.names(paste0("log", sampleBarcodes)),
  upper = list(continuous = "blank"),
  aes(color = sampleBarcodeThresholdLogCpm,
      alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.4)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCountBySampleBarcodeThresholdLogCpmPool1.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  RidgePlot(
    Pool1InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("log", sampleBarcodes)), 
    group.by = "sampleBarcodeThresholdLogCpm",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCountBySampleBarcodeThresholdLogCpmPool1.pdf")),
    height = 8, width = 10)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  ggpairs(
  Pool2InclMultiplets@meta.data,
  columns = make.names(paste0("log", sampleBarcodes)),
  upper = list(continuous = "blank"),
  aes(color = sampleBarcodeThresholdLogCpm,
      alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.4)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCountBySampleBarcodeThresholdLogCpmPool2.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  RidgePlot(
    Pool2InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("log", sampleBarcodes)), 
    group.by = "sampleBarcodeThresholdLogCpm",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCountBySampleBarcodeThresholdLogCpmPool2.pdf")),
    height = 8, width = 10)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggpairs(
  Pool3InclMultiplets@meta.data,
  columns = make.names(paste0("log", sampleBarcodes)),
  upper = list(continuous = "blank"),
  aes(color = sampleBarcodeThresholdLogCpm,
      alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.4)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCountBySampleBarcodeThresholdLogCpmPool3.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  RidgePlot(
    Pool3InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("log", sampleBarcodes)), 
    group.by = "sampleBarcodeThresholdLogCpm",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCountBySampleBarcodeThresholdLogCpmPool3.pdf")),
    height = 8, width = 10)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggpairs(
  Pool4InclMultiplets@meta.data,
  columns = make.names(paste0("log", sampleBarcodes)),
  upper = list(continuous = "blank"),
  aes(color = sampleBarcodeThresholdLogCpm,
      alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.4)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCountBySampleBarcodeThresholdLogCpmPool4.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  RidgePlot(
    Pool4InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("log", sampleBarcodes)), 
    group.by = "sampleBarcodeThresholdLogCpm",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCountBySampleBarcodeThresholdLogCpmPool4.pdf")),
    height = 8, width = 10)
print(plot.tmp)
invisible(dev.off())
rm_tmp(ask=FALSE)
```

This looks excellent!!!

```{r plotBiaxialAndRidgeHtoLogCpmBySampleBarcodeThresholdLogCpm, dependson="hashtagDemultiplexingThresholdLogCpm"}
plot.tmp <-
  ggpairs(
  Pool1InclMultiplets@meta.data,
  columns = make.names(paste0("logCpm", sampleBarcodes)),
  upper = list(continuous = "blank"),
  aes(color = sampleBarcodeThresholdLogCpm,
      alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.4)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCpmBySampleBarcodeThresholdLogCpmPool1.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  RidgePlot(
    Pool1InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("logCpm", sampleBarcodes)), 
    group.by = "sampleBarcodeThresholdLogCpm",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCpmBySampleBarcodeThresholdLogCpmPool1.pdf")),
    height = 8, width = 10)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  ggpairs(
  Pool2InclMultiplets@meta.data,
  columns = make.names(paste0("logCpm", sampleBarcodes)),
  upper = list(continuous = "blank"),
  aes(color = sampleBarcodeThresholdLogCpm,
      alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.4)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCpmBySampleBarcodeThresholdLogCpmPool2.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  RidgePlot(
    Pool2InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("logCpm", sampleBarcodes)), 
    group.by = "sampleBarcodeThresholdLogCpm",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCpmBySampleBarcodeThresholdLogCpmPool2.pdf")),
    height = 8, width = 10)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggpairs(
  Pool3InclMultiplets@meta.data,
  columns = make.names(paste0("logCpm", sampleBarcodes)),
  upper = list(continuous = "blank"),
  aes(color = sampleBarcodeThresholdLogCpm,
      alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.4)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCpmBySampleBarcodeThresholdLogCpmPool3.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  RidgePlot(
    Pool3InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("logCpm", sampleBarcodes)), 
    group.by = "sampleBarcodeThresholdLogCpm",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCpmBySampleBarcodeThresholdLogCpmPool3.pdf")),
    height = 8, width = 10)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggpairs(
  Pool4InclMultiplets@meta.data,
  columns = make.names(paste0("logCpm", sampleBarcodes)),
  upper = list(continuous = "blank"),
  aes(color = sampleBarcodeThresholdLogCpm,
      alpha = 0.4)) +
  scale_color_manual(values = pal.sampleBarcode) +
  scale_fill_manual(values = pal.sampleBarcode) +
  theme(strip.text = element_text(size = rel(0.4)),
        axis.text = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("scatterplotsHtoLogCpmBySampleBarcodeThresholdLogCpmPool4.pdf")),
    height = 20, width = 20)
print(plot.tmp)
invisible(dev.off())


plot.tmp <-
  RidgePlot(
    Pool4InclMultiplets, 
    assay = "Hashes", 
    features =
      make.names(
        paste0("logCpm", sampleBarcodes)), 
    group.by = "sampleBarcodeThresholdLogCpm",
    same.y.lims = TRUE,
    ncol = 5) &
  scale_fill_manual(values = pal.sampleBarcode) &
  labs(x = NULL) &
  theme(axis.text.y = element_text(size = 10))

pdf(
  file.path(
    "plots", 
    paste0("ridgeplotsHtoLogCpmBySampleBarcodeThresholdLogCpmPool4.pdf")),
    height = 8, width = 10)
print(plot.tmp)
invisible(dev.off())
rm_tmp(ask=FALSE)
```

## Use DoubletFinder tool to identify doublets

1) Do this on split data, so that the detection can be sample-specific.
2) Do this before removing doublets identified by double-positive barcodes, as a way to check that it's working. Approximately half of the doublets should be double-positive, as 9/10 of them should have one cell from each.

*Continue here on doublet identification, 2021-12-05*

```{r identifyDoubletsDoubletFinder}
# remotes::install_github("chris-mcginnis-ucsf/DoubletFinder", upgrade = F)
```

```{r saving files}
saveRDS(Pool1InclMultiplets, "Pool1InclMultiplets.RDS")
saveRDS(Pool2InclMultiplets, "Pool2InclMultiplets.RDS")
saveRDS(Pool3InclMultiplets, "Pool3InclMultiplets.RDS")
saveRDS(Pool4InclMultiplets, "Pool4InclMultiplets.RDS")
```
## Assign demultiplexing calls, and remove multiplets identified by double-positive for barcode counts

```{r assignSampleBarcodes, dependson=c("hastagDemultiplexingHTODemux", "hastagDemultiplexingThresholdLogCpm")}
sampleBarcodeUsed <- "sampleBarcodeThresholdLogCpm"
# sampleBarcodeUsed <- "sampleBarcodeHTODemux"

# for the original combined version of the data
Pool1InclMultiplets@meta.data <-
  Pool1InclMultiplets@meta.data %>%
  dplyr::mutate(sampleBarcode = !!rlang::sym(sampleBarcodeUsed)) %>%
  left_join(linkSampleBarcodeToSampleName)
rownames(Pool1InclMultiplets@meta.data) <- Pool1InclMultiplets@meta.data$cellBarcode

Pool2InclMultiplets@meta.data <-
  Pool2InclMultiplets@meta.data %>%
  dplyr::mutate(sampleBarcode = !!rlang::sym(sampleBarcodeUsed)) %>%
  left_join(linkSampleBarcodeToSampleName)
rownames(Pool2InclMultiplets@meta.data) <- Pool2InclMultiplets@meta.data$cellBarcode

Pool3InclMultiplets@meta.data <-
  Pool3InclMultiplets@meta.data %>%
  dplyr::mutate(sampleBarcode = !!rlang::sym(sampleBarcodeUsed)) %>%
  left_join(linkSampleBarcodeToSampleName)
rownames(Pool3InclMultiplets@meta.data) <- Pool3InclMultiplets@meta.data$cellBarcode

Pool4InclMultiplets@meta.data <-
  Pool4InclMultiplets@meta.data %>%
  dplyr::mutate(sampleBarcode = !!rlang::sym(sampleBarcodeUsed)) %>%
  left_join(linkSampleBarcodeToSampleName)
rownames(Pool4InclMultiplets@meta.data) <- Pool4InclMultiplets@meta.data$cellBarcode
```

```{r outputFileCellCategory, dependson="assignSampleBarcodes"}
Pool1CellCategory <-
  `Pool_110x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(cellBarcode) %>%
  left_join(
    Pool1InclMultiplets@meta.data %>% 
      dplyr::select(cellBarcode, sampleBarcode)) %>%
  mutate(
    cellCategory =
      case_when(
        cellBarcode %in% cellsDropPool1 ~ "removedQc",
        cellBarcode %in% cellsKeepPool1 ~ as.character(sampleBarcode))) %>%
  dplyr::select(cellBarcode, cellCategory)

  Pool1CellCategory %>%
    write.csv(
      file = file.path("tables", paste0("P389_sampleBarcode_cellCategoryPool1.csv")),
      row.names = FALSE)

Pool2CellCategory <-
  `Pool_210x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(cellBarcode) %>%
  left_join(
    Pool2InclMultiplets@meta.data %>% 
      dplyr::select(cellBarcode, sampleBarcode)) %>%
  mutate(
    cellCategory =
      case_when(
        cellBarcode %in% cellsDropPool2 ~ "removedQc",
        cellBarcode %in% cellsKeepPool2 ~ as.character(sampleBarcode))) %>%
  dplyr::select(cellBarcode, cellCategory)

Pool2CellCategory %>%
  write.csv(
    file = file.path("tables", paste0("P389_sampleBarcode_cellCategoryPool2.csv")),
    row.names = FALSE)

Pool3CellCategory <-
  `Pool_310x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(cellBarcode) %>%
  left_join(
    Pool3InclMultiplets@meta.data %>% 
      dplyr::select(cellBarcode, sampleBarcode)) %>%
  mutate(
    cellCategory =
      case_when(
        cellBarcode %in% cellsDropPool3 ~ "removedQc",
        cellBarcode %in% cellsKeepPool3 ~ as.character(sampleBarcode))) %>%
  dplyr::select(cellBarcode, cellCategory)

Pool3CellCategory %>%
  write.csv(
    file = file.path("tables", paste0("P389_sampleBarcode_cellCategoryPool3.csv")),
    row.names = FALSE)

Pool4CellCategory <-
  `Pool_410x.P471-1_10x_analysis.2022-04-13`@meta.data %>%
  dplyr::select(cellBarcode) %>%
  left_join(
    Pool4InclMultiplets@meta.data %>% 
      dplyr::select(cellBarcode, sampleBarcode)) %>%
  mutate(
    cellCategory =
      case_when(
        cellBarcode %in% cellsDropPool4 ~ "removedQc",
        cellBarcode %in% cellsKeepPool4 ~ as.character(sampleBarcode))) %>%
  dplyr::select(cellBarcode, cellCategory)

Pool4CellCategory %>%
  write.csv(
    file = file.path("tables", paste0("P389_sampleBarcode_cellCategoryPool4.csv")),
    row.names = FALSE)
```

```{r saveData10xBeforeRemoveDoublets, dependson="assignSampleBarcodes"}
# save version of data10xIncMultiplets for easy recovery
filenamePool1CompleteIncDoublets <-
  file.path("data_saved", 
            paste0("Pool1CompleteIncDoublets.RDS"))
saveRDS(Pool1InclMultiplets, filenamePool1CompleteIncDoublets)
# data10xIncMultiplets <- readRDS(filenameData10xCompleteIncDoublets)

filenamePool2CompleteIncDoublets <-
  file.path("data_saved", 
            paste0("Pool2CompleteIncDoublets.RDS"))
saveRDS(Pool2InclMultiplets, filenamePool2CompleteIncDoublets)

filenamePool3CompleteIncDoublets <-
  file.path("data_saved", 
            paste0("Pool3CompleteIncDoublets.RDS"))
saveRDS(Pool3InclMultiplets, filenamePool3CompleteIncDoublets)

filenamePool4CompleteIncDoublets <-
  file.path("data_saved", 
            paste0("Pool4CompleteIncDoublets.RDS"))
saveRDS(Pool4InclMultiplets, filenamePool4CompleteIncDoublets)

```

```{r removeCellBarcodeMultiplePositive, dependson="assignSampleBarcodes"}
# for the original combined version of the data
rownames(Pool1InclMultiplets@meta.data) <- Pool1InclMultiplets@meta.data$cellBarcode
cells.tmp <-
  WhichCells(Pool1InclMultiplets, expression = sampleBarcode %in% paste0("HTO-M", 1:3))
Pool1 <-
  subset(Pool1InclMultiplets, cells = cells.tmp)
Pool1@meta.data$sampleName <-
  droplevels(Pool1@meta.data$sampleName)

rm_tmp(ask=FALSE)

rownames(Pool2InclMultiplets@meta.data) <- Pool2InclMultiplets@meta.data$cellBarcode
cells.tmp <-
  WhichCells(Pool2InclMultiplets, expression = sampleBarcode %in% paste0("HTO-M", 1:3))
Pool2 <-
  subset(Pool2InclMultiplets, cells = cells.tmp)
Pool2@meta.data$sampleName <-
  droplevels(Pool2@meta.data$sampleName)

rm_tmp(ask=FALSE)

rownames(Pool3InclMultiplets@meta.data) <- Pool3InclMultiplets@meta.data$cellBarcode
cells.tmp <-
  WhichCells(Pool3InclMultiplets, expression = sampleBarcode %in% paste0("HTO-M", 1:3))
Pool3 <-
  subset(Pool3InclMultiplets, cells = cells.tmp)
Pool3@meta.data$sampleName <-
  droplevels(Pool3@meta.data$sampleName)

rm_tmp(ask=FALSE)

rownames(Pool4InclMultiplets@meta.data) <- Pool4InclMultiplets@meta.data$cellBarcode
cells.tmp <-
  WhichCells(Pool4InclMultiplets, expression = sampleBarcode %in% paste0("HTO-M", 1:3))
Pool4 <-
  subset(Pool4InclMultiplets, cells = cells.tmp)
Pool4@meta.data$sampleName <-
  droplevels(Pool4@meta.data$sampleName)
rm_tmp(ask=FALSE)
```

## Output number of VDJ chains for singlets and multiplets

```{r outputTableNumberVdjChainsSingletsMultiplets, dependson=c("assignSampleBarcodes", "loadData10xVdj")}
Pool1VdjUnfiltered %>%
  inner_join(Pool1InclMultiplets@meta.data) %>%
  group_by(cellBarcode, sampleBarcode) %>%
  summarise(n_TRA = sum(chain == "TRA"),
            n_TRB = sum(chain == "TRB")) %>%
  group_by(sampleBarcode, n_TRA, n_TRB) %>%
  summarise(n_cells = n()) %>%
  dplyr::arrange(sampleBarcode, n_TRA, n_TRB) %>%
  kable(row.names = F,
        caption = "Cells with single and multiple chains, by mouse") %>%
  kable_styling(full_width = F, 
                position = "left")

Pool2VdjUnfiltered %>%
  inner_join(Pool2InclMultiplets@meta.data) %>%
  group_by(cellBarcode, sampleBarcode) %>%
  summarise(n_TRA = sum(chain == "TRA"),
            n_TRB = sum(chain == "TRB")) %>%
  group_by(sampleBarcode, n_TRA, n_TRB) %>%
  summarise(n_cells = n()) %>%
  dplyr::arrange(sampleBarcode, n_TRA, n_TRB) %>%
  kable(row.names = F,
        caption = "Cells with single and multiple chains, by mouse") %>%
  kable_styling(full_width = F, 
                position = "left")

Pool3VdjUnfiltered %>%
  inner_join(Pool3InclMultiplets@meta.data) %>%
  group_by(cellBarcode, sampleBarcode) %>%
  summarise(n_TRA = sum(chain == "TRA"),
            n_TRB = sum(chain == "TRB")) %>%
  group_by(sampleBarcode, n_TRA, n_TRB) %>%
  summarise(n_cells = n()) %>%
  dplyr::arrange(sampleBarcode, n_TRA, n_TRB) %>%
  kable(row.names = F,
        caption = "Cells with single and multiple chains, by mouse") %>%
  kable_styling(full_width = F, 
                position = "left")

Pool4VdjUnfiltered %>%
  inner_join(Pool4InclMultiplets@meta.data) %>%
  group_by(cellBarcode, sampleBarcode) %>%
  summarise(n_TRA = sum(chain == "TRA"),
            n_TRB = sum(chain == "TRB")) %>%
  group_by(sampleBarcode, n_TRA, n_TRB) %>%
  summarise(n_cells = n()) %>%
  dplyr::arrange(sampleBarcode, n_TRA, n_TRB) %>%
  kable(row.names = F,
        caption = "Cells with single and multiple chains, by mouse") %>%
  kable_styling(full_width = F, 
                position = "left")
```

Some of the individuals cells show multiple heavy or light chains, despite the cells not being identified as multiplets. We need to look into these. We may want to exclude them entirely, or it's possible that there are biologically valid multiple-BCR B cells. Literature suggests that single cells with multiple light chains (IGK or IGL) are more likely to occur biologically than multiple heavy chains (IGH).

```{r identifyCellBarcodemMultipleChains, dependson=c("assignSampleBarcodes", "loadData10xVdj")}
Pool1MultipleChains <-
  Pool1VdjUnfiltered %>%
  inner_join(Pool1InclMultiplets@meta.data) %>%
  group_by(cellBarcode, sampleBarcode) %>%
  summarise(n_TRA = sum(chain == "TRA"),
            n_TRB = sum(chain == "TRB")) %>%
  dplyr::filter((n_TRA > 1) | (n_TRB > 1)) %>%
  dplyr::filter(sampleBarcode %in% sampleBarcodes) %>%
  dplyr::arrange(sampleBarcode)

Pool2MultipleChains <-
  Pool2VdjUnfiltered %>%
  inner_join(Pool2InclMultiplets@meta.data) %>%
  group_by(cellBarcode, sampleBarcode) %>%
  summarise(n_TRA = sum(chain == "TRA"),
            n_TRB = sum(chain == "TRB")) %>%
  dplyr::filter((n_TRA > 1) | (n_TRB > 1)) %>%
  dplyr::filter(sampleBarcode %in% sampleBarcodes) %>%
  dplyr::arrange(sampleBarcode)

Pool3MultipleChains <-
  Pool3VdjUnfiltered %>%
  inner_join(Pool3InclMultiplets@meta.data) %>%
  group_by(cellBarcode, sampleBarcode) %>%
  summarise(n_TRA = sum(chain == "TRA"),
            n_TRB = sum(chain == "TRB")) %>%
  dplyr::filter((n_TRA > 1) | (n_TRB > 1)) %>%
  dplyr::filter(sampleBarcode %in% sampleBarcodes) %>%
  dplyr::arrange(sampleBarcode)

Pool4MultipleChains <-
  Pool4VdjUnfiltered %>%
  inner_join(Pool4InclMultiplets@meta.data) %>%
  group_by(cellBarcode, sampleBarcode) %>%
  summarise(n_TRA = sum(chain == "TRA"),
            n_TRB = sum(chain == "TRB")) %>%
  dplyr::filter((n_TRA > 1) | (n_TRB > 1)) %>%
  dplyr::filter(sampleBarcode %in% sampleBarcodes) %>%
  dplyr::arrange(sampleBarcode)
```

## Filter VDJ sequences to only high-quality singlets, and pull in sampleName

```{r filterData10xVdj, dependson=c("loadData10xVdj", "removeCellBarcodeMultiplePositive")}
Pool1Vdj <-
  Pool1VdjUnfiltered %>%
  dplyr::filter(cellBarcode %in% Pool1$cellBarcode) %>%
  left_join(
    Pool1@meta.data %>% dplyr::select(cellBarcode, sampleName))

Pool2Vdj <-
  Pool2VdjUnfiltered %>%
  dplyr::filter(cellBarcode %in% Pool2$cellBarcode) %>%
  left_join(
    Pool2@meta.data %>% dplyr::select(cellBarcode, sampleName))

Pool3Vdj <-
  Pool3VdjUnfiltered %>%
  dplyr::filter(cellBarcode %in% Pool3$cellBarcode) %>%
  left_join(
    Pool3@meta.data %>% dplyr::select(cellBarcode, sampleName))

Pool4Vdj <-
  Pool4VdjUnfiltered %>%
  dplyr::filter(cellBarcode %in% Pool4$cellBarcode) %>%
  left_join(
    Pool4@meta.data %>% dplyr::select(cellBarcode, sampleName))
```

## Output VDJ sequences by mouse

```{r outputData10xVdjBySampleBarcode, dependson=c("filterData10xVdj")}
filenameOutputData10xVdjBySampleBarcode <-
  file.path(
    "data_output",
    paste0("Pool1_TCR_sequences_filtered_with_germline_by_mouse.", filenameSuffix, ".xlsx"))
Pool1Vdj %>%
  inner_join(
      Pool1@meta.data %>%
      dplyr::filter(sampleBarcode %in% sampleBarcodes) %>%
      dplyr::select(cellBarcode, mouse = sampleBarcode)) %>%
  dplyr::arrange(mouse, cellBarcode, chain, contigNumPerCell) %>%
  dplyr::select(mouse, cellBarcode, chain, everything()) %>%
  writexl::write_xlsx(path = filenameOutputData10xVdjBySampleBarcode, format_headers = FALSE)

filenameOutputData10xVdjBySampleBarcode <-
  file.path(
    "data_output",
    paste0("Pool2_TCR_sequences_filtered_with_germline_by_mouse.", filenameSuffix, ".xlsx"))
Pool2Vdj %>%
  inner_join(
      Pool2@meta.data %>%
      dplyr::filter(sampleBarcode %in% sampleBarcodes) %>%
      dplyr::select(cellBarcode, mouse = sampleBarcode)) %>%
  dplyr::arrange(mouse, cellBarcode, chain, contigNumPerCell) %>%
  dplyr::select(mouse, cellBarcode, chain, everything()) %>%
  writexl::write_xlsx(path = filenameOutputData10xVdjBySampleBarcode, format_headers = FALSE)

filenameOutputData10xVdjBySampleBarcode <-
  file.path(
    "data_output",
    paste0("Pool3_TCR_sequences_filtered_with_germline_by_mouse.", filenameSuffix, ".xlsx"))
Pool3Vdj %>%
  inner_join(
      Pool3@meta.data %>%
      dplyr::filter(sampleBarcode %in% sampleBarcodes) %>%
      dplyr::select(cellBarcode, mouse = sampleBarcode)) %>%
  dplyr::arrange(mouse, cellBarcode, chain, contigNumPerCell) %>%
  dplyr::select(mouse, cellBarcode, chain, everything()) %>%
  writexl::write_xlsx(path = filenameOutputData10xVdjBySampleBarcode, format_headers = FALSE)

filenameOutputData10xVdjBySampleBarcode <-
  file.path(
    "data_output",
    paste0("Pool4_TCR_sequences_filtered_with_germline_by_mouse.", filenameSuffix, ".xlsx"))
Pool4Vdj %>%
  inner_join(
      Pool4@meta.data %>%
      dplyr::filter(sampleBarcode %in% sampleBarcodes) %>%
      dplyr::select(cellBarcode, mouse = sampleBarcode)) %>%
  dplyr::arrange(mouse, cellBarcode, chain, contigNumPerCell) %>%
  dplyr::select(mouse, cellBarcode, chain, everything()) %>%
  writexl::write_xlsx(path = filenameOutputData10xVdjBySampleBarcode, format_headers = FALSE)
```

## Pull clonotype info into data10x

```{r pullClonotypeData, dependson=c("filterData10xVdj", "removeCellBarcodeMultiplePositive")}
Pool1@meta.data <-
  Pool1@meta.data %>%
  left_join(
    Pool1Vdj %>%
      dplyr::select(
        cellBarcode, 
        enclone_raw_clonotype_id = raw_clonotype_id,
        enclone_exact_subclonotype_id = exact_subclonotype_id) %>%
      unique()
  )
rownames(Pool1@meta.data) <- Pool1@meta.data$cellBarcode

Pool2@meta.data <-
  Pool2@meta.data %>%
  left_join(
    Pool2Vdj %>%
      dplyr::select(
        cellBarcode, 
        enclone_raw_clonotype_id = raw_clonotype_id,
        enclone_exact_subclonotype_id = exact_subclonotype_id) %>%
      unique()
  )
rownames(Pool2@meta.data) <- Pool2@meta.data$cellBarcode

Pool3@meta.data <-
  Pool3@meta.data %>%
  left_join(
    Pool3Vdj %>%
      dplyr::select(
        cellBarcode, 
        enclone_raw_clonotype_id = raw_clonotype_id,
        enclone_exact_subclonotype_id = exact_subclonotype_id) %>%
      unique()
  )
rownames(Pool3@meta.data) <- Pool3@meta.data$cellBarcode

Pool4@meta.data <-
  Pool4@meta.data %>%
  left_join(
    Pool4Vdj %>%
      dplyr::select(
        cellBarcode, 
        enclone_raw_clonotype_id = raw_clonotype_id,
        enclone_exact_subclonotype_id = exact_subclonotype_id) %>%
      unique()
  )
rownames(Pool4@meta.data) <- Pool4@meta.data$cellBarcode
```

## Plot proportion of cells from each mouse with heavy chain, light chain, both, or none

```{r calculateVdjChainsByCellBarcode, dependson="pullClonotypeData"}
Pool1@meta.data <-
  Pool1@meta.data %>%
  left_join(
    Pool1Vdj %>%
      dplyr::select(cellBarcode, chain) %>%
      table() %>%
      as.data.frame() %>%
      pivot_wider(
        names_from = chain, names_prefix = "nChain",
        values_from = Freq)) %>%
  dplyr::mutate(
    nChainTRA = case_when(is.na(nChainTRA) ~ 0L, TRUE ~ nChainTRA),
    nChainTRB = case_when(is.na(nChainTRB) ~ 0L, TRUE ~ nChainTRB),
    tcrGroup =
      case_when(
        ((nChainTRA == 0) & (nChainTRB == 0)) ~ "no TCR",
        ((nChainTRA > 0) & (nChainTRB == 0)) ~ "TRA only",
        ((nChainTRA == 0) & (nChainTRB > 0 )) ~ "TRB only",
        ((nChainTRA > 0) & (nChainTRB > 0)) ~ "Both Chains") %>%
      factor(levels = c("no TCR", "TRA only", "TRB only", "Both Chains")))
rownames(Pool1@meta.data) <- Pool1@meta.data$cellBarcode

Pool2@meta.data <-
  Pool2@meta.data %>%
  left_join(
    Pool2Vdj %>%
      dplyr::select(cellBarcode, chain) %>%
      table() %>%
      as.data.frame() %>%
      pivot_wider(
        names_from = chain, names_prefix = "nChain",
        values_from = Freq)) %>%
  dplyr::mutate(
    nChainTRA = case_when(is.na(nChainTRA) ~ 0L, TRUE ~ nChainTRA),
    nChainTRB = case_when(is.na(nChainTRB) ~ 0L, TRUE ~ nChainTRB),
    tcrGroup =
      case_when(
        ((nChainTRA == 0) & (nChainTRB == 0)) ~ "no TCR",
        ((nChainTRA > 0) & (nChainTRB == 0)) ~ "TRA only",
        ((nChainTRA == 0) & (nChainTRB > 0 )) ~ "TRB only",
        ((nChainTRA > 0) & (nChainTRB > 0)) ~ "Both Chains") %>%
      factor(levels = c("no TCR", "TRA only", "TRB only", "Both Chains")))
rownames(Pool2@meta.data) <- Pool2@meta.data$cellBarcode


Pool3@meta.data <-
  Pool3@meta.data %>%
  left_join(
    Pool3Vdj %>%
      dplyr::select(cellBarcode, chain) %>%
      table() %>%
      as.data.frame() %>%
      pivot_wider(
        names_from = chain, names_prefix = "nChain",
        values_from = Freq)) %>%
  dplyr::mutate(
    nChainTRA = case_when(is.na(nChainTRA) ~ 0L, TRUE ~ nChainTRA),
    nChainTRB = case_when(is.na(nChainTRB) ~ 0L, TRUE ~ nChainTRB),
    tcrGroup =
      case_when(
        ((nChainTRA == 0) & (nChainTRB == 0)) ~ "no TCR",
        ((nChainTRA > 0) & (nChainTRB == 0)) ~ "TRA only",
        ((nChainTRA == 0) & (nChainTRB > 0 )) ~ "TRB only",
        ((nChainTRA > 0) & (nChainTRB > 0)) ~ "Both Chains") %>%
      factor(levels = c("no TCR", "TRA only", "TRB only", "Both Chains")))
rownames(Pool3@meta.data) <- Pool3@meta.data$cellBarcode


Pool4@meta.data <-
  Pool4@meta.data %>%
  left_join(
    Pool4Vdj %>%
      dplyr::select(cellBarcode, chain) %>%
      table() %>%
      as.data.frame() %>%
      pivot_wider(
        names_from = chain, names_prefix = "nChain",
        values_from = Freq)) %>%
  dplyr::mutate(
    nChainTRA = case_when(is.na(nChainTRA) ~ 0L, TRUE ~ nChainTRA),
    nChainTRB = case_when(is.na(nChainTRB) ~ 0L, TRUE ~ nChainTRB),
    tcrGroup =
      case_when(
        ((nChainTRA == 0) & (nChainTRB == 0)) ~ "no TCR",
        ((nChainTRA > 0) & (nChainTRB == 0)) ~ "TRA only",
        ((nChainTRA == 0) & (nChainTRB > 0 )) ~ "TRB only",
        ((nChainTRA > 0) & (nChainTRB > 0)) ~ "Both Chains") %>%
      factor(levels = c("no TCR", "TRA only", "TRB only", "Both Chains")))
rownames(Pool4@meta.data) <- Pool4@meta.data$cellBarcode


# calculate number of cells with each combination of chains
tcrChainCountsPool1 <-
  Pool1@meta.data %>%
  group_by(sampleName, tcrGroup) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()

tcrChainCountsPool2 <-
  Pool2@meta.data %>%
  group_by(sampleName, tcrGroup) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()

tcrChainCountsPool3 <-
  Pool3@meta.data %>%
  group_by(sampleName, tcrGroup) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()

tcrChainCountsPool4 <-
  Pool4@meta.data %>%
  group_by(sampleName, tcrGroup) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()
```

```{r plotVdjChainsByCellBarcode, dependson="calculateVdjChainsByCellBarcode"}
sampleNameLabels.tmp <-
  tcrChainCountsPool1 %>%
  dplyr::select(sampleName, totalCount) %>%
  unique() %>%
  mutate(
    sampleNameLabel = paste0(sampleName, " (", totalCount, ")"),
    freq = 0.1) %>%
  dplyr::arrange(sampleName)
 
plot.tmp <-
  ggplot(
    data = tcrChainCountsPool1,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity", mapping = aes(fill = tcrGroup)) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(
    values = c("black", "blue", "orange", "red"), 
    labels = c("no TCR", "TRA only", "TRB only", "Both Chains")) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrChainDetectionPool1.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

rm_tmp(ask=FALSE)

sampleNameLabels.tmp <-
  tcrChainCountsPool2 %>%
  dplyr::select(sampleName, totalCount) %>%
  unique() %>%
  mutate(
    sampleNameLabel = paste0(sampleName, " (", totalCount, ")"),
    freq = 0.1) %>%
  dplyr::arrange(sampleName)
 
plot.tmp <-
  ggplot(
    data = tcrChainCountsPool2,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity", mapping = aes(fill = tcrGroup)) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(
    values = c("black", "blue", "orange", "red"), 
    labels = c("no TCR", "TRA only", "TRB only", "Both Chains")) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrChainDetectionPool2.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

rm_tmp(ask=FALSE)

sampleNameLabels.tmp <-
  tcrChainCountsPool3 %>%
  dplyr::select(sampleName, totalCount) %>%
  unique() %>%
  mutate(
    sampleNameLabel = paste0(sampleName, " (", totalCount, ")"),
    freq = 0.1) %>%
  dplyr::arrange(sampleName)
 
plot.tmp <-
  ggplot(
    data = tcrChainCountsPool3,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity", mapping = aes(fill = tcrGroup)) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(
    values = c("black", "blue", "orange", "red"), 
    labels = c("no TCR", "TRA only", "TRB only", "Both Chains")) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrChainDetectionPool3.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

rm_tmp(ask=FALSE)

sampleNameLabels.tmp <-
  tcrChainCountsPool4 %>%
  dplyr::select(sampleName, totalCount) %>%
  unique() %>%
  mutate(
    sampleNameLabel = paste0(sampleName, " (", totalCount, ")"),
    freq = 0.1) %>%
  dplyr::arrange(sampleName)
 
plot.tmp <-
  ggplot(
    data = tcrChainCountsPool4,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity", mapping = aes(fill = tcrGroup)) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(
    values = c("black", "blue", "orange", "red"), 
    labels = c("no TCR", "TRA only", "TRB only", "Both Chains")) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrChainDetectionPool4.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

rm_tmp(ask=FALSE)
```

## Plot V and J gene usage from each mouse for heavy and light chain

```{r calculateVdjGeneUsageBySampleName, dependson="pullClonotypeData"}
TcrGeneCountsPool1 <- list()
TcrGeneCountsPool1[["TRA"]] <- list()
TcrGeneCountsPool1[["TRA"]][["v_gene"]] <-
  Pool1Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool1[["TRA"]][["v_gene_family"]] <-
  Pool1Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  mutate(gene = str_extract(gene, "^[A-Z0-9]+")) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool1[["TRA"]][["j_gene"]] <-
  Pool1Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = j_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))

TcrGeneCountsPool1[["TRB"]] <- list()
TcrGeneCountsPool1[["TRB"]][["v_gene"]] <-
  Pool1Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool1[["TRB"]][["v_gene_family"]] <-
  Pool1Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  mutate(gene = str_extract(gene, "^[A-Z0-9]+")) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool1[["TRB"]][["j_gene"]] <-
  Pool1Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = j_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))

TcrGeneCountsPool2 <- list()
TcrGeneCountsPool2[["TRA"]] <- list()
TcrGeneCountsPool2[["TRA"]][["v_gene"]] <-
  Pool2Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool2[["TRA"]][["v_gene_family"]] <-
  Pool2Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  mutate(gene = str_extract(gene, "^[A-Z0-9]+")) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool2[["TRA"]][["j_gene"]] <-
  Pool2Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = j_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))

TcrGeneCountsPool2[["TRB"]] <- list()
TcrGeneCountsPool2[["TRB"]][["v_gene"]] <-
  Pool2Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool2[["TRB"]][["v_gene_family"]] <-
  Pool2Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  mutate(gene = str_extract(gene, "^[A-Z0-9]+")) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool2[["TRB"]][["j_gene"]] <-
  Pool1Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = j_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))

TcrGeneCountsPool3 <- list()
TcrGeneCountsPool3[["TRA"]] <- list()
TcrGeneCountsPool3[["TRA"]][["v_gene"]] <-
  Pool3Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool3[["TRA"]][["v_gene_family"]] <-
  Pool3Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  mutate(gene = str_extract(gene, "^[A-Z0-9]+")) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool3[["TRA"]][["j_gene"]] <-
  Pool3Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = j_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))

TcrGeneCountsPool3[["TRB"]] <- list()
TcrGeneCountsPool3[["TRB"]][["v_gene"]] <-
  Pool3Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool3[["TRB"]][["v_gene_family"]] <-
  Pool3Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  mutate(gene = str_extract(gene, "^[A-Z0-9]+")) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool3[["TRB"]][["j_gene"]] <-
  Pool3Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = j_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))

TcrGeneCountsPool4 <- list()
TcrGeneCountsPool4[["TRA"]] <- list()
TcrGeneCountsPool4[["TRA"]][["v_gene"]] <-
  Pool4Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool4[["TRA"]][["v_gene_family"]] <-
  Pool4Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  mutate(gene = str_extract(gene, "^[A-Z0-9]+")) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool4[["TRA"]][["j_gene"]] <-
  Pool4Vdj %>%
  dplyr::filter(chain == "TRA") %>%
  dplyr::select(sampleName, gene = j_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))

TcrGeneCountsPool4[["TRB"]] <- list()
TcrGeneCountsPool4[["TRB"]][["v_gene"]] <-
  Pool4Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool4[["TRB"]][["v_gene_family"]] <-
  Pool4Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = v_gene) %>%
  mutate(gene = str_extract(gene, "^[A-Z0-9]+")) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
TcrGeneCountsPool4[["TRB"]][["j_gene"]] <-
  Pool4Vdj %>%
  dplyr::filter(chain == "TRB") %>%
  dplyr::select(sampleName, gene = j_gene) %>%
  group_by(sampleName, gene) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  ungroup() %>%
  mutate(
    gene = gene %>%
      factor(levels = str_sort(unique(.), numeric=TRUE)))
```

```{r plotVdjGeneUsageBySampleName, dependson="calculateVdjGeneUsageBySampleName"}
for (chain.tmp in names(TcrGeneCountsPool1)) {
  for (gene.tmp in names(TcrGeneCountsPool1[[chain.tmp]])) {
    plot.tmp <-
      ggplot(
        data = TcrGeneCountsPool1[[chain.tmp]][[gene.tmp]],
        aes(x = sampleName, y = freq*100)) +
      geom_bar(stat = "identity", mapping = aes(fill = gene)) +
      scale_fill_manual(
        values = big_colorblind_pal(length(levels(TcrGeneCountsPool1[[chain.tmp]][[gene.tmp]]$gene))),
        labels = levels(TcrGeneCountsPool1[[chain.tmp]][[gene.tmp]]$gene)) +
      guides(fill = "none") +
      labs(x = NULL, y = "percent of cells", fill = NULL,
           title = paste0(chain.tmp, " chain, ", gene.tmp)) +
      theme(axis.text.x = element_text(angle = -45, hjust = 0),
            text = element_text(size=14))
   
    # output plot
    pdf(
      file.path("plots", 
                paste("barplotTcrGeneUsagePool1", chain.tmp, gene.tmp, filenameSuffix, "pdf", sep = ".")),
      height = 5, width = 10)
    print(plot.tmp)
    invisible(dev.off())
    
  }
}

for (chain.tmp in names(TcrGeneCountsPool2)) {
  for (gene.tmp in names(TcrGeneCountsPool2[[chain.tmp]])) {
    plot.tmp <-
      ggplot(
        data = TcrGeneCountsPool2[[chain.tmp]][[gene.tmp]],
        aes(x = sampleName, y = freq*100)) +
      geom_bar(stat = "identity", mapping = aes(fill = gene)) +
      scale_fill_manual(
        values = big_colorblind_pal(length(levels(TcrGeneCountsPool2[[chain.tmp]][[gene.tmp]]$gene))),
        labels = levels(TcrGeneCountsPool2[[chain.tmp]][[gene.tmp]]$gene)) +
      guides(fill = "none") +
      labs(x = NULL, y = "percent of cells", fill = NULL,
           title = paste0(chain.tmp, " chain, ", gene.tmp)) +
      theme(axis.text.x = element_text(angle = -45, hjust = 0),
            text = element_text(size=14))
   
    # output plot
    pdf(
      file.path("plots", 
                paste("barplotTcrGeneUsagePool2", chain.tmp, gene.tmp, filenameSuffix, "pdf", sep = ".")),
      height = 5, width = 10)
    print(plot.tmp)
    invisible(dev.off())
    
  }
}

for (chain.tmp in names(TcrGeneCountsPool3)) {
  for (gene.tmp in names(TcrGeneCountsPool3[[chain.tmp]])) {
    plot.tmp <-
      ggplot(
        data = TcrGeneCountsPool3[[chain.tmp]][[gene.tmp]],
        aes(x = sampleName, y = freq*100)) +
      geom_bar(stat = "identity", mapping = aes(fill = gene)) +
      scale_fill_manual(
        values = big_colorblind_pal(length(levels(TcrGeneCountsPool3[[chain.tmp]][[gene.tmp]]$gene))),
        labels = levels(TcrGeneCountsPool3[[chain.tmp]][[gene.tmp]]$gene)) +
      guides(fill = "none") +
      labs(x = NULL, y = "percent of cells", fill = NULL,
           title = paste0(chain.tmp, " chain, ", gene.tmp)) +
      theme(axis.text.x = element_text(angle = -45, hjust = 0),
            text = element_text(size=14))
   
    # output plot
    pdf(
      file.path("plots", 
                paste("barplotTcrGeneUsagePool3", chain.tmp, gene.tmp, filenameSuffix, "pdf", sep = ".")),
      height = 5, width = 10)
    print(plot.tmp)
    invisible(dev.off())
    
  }
}

for (chain.tmp in names(TcrGeneCountsPool4)) {
  for (gene.tmp in names(TcrGeneCountsPool4[[chain.tmp]])) {
    plot.tmp <-
      ggplot(
        data = TcrGeneCountsPool4[[chain.tmp]][[gene.tmp]],
        aes(x = sampleName, y = freq*100)) +
      geom_bar(stat = "identity", mapping = aes(fill = gene)) +
      scale_fill_manual(
        values = big_colorblind_pal(length(levels(TcrGeneCountsPool4[[chain.tmp]][[gene.tmp]]$gene))),
        labels = levels(TcrGeneCountsPool4[[chain.tmp]][[gene.tmp]]$gene)) +
      guides(fill = "none") +
      labs(x = NULL, y = "percent of cells", fill = NULL,
           title = paste0(chain.tmp, " chain, ", gene.tmp)) +
      theme(axis.text.x = element_text(angle = -45, hjust = 0),
            text = element_text(size=14))
   
    # output plot
    pdf(
      file.path("plots", 
                paste("barplotTcrGeneUsagePool4", chain.tmp, gene.tmp, filenameSuffix, "pdf", sep = ".")),
      height = 5, width = 10)
    print(plot.tmp)
    invisible(dev.off())
    
  }
}

rm_tmp(ask=FALSE)
```

*Continue here, adding plots of BCR gene usage by links, like Mark Anderson did*

## Call BCR isotypes from heavy chain constant gene

```{r callBcrIsotype, dependson=c("filterData10xVdj", "removeCellBarcodeMultiplePositive")}
# this defines the BCR isotype based on the C gene in the heavy chain of the BCR sequence
# if there is not heavy chain detected, or if there are multipe heavy chains with different C genes, the isotype is left ambiguous
Pool1@meta.data <-
  Pool1@meta.data %>%
  left_join(
    Pool1Vdj %>%
      dplyr::filter(chain == "TRB") %>%
      dplyr::select(cellBarcode, c_gene) %>%
      unique() %>%
      group_by(cellBarcode) %>%
      mutate(count = n()) %>%
      dplyr::filter(count == 1) %>%
      mutate(
        isotypeFromTcrBChain =
          c_gene %>%
          str_replace("TRBC2", "TRbc2") %>%
          #str_replace("TRAC", "TRaC") %>%
          str_replace("TRBC1", "TRbc1")) %>%
      dplyr::select(cellBarcode, isotypeFromTcrBChain)
  ) %>%
  # replace NAs with "undetermined"
  mutate(
    isotypeFromTcrBChain = 
      case_when(
        is.na(isotypeFromTcrBChain) ~ "undetermined",
        isotypeFromTcrBChain == "" ~ "undetermined", # some blanks
        TRUE ~ isotypeFromTcrBChain) %>%
      factor(levels = names(pal.isotype)))
rownames(Pool1@meta.data) <- Pool1@meta.data$cellBarcode

Pool2@meta.data <-
  Pool2@meta.data %>%
  left_join(
    Pool2Vdj %>%
      dplyr::filter(chain == "TRB") %>%
      dplyr::select(cellBarcode, c_gene) %>%
      unique() %>%
      group_by(cellBarcode) %>%
      mutate(count = n()) %>%
      dplyr::filter(count == 1) %>%
      mutate(
        isotypeFromTcrBChain =
          c_gene %>%
          str_replace("TRBC2", "TRbc2") %>%
          #str_replace("TRAC", "TRaC") %>%
          str_replace("TRBC1", "TRbc1")) %>%
      dplyr::select(cellBarcode, isotypeFromTcrBChain)
  ) %>%
  # replace NAs with "undetermined"
  mutate(
    isotypeFromTcrBChain = 
      case_when(
        is.na(isotypeFromTcrBChain) ~ "undetermined",
        isotypeFromTcrBChain == "" ~ "undetermined", # some blanks
        TRUE ~ isotypeFromTcrBChain) %>%
      factor(levels = names(pal.isotype)))
rownames(Pool2@meta.data) <- Pool2@meta.data$cellBarcode

Pool3@meta.data <-
  Pool3@meta.data %>%
  left_join(
    Pool3Vdj %>%
      dplyr::filter(chain == "TRB") %>%
      dplyr::select(cellBarcode, c_gene) %>%
      unique() %>%
      group_by(cellBarcode) %>%
      mutate(count = n()) %>%
      dplyr::filter(count == 1) %>%
      mutate(
        isotypeFromTcrBChain =
          c_gene %>%
          str_replace("TRBC2", "TRbc2") %>%
          #str_replace("TRAC", "TRaC") %>%
          str_replace("TRBC1", "TRbc1")) %>%
      dplyr::select(cellBarcode, isotypeFromTcrBChain)
  ) %>%
  # replace NAs with "undetermined"
  mutate(
    isotypeFromTcrBChain = 
      case_when(
        is.na(isotypeFromTcrBChain) ~ "undetermined",
        isotypeFromTcrBChain == "" ~ "undetermined", # some blanks
        TRUE ~ isotypeFromTcrBChain) %>%
      factor(levels = names(pal.isotype)))
rownames(Pool3@meta.data) <- Pool3@meta.data$cellBarcode

Pool4@meta.data <-
  Pool4@meta.data %>%
  left_join(
    Pool4Vdj %>%
      dplyr::filter(chain == "TRB") %>%
      dplyr::select(cellBarcode, c_gene) %>%
      unique() %>%
      group_by(cellBarcode) %>%
      mutate(count = n()) %>%
      dplyr::filter(count == 1) %>%
      mutate(
        isotypeFromTcrBChain =
          c_gene %>%
          str_replace("TRBC2", "TRbc2") %>%
          #str_replace("TRAC", "TRaC") %>%
          str_replace("TRBC1", "TRbc1")) %>%
      dplyr::select(cellBarcode, isotypeFromTcrBChain)
  ) %>%
  # replace NAs with "undetermined"
  mutate(
    isotypeFromTcrBChain = 
      case_when(
        is.na(isotypeFromTcrBChain) ~ "undetermined",
        isotypeFromTcrBChain == "" ~ "undetermined", # some blanks
        TRUE ~ isotypeFromTcrBChain) %>%
      factor(levels = names(pal.isotype)))
rownames(Pool4@meta.data) <- Pool4@meta.data$cellBarcode

# calculate number of cells with each isotype
tcrIsotypeCountsPool1 <-
  Pool1@meta.data %>%
  group_by(sampleName, isotypeFromTcrBChain) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()

tcrIsotypeCountsPool4 <-
  Pool4@meta.data %>%
  group_by(sampleName, isotypeFromTcrBChain) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()

tcrIsotypeCountsPool2 <-
  Pool2@meta.data %>%
  group_by(sampleName, isotypeFromTcrBChain) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()

tcrIsotypeCountsPool3 <-
  Pool3@meta.data %>%
  group_by(sampleName, isotypeFromTcrBChain) %>%
  summarise(count = n()) %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()

# calculate number of cells with each isotype
tcrIsotypeCountsCalledOnlyPool1 <-
  tcrIsotypeCountsPool1 %>%
  dplyr::filter(isotypeFromTcrBChain %nin% "undetermined") %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()

tcrIsotypeCountsCalledOnlyPool2 <-
  tcrIsotypeCountsPool2 %>%
  dplyr::filter(isotypeFromTcrBChain %nin% "undetermined") %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()

tcrIsotypeCountsCalledOnlyPool3 <-
  tcrIsotypeCountsPool3 %>%
  dplyr::filter(isotypeFromTcrBChain %nin% "undetermined") %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()

tcrIsotypeCountsCalledOnlyPool4 <-
  tcrIsotypeCountsPool4 %>%
  dplyr::filter(isotypeFromTcrBChain %nin% "undetermined") %>%
  group_by(sampleName) %>%
  mutate(
    totalCount = sum(count),
    freq = count / totalCount) %>%
  droplevels()
```

```{r plotBcrIsotypeByCellBarcode, dependson="callBcrIsotype"}
sampleNameLabels.tmp <-
  tcrIsotypeCountsPool1 %>%
  dplyr::select(sampleName, totalCount) %>%
  unique() %>%
  mutate(
    sampleNameLabel = paste0(sampleName, " (", totalCount, ")"),
    freq = 0.9) %>%
  dplyr::arrange(sampleName)
 
plot.tmp <-
  ggplot(
    data = tcrIsotypeCountsPool1,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity",
           mapping = aes(fill = forcats::fct_rev(isotypeFromTcrBChain))) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(values = pal.isotype) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrBetaByCellBarcodePool1.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

rm_tmp(ask=FALSE)

sampleNameLabels.tmp <-
  tcrIsotypeCountsPool2 %>%
  dplyr::select(sampleName, totalCount) %>%
  unique() %>%
  mutate(
    sampleNameLabel = paste0(sampleName, " (", totalCount, ")"),
    freq = 0.9) %>%
  dplyr::arrange(sampleName)
 
plot.tmp <-
  ggplot(
    data = tcrIsotypeCountsPool2,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity",
           mapping = aes(fill = forcats::fct_rev(isotypeFromTcrBChain))) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(values = pal.isotype) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrBetaByCellBarcodePool2.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

rm_tmp(ask=FALSE)

sampleNameLabels.tmp <-
  tcrIsotypeCountsPool3 %>%
  dplyr::select(sampleName, totalCount) %>%
  unique() %>%
  mutate(
    sampleNameLabel = paste0(sampleName, " (", totalCount, ")"),
    freq = 0.9) %>%
  dplyr::arrange(sampleName)
 
plot.tmp <-
  ggplot(
    data = tcrIsotypeCountsPool3,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity",
           mapping = aes(fill = forcats::fct_rev(isotypeFromTcrBChain))) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(values = pal.isotype) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrBetaByCellBarcodePool3.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

rm_tmp(ask=FALSE)

sampleNameLabels.tmp <-
  tcrIsotypeCountsPool1 %>%
  dplyr::select(sampleName, totalCount) %>%
  unique() %>%
  mutate(
    sampleNameLabel = paste0(sampleName, " (", totalCount, ")"),
    freq = 0.9) %>%
  dplyr::arrange(sampleName)
 
plot.tmp <-
  ggplot(
    data = tcrIsotypeCountsPool4,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity",
           mapping = aes(fill = forcats::fct_rev(isotypeFromTcrBChain))) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(values = pal.isotype) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrBetaByCellBarcodePool4.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

rm_tmp(ask=FALSE)
```

```{r plotBcrIsotypeByCellBarcodeCalledOnly, dependson="callBcrIsotype"}
sampleNameLabels.tmp <-
  tcrIsotypeCountsCalledOnlyPool1 %>%
  dplyr::select(sampleName, totalCount) %>%
  unique() %>%
  mutate(
    sampleNameLabel = paste0(sampleName, " (", totalCount, ")"),
    freq = 0.9) %>%
  dplyr::arrange(sampleName)
 
plot.tmp <-
  ggplot(
    data =
      tcrIsotypeCountsCalledOnlyPool1,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity", mapping = aes(fill = forcats::fct_rev(isotypeFromTcrBChain))) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(
    values = 
      pal.isotype[names(pal.isotype) %in% tcrIsotypeCountsCalledOnlyPool1$isotypeFromTcrBChain]) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrBetaByCellBarcodeCalledOnlyPool1.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggplot(
    data =
      tcrIsotypeCountsCalledOnlyPool2,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity", mapping = aes(fill = forcats::fct_rev(isotypeFromTcrBChain))) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(
    values = 
      pal.isotype[names(pal.isotype) %in% tcrIsotypeCountsCalledOnlyPool2$isotypeFromTcrBChain]) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrBetaByCellBarcodeCalledOnlyPool2.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggplot(
    data =
      tcrIsotypeCountsCalledOnlyPool3,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity", mapping = aes(fill = forcats::fct_rev(isotypeFromTcrBChain))) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(
    values = 
      pal.isotype[names(pal.isotype) %in% tcrIsotypeCountsCalledOnlyPool3$isotypeFromTcrBChain]) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrBetaByCellBarcodeCalledOnlyPool3.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

plot.tmp <-
  ggplot(
    data =
      tcrIsotypeCountsCalledOnlyPool4,
    aes(x = sampleName, y = freq*100)) +
  geom_bar(stat = "identity", mapping = aes(fill = forcats::fct_rev(isotypeFromTcrBChain))) +
  geom_text(
    data = sampleNameLabels.tmp,
    mapping = aes(label = totalCount),
    size = 4) +
  scale_fill_manual(
    values = 
      pal.isotype[names(pal.isotype) %in% tcrIsotypeCountsCalledOnlyPool4$isotypeFromTcrBChain]) +
  labs(x = NULL, y = "percent of cells", fill = NULL, title = NULL) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0),
        text = element_text(size=14))

# output plot
pdf(
  file.path("plots", paste0("barplotTcrBetaByCellBarcodeCalledOnlyPool4.pdf")),
    height = 5, width = 10)
print(plot.tmp)
invisible(dev.off())

rm_tmp(ask=FALSE)
```

# Calculate gene set scores

*Continue from here, 2021-11-22, need to update gene sets and calculate gene sets scores*

```{r calculateGeneSetScoresBCells, dependson=c("addMetadataFromDesign", "loadGeneSets"), eval=FALSE}
# need to ensure that RNA data are normalized before doing this calculation
Pool1 <-
  NormalizeData(
    Pool1, assay = "RNA", normalization.method = "LogNormalize")

Pool2 <-
  NormalizeData(
    Pool2, assay = "RNA", normalization.method = "LogNormalize")

Pool3 <-
  NormalizeData(
    Pool3, assay = "RNA", normalization.method = "LogNormalize")

Pool4 <-
  NormalizeData(
    Pool4, assay = "RNA", normalization.method = "LogNormalize")

Pool1 <- FindVariableFeatures(Pool1, selection.method = "vst", nfeatures = 2000)

Pool2 <- FindVariableFeatures(Pool2, selection.method = "vst", nfeatures = 2000)

Pool3 <- FindVariableFeatures(Pool3, selection.method = "vst", nfeatures = 2000)

Pool4 <- FindVariableFeatures(Pool4, selection.method = "vst", nfeatures = 2000)

features <- SelectIntegrationFeatures(object.list = list(Pool1, Pool2, Pool3, Pool4))

anchors <- FindIntegrationAnchors(object.list = list(Pool1, Pool2, Pool3, Pool4), anchor.features = features)

Pool1$cellBarcode <- sub("^", "Pool1", Pool1$cellBarcode)

Pool2$cellBarcode <- sub("^", "Pool2", Pool2$cellBarcode)
Pool3$cellBarcode <- sub("^", "Pool3", Pool3$cellBarcode)
Pool4$cellBarcode <- sub("^", "Pool4", Pool4$cellBarcode)

combinedPools <- IntegrateData(anchorset = anchors)

combinedPools@meta.data$geneSetMean... <-
  combinedPools@assays$RNA@data[
    which(rownames(combinedPools@assays$RNA@data) %in%
            geneSet...),] %>%
  apply(MARGIN = 2, FUN = mean)

saveRDS(combinedPools, "poolsAfterIntegration.rd")
```

# Normalize, cluster, project, and plot data

```{r createClusterObjects}
clusterResolution <-
    c("RNA" = 0.25)
# running with 0.8 yielded 11 clusters, too many!
# running with 0.5 yielded 10 clusters, still too many
# running with 0.3 yielded 7 clusters, which looks right

clusterName <- character()

clusterMarkers <- list()
```

## Normalize, cluster, project, and plot RNAseq data

```{r normalizeClusterProjectRnaData, dependson=c("removeCellBarcodeMultiplePositive", "createClusterObjects")}
# perform visualization and clustering steps

# normalize data (skip for integrated data, as data already normalized)

DefaultAssay(combinedPools) <- "RNA"

combinedPools <- NormalizeData(combinedPools, assay = "RNA", normalization.method = "LogNormalize")

# find variable featuers (skip for integrated data, as variable features already found)
combinedPools <- FindVariableFeatures(combinedPools, assay = "RNA")

# scale data (skip for integrated data, as data already scaled)
combinedPools <- ScaleData(combinedPools, assay = "RNA")

# run PCA
combinedPools <- 
  RunPCA(combinedPools, assay = "RNA", verbose = FALSE,
         reduction.name = "pca_RNA", reduction.key = "pcRNA_")

# find nearest neighbors and shared nearest neighbors
combinedPools <-
  FindNeighbors(
    combinedPools, reduction = "pca_RNA", dims = 1:30,
    graph.name = c("nn_RNA", "snn_RNA"))

# find clusters
#rownames(combinedPools@meta.data) <- combinedPools@meta.data$cellBarcode
#rownames(combinedPools) <- combinedPools$cellBarcode
combinedPools <- FindClusters(
    combinedPools, graph.name = "snn_RNA",
    resolution = .2,
    verbose = TRUE)

# make cluster order more logical; will need to be tweaked if anything changes upstream
Idents(combinedPools) <-
  Idents(combinedPools) %>%
  dplyr::recode(
    "6" = "0",
    "1" = "1",
    "2" = "2",
    "3" = "3",
    "0" = "4",
    "4" = "5",
    "5" = "6") %>%
  factor(levels = str_sort(unique(.), numeric = TRUE))

# store clusters from RNA in metadata
clusterName[["RNA"]] <- paste0("seurat_clusters_RNA_", str_replace(clusterResolution[["RNA"]], "\\.", "p"))
combinedPools <- combinedPools %>%
  AddMetaData(Idents(combinedPools), col.name = clusterName[["RNA"]])

# run UMAP
combinedPools <-
  RunUMAP(combinedPools, reduction = "pca_RNA", dims = 1:30,
          reduction.name = "umap_RNA", reduction.key = "umapRNA_")

# run tSNE
combinedPools <- 
  RunTSNE(combinedPools, reduction = "pca_RNA", dims = 1:30,
          reduction.name = "tsne_RNA", reduction.key = "tsneRNA_")

# save version of data10x for easy recovery
CombinedPoolsClusterProjectRna <-
  file.path("data_saved",
            paste0("aggrData10xClusterProjectRna.", filenameSuffix, ".RDS"))
saveRDS(combinedPools, CombinedPoolsClusterProjectRna)
# data10x <- readRDS(filenameData10xClusterProjectRna)
```

```{r elbowPlotPcaRnaData, dependson="normalizeClusterProjectRnaData", fig.width=8, fig.height=6}
ElbowPlot(combinedPools, reduction = "pca_RNA")
# looks like 4 or 6 or 13 PCs would be best
```

```{r findClusterMarkersRnaData, dependson=c("normalizeClusterProjectRnaData", "createClusterObjects")}
Idents(combinedPools) <- combinedPools@meta.data[[clusterName[["RNA"]]]]
clusterMarkers[[clusterName[["RNA"]]]] <-
  FindAllMarkers(combinedPools, assay = "RNA")
```

```{r writeOutClusterMarkersRnaData, dependson="findClusterMarkersRnaData"}
nGenesPerCluster.tmp <- 500

file.tmp <-
  file.path(
    "data_output",
    paste0("clusterMarkers.",
           clusterName[["RNA"]], ".",
           filenameSuffix, ".xlsx"))

clusterMarkers[[clusterName]] %>%
  dplyr::arrange(cluster, p_val) %>%
  dplyr::group_by(cluster) %>%
  dplyr::slice(1:nGenesPerCluster.tmp) %>%
  dplyr::select(cluster, gene, everything()) %>%
  writexl::write_xlsx(path = file.tmp, format_headers = FALSE)
```

### Plot UMAP of RNAseq data by cluster

```{r plotRnaUmapColorByCluster, dependson="normalizeClusterProjectRnaData", fig.width=9.2, fig.height=7}
plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = clusterName[["RNA"]]), size = .75, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "T-reg Category", 
    values = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]])),
    labels = c("Central", "Effector", "STAT5-related", "Cycling_1", "Cycling_2", "Recently activated")) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("rnaUMAP_BiologicalClass", filenameSuffix),
  format = plotOutputFormat,
  w = 9, h = 7)

rm_tmp(ask=FALSE)

plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = clusterName[["RNA"]]), size = .5, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) +
  facet_wrap(~treatment, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapColorByClusterByFacet", filenameSuffix),
  format = plotOutputFormat,
  w = 9, h = 7)
```

```{feature plots}

plot.tmp <- FeaturePlot(combinedPools, "Foxp3")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Foxp3"),
  format = plotOutputFormat,
  w = 9, h = 7)

VlnPlot(combinedPools, "Il2ra", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)


plot.tmp <- FeaturePlot(combinedPools, "Mki67")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Mki67"),
  format = plotOutputFormat,
  w = 9, h = 7)


plot.tmp <- FeaturePlot(combinedPools, "Il2ra")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Il2ra"),
  format = plotOutputFormat,
  w = 9, h = 7)


plot.tmp <- FeaturePlot(combinedPools, "Ly6c1")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Ly6c1"),
  format = plotOutputFormat,
  w = 9, h = 7)


plot.tmp <- FeaturePlot(combinedPools, "Ccr7")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Ccr7"),
  format = plotOutputFormat,
  w = 9, h = 7)


plot.tmp <- FeaturePlot(combinedPools, "Icos")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Icos"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Ctla4")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Ctla4"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Pdcd1")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Pdcd1"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Flt3l")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Flt31"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Bach2")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Bach2"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Tcf7")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Tcf7"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Lef1")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Lef1"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Klf2")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Klf2"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Maf")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Maf"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Nfatc1")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Nfatc1"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Batf")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Batf"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- FeaturePlot(combinedPools, "Hif1a")

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Hif1a"),
  format = plotOutputFormat,
  w = 9, h = 7)

#withviolins

plot.tmp <- VlnPlot(combinedPools, "Foxp3", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Foxp3") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Foxp3Viol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Mki67", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Mki67") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Mki67Viol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Il2ra", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Il2ra") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Il2raViol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Ly6c1", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Ly6c1") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Ly6c1Vio"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Ccr7", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Ccr7") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Ccr7Viol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Icos", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Icos") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("IcosViol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Ctla4", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Ctla4") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Ctla4Viol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Pdcd1", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Pdcd1") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Pdcd1Viol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Flt3l", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Flt3l") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Flt3lViol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Bach2", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Bach2") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Bach2Viol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Tcf7", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Tcf7") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Tcf7"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Lef1", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Lef1") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Lef1Viol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Klf2", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Klf2") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Klf2Viol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Maf", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Maf") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("MafViol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Nfatc1", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Nfatc1") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Nfatc1Viol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Batf", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Batf") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("BatfViol"),
  format = plotOutputFormat,
  w = 9, h = 7)

plot.tmp <- VlnPlot(combinedPools, "Hif1a", group.by = "treatment", pt.size = FALSE) + geom_boxplot(width = .1)
plot.tmp <- FeaturePlot(combinedPools, "Hif1a") + plot.tmp

outputPlots(
  plot.tmp, fileDir = "plots/featureplots",
  fileStem = 
    paste0("Hif1aViol"),
  format = plotOutputFormat,
  w = 9, h = 7)

#Viols by Cluster

 plot3.tmp <- ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = clusterName[["RNA"]]), size = .75, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))
  
plot.tmp <- VlnPlot(combinedPools, "Foxp3", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Foxp3")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)


grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))


pdf(
  file.path("plots/featureplots", paste0("Foxp3ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Mki67", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Mki67")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Mki67ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Il2ra", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Il2ra")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Il2raViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Ly6c1", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Ly6c1")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Ly6c1ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Ccr7", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Ccr7")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Ccr71ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Icos", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Icos")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Icos1ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Ctla4", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Ctla4")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Ctla4ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Pdcd1", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Pdcd1")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Pdcd1ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Flt3l", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Flt3l")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Flt3lViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Bach2", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Bach2")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Bach2ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Tcf7", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Tcf7")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Tcf7ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Lef1", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Lef1")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Lef1ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()


plot.tmp <- VlnPlot(combinedPools, "Klf2", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Klf2")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Klf2ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Maf", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Maf")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("MafViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()

plot.tmp <- VlnPlot(combinedPools, "Nfatc1", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Nfatc1")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Nfatc1ViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()


plot.tmp <- VlnPlot(combinedPools, "Batf", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Batf")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("BatfViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()


plot.tmp <- VlnPlot(combinedPools, "Hif1a", group.by = "seurat_clusters", pt.size = FALSE, cols = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) + geom_boxplot(width = .1)
plot2.tmp <- FeaturePlot(combinedPools, "Hif1a")
plots <- list(plot3.tmp, plot2.tmp, plot.tmp)

pdf(
  file.path("plots/featureplots", paste0("Hif1aViolCluster.pdf")),
    height = 7, width = 10)
grid.arrange(grobs = plots, nrow = 2, ncol = 2, layout_matrix = rbind(c(1,2),c(3,3)))
dev.off()


```

### Plot UMAP of RNAseq data by other variables

```{r plotRnaUmapColorBySampleName, dependson="normalizeClusterProjectRnaData", fig.width=9.2, fig.height=7}
plot.tmp <-
ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
          color = treatment)) +
  ggrastr::rasterise(
    geom_point(size = .5, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual("Sample", values = pal.treatmentName) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))
# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapColorByTreatment.", filenameSuffix),
  format = plotOutputFormat,
  w = 9.2, h = 7)

rm_tmp(ask=FALSE)
```

```{r plotRnaUmapFacetBySampleName, dependson="normalizeClusterProjectRnaData", fig.width=15, fig.height=7}
plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
          color = treatment)) +
  ggrastr::rasterise(
    geom_point(size = .25, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(values = pal.treatmentName) +
  facet_wrap(~treatment, nrow = 2) +
  guides(color = "none")

# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapFacetByTreatment.", filenameSuffix),
  format = plotOutputFormat,
  w = 15, h = 7)

rm_tmp(ask=FALSE)
```

```{r plotRnaUmapByIghConstantGenes, dependson="normalizeClusterProjectRnaData", fig.width=11.5, fig.height=10}
plot.tmp <-
  FeaturePlot(
    combinedPools,
    reduction = "umap_RNA",
    slot = "data",
    features = 
      c("ICOS", "GITR", "PD-1", "CD25", "CD44", "CD62L"), # Ighe is all 0s, Ighg2a is missing
        # "Cd3d", "Cd3e"), # used these in earlier plots before excluding T cells
    pt.size = 0.5,
    cols = cols.geneCounts) +
  patchwork::plot_layout(ncol = 3)

# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapByADT.", filenameSuffix),
  format = plotOutputFormat,
  w = 12, h = 7)


rm_tmp(ask=FALSE)
```

```{r plotRnaUmapColorByIsotype, dependson="normalizeClusterProjectRnaData", fig.width=9.2, fig.height=7}
plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
          color = isotypeFromTcrBChain)) +
  ggrastr::rasterise(
    geom_point(size = .5, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual("TCR B chain", values = pal.isotype) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapColorByBchain.", filenameSuffix),
  format = plotOutputFormat,
  w = 9.6, h = 7)

rm_tmp(ask=FALSE)
```

```{r plotRnaUmapColorByNFeatureRna, dependson="normalizeClusterProjectRnaData", fig.width=9.2, fig.height=7}
plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
          color = nFeature_RNA)) +
  ggrastr::rasterise(
    geom_point(size = .5, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_viridis_c("# genes\ndetected") +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

# print(plot.tmp)

as.pdf(
  plot.tmp,
  file = file.path("plots", paste0("plotRnaUmapColorByNFeatureRna.", filenameSuffix, ".pdf")),
  width = 5, height = 5)

rm_tmp(ask=FALSE)
```

```{r plotRnaUmapColorByNCountRna, dependson="normalizeClusterProjectRnaData", fig.width=9.2, fig.height=7}
plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
          color = nCount_RNA)) +
  ggrastr::rasterise(
    geom_point(size = .5, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_viridis_c("total UMIs") +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

# print(plot.tmp)

as.pdf(
  plot.tmp,
  file = file.path("plots", paste0("plotRnaUmapColorByNCountRna.", filenameSuffix, ".pdf")),
  width = 5, height = 5)

# outputPlots(
#   plot.tmp, fileDir = "plots",
#   fileStem = 
#     paste0("plotRnaUmapColorByNCountRna.", filenameSuffix),
#   format = plotOutputFormat,
#   w = 9.2, h = 7)

rm_tmp(ask=FALSE)
```

### Plot UMAP of RNAseq data by Tox and other genes

```{r plotRnaUmapByToxAndRelatedGenes, dependson="normalizeClusterProjectRnaData", fig.width=8, fig.height=7}
plot.tmp <-
  FeaturePlot(
    data10x,
    reduction = "umap_RNA",
    slot = "data",
    features = 
      c("Tox", "Tox2", # Tox genes
        "Prdm1", # Blimp-1, indicates plasma cells
        "Aicda"),
    pt.size = 0.5,
    cols = cols.geneCounts) +
  patchwork::plot_layout(ncol = 2)

# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapByToxAndRelatedGenes.", filenameSuffix),
  format = plotOutputFormat,
  w = 8, h = 7)

rm_tmp(ask=FALSE)
```

### Plot t-SNE of RNAseq data by cluster

```{r plotRnaTsneColorByCluster, dependson="normalizeClusterProjectRnaData", fig.width=9.2, fig.height=7}
plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$tsne_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$tsne_RNA@cell.embeddings[,2])) +
  ggrastr::rasterise(
    mapping = aes_string(color = combinedPools@meta.data[[clusterName[["RNA"]]]]),
    geom_point(size = 2, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "Tsne 1", y = "Tsne 2") +
  scale_color_manual(
    "cluster", 
    values = big_colorblind_pal(n_distinct(combinedPools@meta.data[[clusterName[["RNA"]]]]))) +
 # guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaTsneColorByCluster.", filenameSuffix),
  format = plotOutputFormat,
  w = 9, h = 7)

rm_tmp(ask=FALSE)
```


## Heatmap of top genes defining each cluster

The cells within a given cluster are not completely homogeneous. In order to get at the variation among cells within a cluster, we plot the expression of the cluster marker genes in each cell assigned to that cluster.

```{r heatmapSetup1.clusterMarkersRnaData, dependson=c("findClusterMarkersRnaData")}
heatmapParams.clusterMarkersRnaData <- list()
heatmapParams.clusterMarkersRnaData[["nGenesToPlotPerCluster"]] <- 5

heatmapParams.clusterMarkersRnaData[["genesToPlot"]] <-
  clusterMarkers[[clusterName[["RNA"]]]] %>%
  dplyr::arrange(cluster, p_val) %>%
  group_by(cluster) %>%
  dplyr::slice(1:heatmapParams.clusterMarkersRnaData[["nGenesToPlotPerCluster"]]) %>%
  dplyr::pull(gene) %>%
  unique()
```

The heatmaps below show the relative expression of the top `r heatmapParams.clusterMarkersRnaData[["nGenesToPlotPerCluster"]]` genes defining each cluster.

```{r heatmapSetup2.clusterMarkersRnaData,  dependson="heatmapSetup1.clusterMarkersRnaData"}
## check for problem cells that throw off the gene expression scale; more relevant for 10X and similar

## sum the range01-scaled residual counts
# sort(
#   colSums(
#     t(apply(
#       residuals_cdr_prop_mtDNA_reads.select_genes_for_cluster_definitions[
#         genes_to_plot.select_genes_for_cluster_definitions, ], 1, range01))), decreasing=TRUE) %>%
#   head(50)
# no cells with super high expression here

heatmapParams.clusterMarkersRnaData[["cellsToExclude"]] <-
  c()

# set up color scheme
heatmapParams.clusterMarkersRnaData[["annotationCols"]] <- list()

heatmapParams.clusterMarkersRnaData[["annotationCols"]][["cluster"]] <-
  big_colorblind_pal(n_distinct(clusterMarkers[[clusterName[["RNA"]]]]$cluster)) %>%
  setNames(as.character(sort(unique(clusterMarkers[[clusterName[["RNA"]]]]$cluster))))

heatmapParams.clusterMarkersRnaData[["annotationCols"]][["sampleName"]] <-
  pal.sampleName

heatmapParams.clusterMarkersRnaData[[ "annotationCols"]][["treatment"]] <-
  pal.treatmentName


# Determine order of libraries (for use with counts and heatmap annotation)
heatmapParams.clusterMarkersRnaData[["cellBarcodeOrder"]] <-
  combinedPools@meta.data %>%
  dplyr::filter(
    seurat_cellBarcode %nin% heatmapParams.clusterMarkersRnaData[["cellsToExclude"]]) %>%
  dplyr::arrange(!!rlang::sym(clusterName[["RNA"]]), sampleName, treatment, seurat_cellBarcode) %>%
  pull(seurat_cellBarcode)

# Make a column annotation object
heatmapParams.clusterMarkersRnaData[["col_anno"]] <-
  HeatmapAnnotation(
    df = 
      combinedPools@meta.data[
        match(
          heatmapParams.clusterMarkersRnaData[["cellBarcodeOrder"]],
          combinedPools@meta.data$seurat_cellBarcode),] %>%
      dplyr::select(
        cluster = !!rlang::sym(clusterName[["RNA"]]),
        sampleName,
        treatment = treatment),
    col = list(
      cluster = 
        heatmapParams.clusterMarkersRnaData[["annotationCols"]][["cluster"]],
      sampleName =
        heatmapParams.clusterMarkersRnaData[["annotationCols"]][["sampleName"]],
      treatment =
        heatmapParams.clusterMarkersRnaData[["annotationCols"]][["treatment"]]
     ),
    show_legend=c(TRUE, TRUE, TRUE))



# Set up heatmap counts matrix
heatmapParams.clusterMarkersRnaData[["count_matrix"]] <-
  combinedPools@assays$RNA@data[
    heatmapParams.clusterMarkersRnaData[["genesToPlot"]],
    match(
      heatmapParams.clusterMarkersRnaData[["cellBarcodeOrder"]],
      colnames(combinedPools@assays$RNA@data))]

heatmapParams.clusterMarkersRnaData[["count_matrix.range01"]] <-
  heatmapParams.clusterMarkersRnaData[["count_matrix"]] %>%
  apply(MARGIN=1, range01) %>%
  t()

# set gene name display options
heatmapParams.clusterMarkersRnaData[["max_genes_for_names"]] <- 34
heatmapParams.clusterMarkersRnaData[["gene_name_font_size"]] <- 7
```

### Heatmaps of select cluster marker genes, cluster by gene expression

For the following heatmap, we cluster the cells based on similarity of expression of these cluster marker genes.

```{r heatmap.clusterMarkersRnaData, dependson=c("heatmapSetup2.clusterMarkersRnaData", "setHeatmapColors"), fig.width=11, fig.height=8}
heatmap.clusterMarkersRnaData <-
  Heatmap(
    heatmapParams.clusterMarkersRnaData[["count_matrix.range01"]],
    name = "Expression",
    col = heatmapColors,
    cluster_columns = TRUE,
    show_row_names =
      length(heatmapParams.clusterMarkersRnaData[["genesToPlot"]]) <=
      heatmapParams.clusterMarkersRnaData[["max_genes_for_names"]],
    row_names_gp =
      gpar(fontsize = heatmapParams.clusterMarkersRnaData[["gene_name_font_size"]]),
    show_column_names = FALSE, 
    top_annotation =
      heatmapParams.clusterMarkersRnaData[["col_anno"]])

print(heatmap.clusterMarkersRnaData)

# Save heatmap plot
pdf(
  file.path(
    "plots",
    paste0(
      "heatmap.clusterMarkersRnaData",
      ".", filenameSuffix, ".pdf")),
  w=11, h=6,
  useDingbats=F)
print(heatmap.clusterMarkersRnaData)
invisible(dev.off())
```

```{r other heatmap method}
clusterMarkersHmap <- clusterMarkers$seurat_clusters_RNA_0p3 %>%
  #dplyr::group_by(sample) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 30) %>%
  dplyr::slice_max(avg_log2FC, n = 30) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

features <- heatmapParams.clusterMarkersRnaData[["genesToPlot"]]



clusterAverages <- AverageExpression(combinedPools, 
                                    features = features,
                                    assays = 'RNA',
                                    group.by = "seurat_clusters",
                                    use.scale = TRUE,
                                    return.seurat = T) 

clusterAvgExprMat <- clusterAverages[["RNA"]]@data %>% as.matrix()

clusterScaledExprMat <- t(scale(t(clusterAvgExprMat)))


clusterAnno <- rownames(clusterAverages@meta.data)

ColumnAnno <- rowAnnotation(Cluster = clusterAnno, col =  list(Cluster = c("0" = "#000000",
                                                      "1" = "#E69F00",
                                                      "2" = "#56B4E9",
                                                      "3" = "#009E73",
                                                      "4" = "#0072B2",
                                                      "5" = "#D55E00",
                                                      "6" = "#CC79A7")),
                              simple_anno_size = unit(0.25, "cm"))


# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path("plots", 
              "HeatmapClusterMarkers.pdf"),
    height = 4,
    width = 4)

Heatmap(t(clusterScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = FALSE,
        show_row_dend = TRUE,
        left_annotation = ColumnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))
invisible(dev.off())
```
### Heatmaps of select cluster marker genes, ordered by cluster, isotype, and sample/mouse

For the following heatmaps, we arrange the cells by cluster. Within each cluster, we arrange the cells according to isotype and sample/mouse.

```{r heatmap.clusterMarkersRnaData.orderByClusterIsotypeSampleName, dependson=c("heatmapSetup2.clusterMarkersRnaData", "setHeatmapColors"), fig.width=11, fig.height=8}
heatmap.clusterMarkersRnaData.orderByClusterIsotypeSampleName <-
  Heatmap(
    heatmapParams.clusterMarkersRnaData[["count_matrix.range01"]],
    name = "Expression",
    col = heatmapColors,
    cluster_columns = FALSE,
    show_row_names =
      length(heatmapParams.clusterMarkersRnaData[["genesToPlot"]]) <=
      heatmapParams.clusterMarkersRnaData[["max_genes_for_names"]],
    row_names_gp =
      gpar(fontsize = heatmapParams.clusterMarkersRnaData[["gene_name_font_size"]]),
    show_column_names = FALSE, 
    top_annotation =
      heatmapParams.clusterMarkersRnaData[["col_anno"]])

print(heatmap.clusterMarkersRnaData.orderByClusterIsotypeSampleName)

# Save heatmap plot
pdf(
  file.path(
    "plots",
    paste0(
      "heatmap.clusterMarkersRnaData.orderByClusterIsotypeSampleName",
      ".", filenameSuffix, ".pdf")),
  w=11, h=6,
  useDingbats=F)
print(heatmap.clusterMarkersRnaData.orderByClusterIsotypeSampleName)
invisible(dev.off())
```


## Plot co-expression of Tox and Tox2 by cluster

```{r plotCoexpressionToxTox2FacetByCluster, dependson="normalizeClusterProjectRnaData", fig.width=8, fig.height=7}
plot.tmp <-
  ggplot(
    data10x@meta.data %>%
      left_join(
        data10x@assays$RNA@data[c("Tox", "Tox2"),] %>%
          as.matrix() %>%
          t() %>%
          as.data.frame() %>%
          tibble::rownames_to_column(var = "cellBarcode")),
    mapping = aes(x = Tox, y = Tox2)) +
  geom_jitter(
    size = 0.5, alpha = 0.7,
    width = 0.02, height = 0.02) +
  facet_wrap(vars(!!rlang::sym(clusterName[["RNA"]])), nrow = 2) +
  labs(x = "Tox expression (log-normalized)", y = "Tox2 expression (log-normalized)",
       title = "Coexpression of Tox and Tox2 by cluster")


# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotCoexpressionToxTox2FacetByCluster.", filenameSuffix),
  format = plotOutputFormat,
  w = 11, h = 7)

rm_tmp(ask=FALSE)
```

```{r seurat cell cycle stuff}
s.genes <- str_to_title(cc.genes$s.genes)
g2m.genes <- str_to_title(cc.genes$g2m.genes)


combinedPools <- CellCycleScoring(combinedPools, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
          color = Phase)) +
  ggrastr::rasterise(
    geom_point(size = .5, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual("Cell Cycle Phase", values = pal.cycle) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapCellCycle.", filenameSuffix),
  format = plotOutputFormat,
  w = 8, h = 7)

plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
          color = G2M.Score)) +
  ggrastr::rasterise(
    geom_point(size = .5, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2")  +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapG2MScore.", filenameSuffix),
  format = plotOutputFormat,
  w = 8, h = 7)

plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
          color = S.Score)) +
  ggrastr::rasterise(
    geom_point(size = .5, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2")  +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapSScore.", filenameSuffix),
  format = plotOutputFormat,
  w = 8, h = 7)
```

```{r}
avgExpCluster <- AverageExpression(combinedPools, group.by = "seurat_clusters")
avgExpTreatment <- AverageExpression(combinedPools, group.by = "treatment")

```
# Cell cycle scoring

```{r setupPaletteCellCyclePhase}
pal.cellCyclePhase <-
  big_colorblind_pal(3, drop_black = TRUE) %>%
  setNames(c("G1", "S", "G2M"))
```

## Cell cycle scoring using scran cyclone

```{r estimateCellCycleScranCyclone, dependson="normalizeClusterProjectRnaData"}
## use scran::cyclone to estimate cell cycle phase

# cyclone has a pre-defined set of gene pairs for cell cycle phase assignment
mm.pairs.tmp <- readRDS(system.file("exdata", "mouse_cycle_markers.rds", package="scran"))
# because the pre-defined pairs use Ensembl identifiers, need to convert back to those here
# using normalized, not scaled data here; need to check algorithm for suitability
counts.tmp <-
  data10x@assays$RNA@data
rownames(counts.tmp) <-
  annotables::grcm38$ensgene[
    match(rownames(counts.tmp), annotables::grcm38$symbol)]

cellCycleScranCyclone.tmp <-
  scran::cyclone(as.matrix(counts.tmp), pairs = mm.pairs.tmp)
cellCycleScranCyclone.tmp <- 
  data.frame(
    phase.scranCyclone = cellCycleScranCyclone.tmp$phases,
    score.G1.scranCyclone = cellCycleScranCyclone.tmp$scores$G1,
    score.S.scranCyclone = cellCycleScranCyclone.tmp$scores$S,
    score.G2M.scranCyclone = cellCycleScranCyclone.tmp$scores$G2M,
    normalized.score.G1.scranCyclone = cellCycleScranCyclone.tmp$normalized.scores$G1,
    normalized.score.S.scranCyclone = cellCycleScranCyclone.tmp$normalized.scores$S,
    normalized.score.G2M.scranCyclone = cellCycleScranCyclone.tmp$normalized.scores$G2M,
    row.names = colnames(counts.tmp))

data10x <- data10x %>%
  AddMetaData(cellCycleScranCyclone.tmp)

rm_tmp(ask=FALSE)
```

```{r plotRnaUmapColorByCellCyclePhaseScranCyclone, dependson=c("estimateCellCycleScranCyclone", "setupPaletteCellCyclePhase"), fig.width=9.5, fig.height=7}
plot.tmp <-
  ggplot(
    data = data10x@meta.data,
    mapping =
      aes(x = data10x@reductions$umap_RNA@cell.embeddings[,1],
          y = data10x@reductions$umap_RNA@cell.embeddings[,2],
          color = phase.scranCyclone)) +
  geom_point(size = 2, alpha = 0.8) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual("Cell cycle\nphase from\nscran cyclone", values = pal.cellCyclePhase) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapColorByCellCyclePhaseScranCyclone.", filenameSuffix),
  format = plotOutputFormat,
  w = 9.5, h = 7)

rm_tmp(ask=FALSE)
```

## Cell cycle scoring using Seurat

```{r estimateCellCycleSeurat, dependson="normalizeClusterProjectRnaData"}
filenameSeuratMouseGenesConvertedCcGenesUpdated2019 <-
  file.path(dirBoxBase, "Tools", "Gene_sets", "Seurat_cell_cycle_genes_mouse.RDS")

# generate mouse cell cycle phase gene lists if needed (cached because biomaRt is buggy)
if (file.exists(filenameSeuratMouseGenesConvertedCcGenesUpdated2019)) {
  mGenesConvertedCcGenesUpdated2019 <-
    readRDS(filenameSeuratMouseGenesConvertedCcGenesUpdated2019)
} else {
  # Basic function to convert human to mouse gene names 
  # from https://www.r-bloggers.com/2016/10/converting-mouse-to-human-gene-names-with-biomart-package/
  convertHumanGeneList <- function(x){
    require("biomaRt")
    human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
    mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
    genesV2 = 
      getLDS(attributes = c("hgnc_symbol"), 
             filters = "hgnc_symbol", 
             values = x,
             mart = human,
             attributesL = c("mgi_symbol"),
             martL = mouse,
             uniqueRows=T)
    humanx <- unique(genesV2[, 2])
    # Print the first 6 genes found to the screen
    # print(head(humanx))
    return(humanx)
  }
  
  # convert human cell cycle gene lists to mouse
  mGenesConvertedCcGenesUpdated2019 <-
    list(
      s.genes = convertHumanGeneList(cc.genes.updated.2019$s.genes),
      g2m.genes = convertHumanGeneList(cc.genes.updated.2019$g2m.genes))
  
  saveRDS(mGenesConvertedCcGenesUpdated2019,
          file = filenameSeuratMouseGenesConvertedCcGenesUpdated2019)
}

## use Seurat::cellCycleScoring to estimate cell cycle phase
data10x <-
  data10x %>%
  CellCycleScoring(
    s.features = mGenesConvertedCcGenesUpdated2019$s.genes,
    g2m.features = mGenesConvertedCcGenesUpdated2019$g2m.genes)
```

```{r plotRnaUmapColorByCellCyclePhaseSeurat, dependson=c("estimateCellCycleSeurat", "setupPaletteCellCyclePhase"), fig.width=9.3, fig.height=7}
plot.tmp <-
  ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = data10x@reductions$umap_RNA@cell.embeddings[,1],
          y = data10x@reductions$umap_RNA@cell.embeddings[,2],
          color = Phase)) +
  geom_point(size = 2, alpha = 0.8) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual("Cell cycle\nphase from\nSeurat", values = pal.cellCyclePhase) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

# print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = 
    paste0("plotRnaUmapColorByCellCyclePhaseSeurat", filenameSuffix),
  format = plotOutputFormat,
  w = 9.3, h = 7)

rm_tmp(ask=FALSE)
```

```{r plotScoresCellCyclePhaseSeurat, dependson=c("estimateCellCycleSeurat", "setupPaletteCellCyclePhase")}
ggplot(
  data10x@meta.data,
  mapping = aes(x = S.Score, y = G2M.Score, color = Phase)
)
```


## Plot cell cycle scores from scran cyclone and Seurat, colored by inferred phased from either one
## Normalize, cluster, project, and plot RNAseq data, with cell cycle effects regressed out

*Continue here, 2022-01-18, copying and modifying code from above to run with cell cycle regressed out*

# BCR diversity and lineage inference

*Continue here, 2022-01-18*

## BCR lineage inference using Victora Lab's clonality package

```{r inferClonalityVictora, dependson="estimateCellCycleScranCyclone"}
# wasn't working because I had the CRAN version of the package safejoin, which does not have the needed function (safejoin::safe_full_join); need to install from github.com/moodymudskipper/safejoin
clonalityVictora <-
  clonality::clonality(
    data = data10xVdj,
    id_col = "cellBarcode",
    vgene_col = "v_gene",
    jgene_col = "j_gene",
    cdr3_col = "cdr3",
    cell = "B",
    search_gname = FALSE)

# I don't understand the output of this, the clonality values only include a heavy chain or a light chain, not both, so they don't seem like "clones"
# presumably it's inferring "clones" within heavy chain and light chain lineages
```   

## BCR lineage inference using BrepPhylo

```{r inferCloneLineageBrepPhylo, dependson="estimateCellCycleScranCyclone", eval=FALSE}
BrepPhylo::cloneLineage(
  data10xVdj,
  minCloneSize = 1,
  sequenceColumn = 
)
```

# Save data10x with all additions, clustering, projection of all data types

```{r saveData10xAll, dependson="normalizeClusterProjectRnaData"}
# save version of data10x for easy recovery
filenameData10xAll <-
  file.path("data_saved",
            paste0("aggrData10xAll.", filenameSuffix, ".RDS"))
saveRDS(data10x, filenameData10xAll)
# data10x <- readRDS(filenameData10xAll)
```

# Output files for Loupe

```{r outputLoupeFilesData10xAll, dependson="normalizeClusterProjectRnaData"}
# funcutils::seurat2cloupe # this function is a basic export of data from a Seurat object, but it's not very good

# write out coordinates
filenameCoordsForLoupe.tmp <-
  file.path("data_output", paste0("umapCoordsForLoupe.", filenameSuffix, ".csv"))
umapCoordsForLoupe.tmp <-
  data10x@reductions$umap_RNA@cell.embeddings %>%
  # t() %>%
  as.data.frame() %>%
  magrittr::set_colnames(c("UMAP-1", "UMAP-2")) %>%
  tibble::rownames_to_column("Barcode")
write.csv(
  umapCoordsForLoupe.tmp, file = filenameCoordsForLoupe.tmp,
  row.names = F, quote = F)

# write out metadata
filenameMetadataForLoupe.tmp <-
  file.path("data_output", paste0("metadataForLoupe.", filenameSuffix, ".csv"))
metadataForLoupe.tmp <-
  data10x@meta.data %>%
  dplyr::select(
    -matches("^clrMdHTO"), -matches("^logHTO"), -matches("^logCpmHTO"), -matches("^clrSeuratHTO"),
    -(HTO_maxID:hash.ID)) %>%
  dplyr::select_if(function(x) {!is.numeric(x)}) %>% # remove numeric columns
  dplyr::select(-enclone_raw_clonotype_id) %>% # too many unique values
  dplyr::select(Barcode = cellBarcode, everything())
write.csv(
  metadataForLoupe.tmp, file = filenameMetadataForLoupe.tmp,
  row.names = F, quote = F)

rm_tmp(ask=FALSE)
```

# Plots of clonotype distribution by sample

*limit to only those with ? cells, and drop the rest or make them a common color (gray?)*

# Save data objects excluding large Seurat objects

The 10x data to be loaded immediately after this step include very large data matrices. It's easiest to save the data loaded up to this point, so that they can be easily loaded for later use.

```{r saveProjectDataEx10xEnd, cache=FALSE}
filenameDataEx10xEnd <-
  file.path("data_saved", paste0("dataEx10xEnd.", filenameSuffix, ".RData"))

save(list = ls_grep(pattern = "^data10x", invert=TRUE), file = filenameDataEx10xEnd)
# load(file = filenameDataEx10xEnd)
```


# Output R session information

```{r output_session_info, cache=FALSE}
miscHelpers::print_session_info()
```

```{r featureplots of ADT}

plot.tmp <- FeaturePlot(combinedPools, features = "PD-1", split.by = "treatment")
  plot.tmp +
  ggrastr::rasterise(
    geom_point(size = .25, alpha = 0.8),
    dpi = rasterResolutionDpi) +
      labs(x = "UMAP 1", y = "UMAP 2") 
  #facet_wrap(~treatment, nrow = 2) 
    ```


```{r pseudobulking by cluster 2}

Clust2PBS <- subset(combinedPools, idents = "PBS_2")
Clust2WT <- subset(combinedPools, idents = "Fc.WT_2")
Clust2Mut24 <- subset(combinedPools, idents = "Fc.Mut24_2")

Clust2PBS@meta.data$Pool <- str_extract(Clust2PBS@meta.data$cellBarcode, "\\d$")

Clust2WT@meta.data$Pool <- str_extract(Clust2WT@meta.data$cellBarcode, "\\d$")

Clust2Mut24@meta.data$Pool <- str_extract(Clust2Mut24@meta.data$cellBarcode, "\\d$")

for_psuedobulkPBS <- list()
for(i in levels(Clust2PBS)){
  temp <- subset(Clust2PBS, idents = i)
  name <- i
  temp <- temp@assays$RNA@counts
  temp <- Matrix::rowSums(temp)
  temp <- as.data.frame(temp)
  colnames(temp) <- name
  for_psuedobulkPBS[[i]] <- as.matrix(temp)
}

pseudobulkPBS <- for_psuedobulkPBS$`1`
pseudobulkPBS <- merge(pseudobulkPBS, for_psuedobulkPBS$`2`, by= 0, all=TRUE )
pseudobulkPBS <- merge(pseudobulkPBS, for_psuedobulkPBS$`3`, by.x = 'Row.names', by.y=0, all = TRUE)
pseudobulkPBS <- merge(pseudobulkPBS, for_psuedobulkPBS$`4`, by.x = 'Row.names', by.y =0, all= TRUE)
rownames(pseudobulkPBS) <- pseudobulkPBS$Row.names
pseudobulkPBS <- pseudobulkPBS[,2:5]
colnames(pseudobulkPBS) <- c("PBS_1", "PBS_2", "PBS_3", "PBS_4")

for_psuedobulkPBS <- list()
for(i in levels(Clust2PBS)){
  temp <- subset(Clust2PBS, idents = i)
  name <- i
  temp <- temp@assays$RNA@counts
  temp <- Matrix::rowSums(temp)
  temp <- as.data.frame(temp)
  colnames(temp) <- name
  for_psuedobulkPBS[[i]] <- as.matrix(temp)
}

pseudobulkPBS <- for_psuedobulkPBS$`1`
pseudobulkPBS <- merge(pseudobulkPBS, for_psuedobulkPBS$`2`, by= 0, all=TRUE )
pseudobulkPBS <- merge(pseudobulkPBS, for_psuedobulkPBS$`3`, by.x = 'Row.names', by.y=0, all = TRUE)
pseudobulkPBS <- merge(pseudobulkPBS, for_psuedobulkPBS$`4`, by.x = 'Row.names', by.y =0, all= TRUE)
rownames(pseudobulkPBS) <- pseudobulkPBS$Row.names
pseudobulkPBS <- pseudobulkPBS[,2:5]
colnames(pseudobulkPBS) <- c("PBS_1", "PBS_2", "PBS_3", "PBS_4")

for_psuedobulkPBS <- list()
for(i in levels(Clust2PBS)){
  temp <- subset(Clust2PBS, idents = i)
  name <- i
  temp <- temp@assays$RNA@counts
  temp <- Matrix::rowSums(temp)
  temp <- as.data.frame(temp)
  colnames(temp) <- name
  for_psuedobulkPBS[[i]] <- as.matrix(temp)
}

pseudobulkPBS <- for_psuedobulkPBS$`1`
pseudobulkPBS <- merge(pseudobulkPBS, for_psuedobulkPBS$`2`, by= 0, all=TRUE )
pseudobulkPBS <- merge(pseudobulkPBS, for_psuedobulkPBS$`3`, by.x = 'Row.names', by.y=0, all = TRUE)
pseudobulkPBS <- merge(pseudobulkPBS, for_psuedobulkPBS$`4`, by.x = 'Row.names', by.y =0, all= TRUE)
rownames(pseudobulkPBS) <- pseudobulkPBS$Row.names
pseudobulkPBS <- pseudobulkPBS[,2:5]
colnames(pseudobulkPBS) <- c("PBS_1", "PBS_2", "PBS_3", "PBS_4")

Idents(Clust2WT) <- "Pool"
for_psuedobulkWT <- list()
for(i in levels(Clust2WT)){
  temp <- subset(Clust2WT, idents = i)
  name <- i
  temp <- temp@assays$RNA@counts
  temp <- Matrix::rowSums(temp)
  temp <- as.data.frame(temp)
  colnames(temp) <- name
  for_psuedobulkWT[[i]] <- as.matrix(temp)
}

pseudobulkWT <- for_psuedobulkWT$`1`
pseudobulkWT <- merge(pseudobulkWT, for_psuedobulkWT$`2`, by= 0, all=TRUE )
pseudobulkWT <- merge(pseudobulkWT, for_psuedobulkWT$`3`, by.x = 'Row.names', by.y=0, all = TRUE)
pseudobulkWT <- merge(pseudobulkWT, for_psuedobulkWT$`4`, by.x = 'Row.names', by.y =0, all= TRUE)
rownames(pseudobulkWT) <- pseudobulkWT$Row.names
pseudobulkWT <- pseudobulkWT[,2:5]
colnames(pseudobulkWT) <- c("WT_1", "WT_2", "WT_3", "WT_4")

Idents(Clust2Mut24) <- "Pool"
for_psuedobulkMut24 <- list()
for(i in levels(Clust2Mut24)){
  temp <- subset(Clust2Mut24, idents = i)
  name <- i
  temp <- temp@assays$RNA@counts
  temp <- Matrix::rowSums(temp)
  temp <- as.data.frame(temp)
  colnames(temp) <- name
  for_psuedobulkMut24[[i]] <- as.matrix(temp)
}

pseudobulkMut24 <- for_psuedobulkMut24$`1`
pseudobulkMut24 <- merge(pseudobulkMut24, for_psuedobulkMut24$`2`, by= 0, all=TRUE )
pseudobulkMut24 <- merge(pseudobulkMut24, for_psuedobulkMut24$`3`, by.x = 'Row.names', by.y=0, all = TRUE)
pseudobulkMut24 <- merge(pseudobulkMut24, for_psuedobulkMut24$`4`, by.x = 'Row.names', by.y =0, all= TRUE)
rownames(pseudobulkMut24) <- pseudobulkMut24$Row.names
pseudobulkMut24 <- pseudobulkMut24[,2:5]
colnames(pseudobulkMut24) <- c("Mut24_1", "Mut24_2", "Mut24_3", "Mut24_4")

Clust2PseudobulkFinal <- merge(pseudobulkMut24, pseudobulkPBS, by=0, all=TRUE)
Clust2PseudobulkFinal <- merge(Clust2PseudobulkFinal, pseudobulkWT, by.x= "Row.names", by.y=0, all=TRUE)
rownames(Clust2PseudobulkFinal) <- Clust2PseudobulkFinal$Row.names
Clust2PseudobulkFinal <- Clust2PseudobulkFinal[,2:ncol(Clust2PseudobulkFinal)]

group <- factor(c(1,1,1,1,2,2,2,2,3,3,3,3))
y <- DGEList(Clust2PseudobulkFinal, group = group)
design <- model.matrix(~0+group, data=y$samples)
keep <- filterByExpr(y)
y <- y[keep,,keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
y <- estimateDisp(y,design)
colnames(design) <- c("Mut24", "PBS", "WT")
contr.matrix <- makeContrasts(Mut24vsPBS = Mut24-PBS,
                              Mut24vsWT = Mut24-WT,
                              WTvsPBS = WT-PBS,
                              levels=designTreat)

v<- voom(y, design, plot=TRUE)

fit <- lmFit(v, design)
fit_c <- contrasts.fit(fit, contr.matrix)
fit_c_eb <- eBayes(fit_c)
topMutvsWT <- topTable(fit_c_eb, coef = "Mut24vsWT", number = 20000)
topWTvsPBS <- topTable(fit_c_eb, coef = "WTvsPBS", number = 20000)
topMutvsPBS <- topTable(fit_c_eb, coef = "Mut24vsPBS", number = 20000)
markersMutvsWT <- topMutvsWT[c("Ly6c1","Batf"),]
markersWTvsPBS <- topWTvsPBS[c("Ly6c1","Batf"),]
markersMutvsPBS <- topMutvsPBS[c("Ly6c1","Batf"),]

MutvsWT2 <- ggplot(data = topMutvsWT, aes(x=logFC, y=-log10(adj.P.Val), color = logFC>0)) +
    geom_point(alpha=0.7, size=1, shape = 19) +
  geom_text_repel(data=head(markersMutvsWT, 20), label = rownames(head(markersMutvsWT,20)), max.overlaps = Inf, min.segment.length = unit(0, 'lines'), size = 3, col = "black") +
    scale_color_manual(values = c("darkcyan", "red"), name = "", label = NULL) +
        labs(x = "Log Fold-change",
         y = "-log10 adj.P.Val",
         title = "Fc.Mut24 vs Fc.WT") +
    geom_hline(yintercept= -log10(.05), color="black",linetype="dotted",size=1.0) +
    theme(plot.title = element_text(hjust = .5)) +
    NoLegend()

WTvsPBS2 <- ggplot(data = topWTvsPBS, aes(x=logFC, y= -log10(adj.P.Val), color = logFC>0)) +
    geom_point(alpha=0.7, size=1, shape = 19) +
  geom_text_repel(data=head(markersWTvsPBS, 20), label = rownames(head(markersWTvsPBS,20)), max.overlaps = Inf, min.segment.length = unit(0, 'lines'), size = 3, col = "black") +
    scale_color_manual(values = c("darkcyan", "red"), name = "", label = NULL) +
        labs(x = "Log Fold-change",
         y = "-log10 adj.P.Val",
         title = "Fc.WT vs PBS") +
    geom_hline(yintercept= -log10(.05), color="black",linetype="dotted",size=1.0) +
    theme(plot.title = element_text(hjust = .5)) +
    NoLegend()

MutvsPBS2 <- ggplot(data = topMutvsPBS, aes(x=logFC, y= -log10(adj.P.Val), color = logFC>0)) +
    geom_point(alpha=0.7, size=1, shape = 19) +
  geom_text_repel(data=head(markersMutvsPBS, 20), label = rownames(head(markersMutvsPBS,20)), max.overlaps = Inf, min.segment.length = unit(0, 'lines'), size = 3, col = "black") +
    scale_color_manual(values = c("darkcyan", "red"), name = "", label = NULL) +
        labs(x = "Log Fold-change",
         y = "-log10 adj.P.Val",
         title = "Fc.Mut24 vs PBS") +
    geom_hline(yintercept= -log10(.05), color="black",linetype="dotted",size=1.0) +
    theme(plot.title = element_text(hjust = .5)) +
    NoLegend()
```

```{r pseudobulk overall}

for_psuedobulktotal <- list()
for(i in levels(combinedPools)){
  temp <- subset(combinedPools, idents = i)
  name <- i
  temp <- temp@assays$RNA@counts
  temp <- Matrix::rowSums(temp)
  temp <- as.data.frame(temp)
  colnames(temp) <- name
  for_psuedobulktotal[[i]] <- as.matrix(temp)
}

pseudobulktotal <- for_psuedobulktotal[[1]]
for(i in names(for_psuedobulktotal)[2:length(for_psuedobulktotal)]){
  name <- i
 pseudobulktotal <- merge(pseudobulktotal,for_psuedobulktotal[[name]],by=0,all=TRUE)
 rownames(pseudobulktotal) <- pseudobulktotal$Row.names
 pseudobulktotal$Row.names <- NULL
}

group_cluster <- factor(as.numeric(str_extract(colnames(pseudobulktotal), "\\d$")) + 1)
group_treatment <- sub(pattern = ".*PBS*..", "1", colnames(pseudobulktotal)) %>% sub(pattern = ".*Mut24*..", "2", .) %>% sub(pattern = ".*Fc.WT*..", "3", .) %>% factor()
group_clustertreatment  <- sub("Pool._", "", colnames(pseudobulktotal)) %>% sub("PBS_0", "1", .) %>% sub("PBS_1", "2", .) %>% sub("PBS_2", "3", .) %>% sub("PBS_3", "4", .) %>% sub("PBS_4", "5", .) %>% sub("PBS_5", "6", .) %>% sub("PBS_6", "7", .) %>% sub("Fc.WT_0", "8", .) %>% sub("Fc.WT_1", "9", .) %>% sub("Fc.WT_2", "10", .) %>% sub("Fc.WT_3", "11", .) %>% sub("Fc.WT_4", "12", .) %>% sub("Fc.WT_5", "13", .) %>% sub("Fc.WT_6", "14", .) %>% sub("Fc.Mut24_0", "15", .)  %>% sub("Fc.Mut24_1", "16", .)  %>% sub("Fc.Mut24_2", "17", .)  %>% sub("Fc.Mut24_3", "18", .)  %>% sub("Fc.Mut24_4", "19", .)  %>% sub("Fc.Mut24_5", "20", .)  %>% sub("Fc.Mut24_6", "21", .) %>% factor()

yClust <- DGEList(pseudobulktotal, group = group_cluster)
designClust <- model.matrix(~0+group_cluster, data=yClust$samples)
keep <- filterByExpr(yClust)
yClust <- yClust[keep,,keep.lib.sizes=FALSE]
yClust <- calcNormFactors(yClust)
colnames(designClust) <- c("cluster0", "cluster1", "cluster2", "cluster3", "cluster4", "cluster5", "cluster6")
yClust <- estimateDisp(yClust,designClust)

contrastsCluster <- vector()
for(i in 1:(length(colnames(designClust)) - 1)){
  entry <- i
  iterate <- colnames(designClust)[entry+1:length(colnames(designClust))]
  iterate <- iterate[!is.na(iterate)]
    for(n in iterate){
      temp <- paste0(colnames(designClust)[entry], "-", colnames(designClust)[which(colnames(designClust) == n)])
      contrastsCluster <- append(contrastsCluster, temp)
    }
}

contr.matrixClust <- makeContrasts(contrasts = contrastsCluster, levels = designClust)
vClust <- voom(yClust, designClust, plot = FALSE)
eFitClust <- lmFit(vClust, designClust) %>% contrasts.fit(., contr.matrixClust) %>% eBayes(.)

topTableClust <- list()
for(i in contrastsCluster){
  entry <- i
  topTableClust[[entry]] <- topTable(eFitClust, entry, 20)
}


topFive <- data.frame()
for(i in names(topTableClust)){
  name <- i
  toadd <- topTableClust[[name]][1:5,]
  droprows <- rownames(toadd) %in% rownames(topFive)
  topFive <- rbind(topFive, toadd[!droprows,])
}


top0 <- data.frame()
for(i in grep( 'cluster0', names(topTableClust), value = TRUE)){
  name <- i
  toadd <- topTableClust[[name]][1:10,]
  droprows <- rownames(toadd) %in% rownames(top0)
  top0 <- rbind(top0, toadd[!droprows,])
}

clusterDGEList <- lmFit(vClust, designClust)
clusterDGEList0 <- clusterDGEList[c(rownames(top0)),]
Heatmap0 <-coolmap(clusterDGEList0, linkage.col = "none")

top1 <- data.frame()
for(i in grep( 'cluster1', names(topTableClust), value = TRUE)){
  name <- i
  toadd <- topTableClust[[name]][1:10,]
  droprows <- rownames(toadd) %in% rownames(top1)
  top1 <- rbind(top1, toadd[!droprows,])
}

clusterDGEList1 <- clusterDGEList[c(rownames(top1)),]
Heatmap1 <-coolmap(clusterDGEList1, linkage.col = "none")

top2 <- data.frame()
for(i in grep( 'cluster2', names(topTableClust), value = TRUE)){
  name <- i
  toadd <- topTableClust[[name]][1:10,]
  droprows <- rownames(toadd) %in% rownames(top2)
  top2 <- rbind(top2, toadd[!droprows,])
}

clusterDGEList2 <- clusterDGEList[c(rownames(top2)),]
Heatmap2 <-coolmap(clusterDGEList2, linkage.col = "none")

top3 <- data.frame()
for(i in grep( 'cluster3', names(topTableClust), value = TRUE)){
  name <- i
  toadd <- topTableClust[[name]][1:10,]
  droprows <- rownames(toadd) %in% rownames(top3)
  top3 <- rbind(top3, toadd[!droprows,])
}

clusterDGEList3 <- clusterDGEList[c(rownames(top3)),]
Heatmap3 <-coolmap(clusterDGEList3, linkage.col = "none")

top4 <- data.frame()
for(i in grep( 'cluster4', names(topTableClust), value = TRUE)){
  name <- i
  toadd <- topTableClust[[name]][1:10,]
  droprows <- rownames(toadd) %in% rownames(top4)
  top4 <- rbind(top4, toadd[!droprows,])
}

clusterDGEList4 <- clusterDGEList[c(rownames(top4)),]
Heatmap4 <-coolmap(clusterDGEList4, linkage.col = "none",main  = "Cluster 4 DE Genes")

top5 <- data.frame()
for(i in grep( 'cluster5', names(topTableClust), value = TRUE)){
  name <- i
  toadd <- topTableClust[[name]][1:10,]
  droprows <- rownames(toadd) %in% rownames(top5)
  top5 <- rbind(top5, toadd[!droprows,])
}

clusterDGEList5 <- clusterDGEList[c(rownames(top5)),]
Heatmap5 <-coolmap(clusterDGEList5, linkage.col = "none", main  = "Cluster 5 DE Genes")

top6 <- data.frame()
for(i in grep( 'cluster6', names(topTableClust), value = TRUE)){
  name <- i
  toadd <- topTableClust[[name]][1:10,]
  droprows <- rownames(toadd) %in% rownames(top6)
  top6 <- rbind(top6, toadd[!droprows,])
}

clusterDGEList6 <- clusterDGEList[c(rownames(top6)),]
Heatmap6 <-coolmap(clusterDGEList6, linkage.col = "none")

CH0 <- Heatmap0$carpet[1:7, -1]- Heatmap0$carpet[rep(1,7),-1]
Heatmap_0 <- ComplexHeatmap::Heatmap(CH0, name = "Cluster 0 DE", row_order = rownames(CH0), column_names_gp = grid::gpar(fontsize = 7), column_title = "DE Cluster 0 Pairwise", column_title_gp = gpar(fontsize = 20, fontface = "bold"))

CH1 <- Heatmap1$carpet[1:7, -1]- Heatmap1$carpet[rep(2,7),-1]
Heatmap_1 <- ComplexHeatmap::Heatmap(CH1, name = "Cluster 1 DE", row_order = rownames(CH1), column_names_gp = grid::gpar(fontsize = 7), column_title = "DE Cluster 1 Pairwise", column_title_gp = gpar(fontsize = 20, fontface = "bold"))


CH2 <- Heatmap2$carpet[1:7, -1]- Heatmap2$carpet[rep(3,7),-1]
Heatmap_2 <- ComplexHeatmap::Heatmap(CH2, name = "Cluster 2 DE", row_order = rownames(CH2), column_names_gp = grid::gpar(fontsize = 7), column_title = "DE Cluster 2 Pairwise", column_title_gp = gpar(fontsize = 20, fontface = "bold"))


CH3 <- Heatmap3$carpet[1:7, -1]- Heatmap3$carpet[rep(4,7),-1]
Heatmap_3 <- ComplexHeatmap::Heatmap(CH3, name = "Cluster 3 DE", row_order = rownames(CH3), column_names_gp = grid::gpar(fontsize = 7), column_title = "DE Cluster 3 Pairwise", column_title_gp = gpar(fontsize = 20, fontface = "bold"))


CH4 <- Heatmap4$carpet[1:7, -1]- Heatmap4$carpet[rep(5,7),-1]
Heatmap_4 <- ComplexHeatmap::Heatmap(CH4, name = "Cluster 4 DE", row_order = rownames(CH4), column_names_gp = grid::gpar(fontsize = 7), column_title = "DE Cluster 4 Pairwise", column_title_gp = gpar(fontsize = 20, fontface = "bold"))


CH5 <- Heatmap5$carpet[1:7, -1]- Heatmap5$carpet[rep(6,7),-1]
Heatmap_5 <- ComplexHeatmap::Heatmap(CH5, name = "Cluster 5 DE", row_order = rownames(CH5), column_names_gp = grid::gpar(fontsize = 7), column_title = "DE Cluster 5 Pairwise", column_title_gp = gpar(fontsize = 20, fontface = "bold"))


CH6 <- Heatmap6$carpet[1:7, -1]- Heatmap6$carpet[rep(7,7),-1]
Heatmap_6 <- ComplexHeatmap::Heatmap(CH6, name = "Cluster 6 DE", row_order = rownames(CH6), column_names_gp = grid::gpar(fontsize = 7), column_title = "DE Cluster 6 Pairwise", column_title_gp = gpar(fontsize = 20, fontface = "bold"))


pdf("plots/Heatmap_Clust_0.pdf", width = 8, height = 8)
print(Heatmap_0)
dev.off()

pdf("plots/Heatmap_Clust_1.pdf", width = 8, height = 8)
print(Heatmap_1)
dev.off()

pdf("plots/Heatmap_Clust_2.pdf", width = 8, height = 8)
print(Heatmap_2)
dev.off()

pdf("plots/Heatmap_Clust_3.pdf", width = 8, height = 8)
print(Heatmap_3)
dev.off()

pdf("plots/Heatmap_Clust_4.pdf", width = 8, height = 8)
print(Heatmap_4)
dev.off()

pdf("plots/Heatmap_Clust_5.pdf", width = 8, height = 8)
print(Heatmap_5)
dev.off()

pdf("plots/Heatmap_Clust_6.pdf", width = 8, height = 8)
print(Heatmap_6)
dev.off()

write.xlsx(topTableClust["cluster0vscluster1"], file="tables/pseudotimePairwiseClusters.xlsx", sheetName="cluster0vscluster1", row.names=TRUE)
for(i in names(topTableClust)[2:length(names(topTableClust))]){
  name <- i
  
  write.xlsx(topTableClust[name], file="tables/pseudotimePairwiseClusters.xlsx", sheetName=name, append = TRUE, row.names=TRUE)

}
```

```{r GSEA Stat5 vs Mu24}

GenesetMatrix <- read.table("GSE84553_matrix.txt")

STAT5matrix <- GenesetMatrix %>% dplyr::select(Gene_Symbol, 
                                               Entrez_ID, STAT5bCA_vs_Control_Treg_Log2FC, 
                                               STAT5bCA_vs_Control_Treg_Padj)

STAT5matrix<- STAT5matrix %>% 
  filter(STAT5bCA_vs_Control_Treg_Padj < .05) %>% 
  filter(STAT5bCA_vs_Control_Treg_Log2FC > 0) %>% 
  dplyr::select(Entrez_ID)

Stat5Convert <- bitr(STAT5matrix$Entrez_ID, 
                     fromType = "ENTREZID", 
                     toType = "SYMBOL", 
                     OrgDb = org.Mm.eg.db)

Stat5MutvsPBS <- topMutvsPBSTotal[rownames(topMutvsPBSTotal) %in% Stat5Convert$SYMBOL,]

yTreat <- DGEList(pseudobulktotal, group = group_treatment)
designTreat <- model.matrix(~0+group_treatment, data=yTreat$samples)
keep <- filterByExpr(yTreat)
yTreat <- yTreat[keep,,keep.lib.sizes=FALSE]
yTreat <- calcNormFactors(yTreat)
colnames(designTreat) <- c("PBS", "Fc.Mut24", "Fc.WT")
yTreat <- estimateDisp(yTreat,designTreat)


mroast(yTreat, index = Stat5Convert$SYMBOL, designTreat, contrast = contrastTreatment[,2])

v<- voom(yTreat, designTreat, plot=TRUE)

fit <- lmFit(v, designTreat)
fit_c <- contrasts.fit(fit, contrastTreatment)
fit_c_eb <- eBayes(fit_c)

Stat5MutvsPBS

NoLabelStat5 <- ggplot(data = topMutvsPBSTotal, aes(x=logFC, y=-log10(adj.P.Val), color = logFC>0)) +
    geom_point(alpha=0.7, size=1, shape = 19) +
    geom_point(data = Stat5MutvsPBS, aes(x=logFC, y= -log10(adj.P.Val)), color = "darkgreen", size=1, shape = 19, alpha = .5) +
  
      scale_color_manual(values = c("darkcyan", "red"), name = "", label = NULL) +
        labs(x = "Log Fold-change",
         y = "-log10 ad.p.val",
         title = "Mut.24 vs PBS, Stat5 Genes") +
    theme(plot.title = element_text(hjust = .5)) +
    geom_hline(yintercept= -log10(.05), color="black",linetype="dotted",size=1.0) +
  
    NoLegend()


LabelStat5 <- ggplot(data = topMutvsPBSTotal, aes(x=logFC, y=-log10(adj.P.Val), color = logFC>0)) +
    geom_point(alpha=0.7, size=3, shape = 19) +
    geom_point(data = Stat5MutvsPBS, aes(x=logFC, y= -log10(adj.P.Val)), color = "darkgreen", size=3, shape = 19, alpha = 1) +
  geom_text_repel(data=head(Stat5MutvsPBS[order(Stat5MutvsPBS$adj.P.Val),], 15), label = rownames(head(Stat5MutvsPBS[order(Stat5MutvsPBS$adj.P.Val),], 15)), max.overlaps = Inf, min.segment.length = unit(0, 'lines'), size = 5, col = "black", nudge_y = .7, nudge_x = .7) +
      scale_color_manual(values = c("darkcyan", "red"), name = "", label = NULL) +
        labs(x = "Log Fold-change",
         y = "-log10 adj.p.val",
         title = "Mut.24 vs PBS, Stat5 Genes\n520 Up | 141 Down\np < 2.2e-16") +
    theme(plot.title = element_text(hjust = .5)) +
    geom_hline(yintercept= -log10(.05), color="black",linetype="dotted",size=1.0) +
  
    NoLegend()

pdf("plots/Stat5GenesMutPBS_NoLabel.pdf", height = 8, width = 8)
print(NoLabelStat5)
dev.off()

pdf("plots/Stat5GenesMutPBS_Label.pdf", height = 8, width = 8)
print(LabelStat5)
dev.off()

SigTCR <- read.csv("TCR-upregulated Treg genes.csv")
TCRList <- SigTCR$Gene.symbol %>% gsub(".*/","", .) %>% unique()
TCRMutvsPBS <- topMutvsPBSTotal[rownames(topMutvsPBSTotal) %in% TCRList,]

LabelTCR <- ggplot(data = topMutvsPBSTotal, aes(x=logFC,
                                                y=-log10(adj.P.Val),
                                                color = logFC>0)) +
    geom_point(alpha=0.7,
               size=3,
               shape = 19) +
    geom_point(data = TCRMutvsPBS,
               aes(x=logFC, y= -log10(adj.P.Val)),
               color = "darkgreen", size=3,
               shape = 19, alpha = 1) +
  geom_text_repel(data=head(TCRMutvsPBS[order(TCRMutvsPBS$adj.P.Val),], 15),
                  label = rownames(head(TCRMutvsPBS[order(TCRMutvsPBS$adj.P.Val),], 15)),
                  max.overlaps = Inf, min.segment.length = unit(0, 'lines'),
                  size = 5,
                  col = "black",
                  nudge_y = .6,
                  nudge_x = .6) +
      scale_color_manual(values = c("darkcyan", "red"),
                         name = "",
                         label = NULL) +
        labs(x = "Log Fold-change",
         y = "-log10 adj.p.val",
         title = "Mut.24 vs PBS, TCR Genes\n431 Up | 315 Down\np = 2.316e-04") +
    theme(plot.title = element_text(hjust = .5)) +
    geom_hline(yintercept= -log10(.05),
               color="black",linetype="dotted",size=1.0) +
    NoLegend()

pdf("plots/TCRGenesMutPBS_Label.pdf", height = 8, width = 8)
print(LabelTCR)
dev.off()

STAT5matrixlogFC1 <- GenesetMatrix %>% dplyr::select(Gene_Symbol, 
                                               Entrez_ID, STAT5bCA_vs_Control_Treg_Log2FC, 
                                               STAT5bCA_vs_Control_Treg_Padj)

STAT5matrixlogFC1<- STAT5matrixlogFC1 %>% 
  filter(STAT5bCA_vs_Control_Treg_Padj < .05) %>% 
  filter(STAT5bCA_vs_Control_Treg_Log2FC > 1) %>% 
  dplyr::select(Entrez_ID)

Stat5ConvertlogFC1 <- bitr(STAT5matrixlogFC1$Entrez_ID, 
                     fromType = "ENTREZID", 
                     toType = "SYMBOL", 
                     OrgDb = org.Mm.eg.db)

Stat5fc1MutvsPBS <- topMutvsPBSTotal[rownames(topMutvsPBSTotal) %in% Stat5ConvertlogFC1$SYMBOL,]

LabelStat5fc1 <- ggplot(data = topMutvsPBSTotal, aes(x=logFC, y=-log10(adj.P.Val), color = logFC>0)) +
    geom_point(alpha=0.7, size=3, shape = 19) +
    geom_point(data = Stat5fc1MutvsPBS, aes(x=logFC, y= -log10(adj.P.Val)), color = "darkgreen", size=3, shape = 19, alpha = 1) +
  geom_text_repel(data=head(Stat5fc1MutvsPBS[order(Stat5fc1MutvsPBS$adj.P.Val),], 15), label = rownames(head(Stat5fc1MutvsPBS[order(Stat5fc1MutvsPBS$adj.P.Val),], 15)), max.overlaps = Inf, min.segment.length = unit(0, 'lines'), size = 5, col = "black", nudge_y = .7, nudge_x = .7) +
      scale_color_manual(values = c("darkcyan", "red"), name = "", label = NULL) +
        labs(x = "Log Fold-change",
         y = "-log10 adj.p.val",
         title = "Mut.24 vs PBS, Stat5 logFC > 1 Genes\n63 Up | 9 Down\np = 1.009e-09") +
    theme(plot.title = element_text(hjust = .5)) +
    geom_hline(yintercept= -log10(.05), color="black",linetype="dotted",size=1.0) +
  
    NoLegend()



LabelTCRfc1 <- ggplot(data = topMutvsPBSTotal, aes(x=logFC,
                                                y=-log10(adj.P.Val),
                                                color = logFC>0)) +
    geom_point(alpha=0.7,
               size=3,
               shape = 19) +
    geom_point(data = TCRfc1MutvsPBS,
               aes(x=logFC, y= -log10(adj.P.Val)),
               color = "darkgreen", size=3,
               shape = 19, alpha = 1) +
  geom_text_repel(data=head(TCRfc1MutvsPBS[order(TCRfc1MutvsPBS$adj.P.Val),], 15),
                  label = rownames(head(TCRfc1MutvsPBS[order(TCRfc1MutvsPBS$adj.P.Val),], 15)),
                  max.overlaps = Inf, min.segment.length = unit(0, 'lines'),
                  size = 5,
                  col = "black",
                  nudge_y = .6,
                  nudge_x = .6) +
      scale_color_manual(values = c("darkcyan", "red"),
                         name = "",
                         label = NULL) +
        labs(x = "Log Fold-change",
         y = "-log10 adj.p.val",
         title = "Mut.24 vs PBS, TCR Genes logFC >1\n55 Up | 45 Down\np = .4447") +
    theme(plot.title = element_text(hjust = .5)) +
    geom_hline(yintercept= -log10(.05),
               color="black",linetype="dotted",size=1.0) +
    NoLegend()

pdf("plots/TCRGenesMutPBSfc1_Label.pdf", height = 8, width = 8)
print(LabelTCRfc1)
dev.off()

pdf("plots/Stat5GenesMutPBSfc1_Label.pdf", height = 8, width = 8)
print(LabelStat5fc1)
dev.off()

pdf("plots/boxplotMissingStat5Genes.pdf", width = 8, height =8)
ggplot(boxData, aes(x=group, y=Mean)) +
  geom_boxplot() + labs(y = "Mean Read Count", title = "log FC > 1 Stat5 Genes", x = "Status in Mut24vsPBS")
dev.off()

```

```{r import to monocle3 and perform gene module}

combinedPools[["UMAP"]] <- combinedPools[["umap_RNA"]]


p471monocle <- as.cell_data_set(combinedPools)
p471monocle <- cluster_cells(p471monocle, resolution=1e-3)

gene_modules <- find_gene_modules(p471monocle, resolution=1e-2)

cell_group_df<- tibble::tibble(cell=row.names(colData(p471monocle)), 
               cell_group=p471monocle@colData$treatment)

agg_mat <- aggregate_gene_expression(p471monocle, gene_modules, cell_group_df)

row.names(agg_mat) <- stringr::str_c("Module ", row.names(agg_mat))

pdf(file = "plots/geneModuleHeatmap.pdf", height = 8, width =6)
pheatmap::pheatmap(agg_mat, cluster_rows=TRUE, cluster_cols=TRUE,
                   scale="column", clustering_method="ward.D2",
                   fontsize=6)
dev.off()

write.xlsx(as.data.frame(gene_modules[gene_modules$module == "1",]), file = "tables/geneModules.xlsx", sheetName = "Module 1")

for(i in 2:20){
  name <- i
  write.xlsx(as.data.frame(gene_modules[gene_modules$module == name,]), file="tables/geneModules.xlsx", sheetName=paste("Module", name), append = TRUE, row.names=TRUE)
}

GeneNames <- as.data.frame(gene_modules[gene_modules$module == "5",])$id

GeneNames <- bitr(GeneNames, fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = org.Mm.eg.db)

enrichGO(gene = GeneNames$ENTREZID,
                  OrgDb = org.Mm.eg.db,
                  ont = "BP",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

egoBP

```

```{r remove cells}

combinedPools <- subset(combinedPools, ident = "5", invert = TRUE)


gene_modules <- find_gene_modules(p471monocle, resolution=1e-2)

cell_group_df<- tibble::tibble(cell=row.names(colData(p471monocle)), 
               cell_group=p471monocle@colData$teatmentbycluster)

agg_mat <- aggregate_gene_expression(p471monocle, gene_modules, cell_group_df)

row.names(agg_mat) <- stringr::str_c("Module ", row.names(agg_mat))


for(i in 1:20){
  name <- paste0("module", i)
  print(paste("Processing", name))
  combinedPools <- AddModuleScore(combinedPools,
               features = list(gene_modules$id[gene_modules$module == i]),
               ctrl = 200,
               name = paste0("GeneModule", i, "_")
               )
  for(i in c(0:4,6)){
    for(z in 1:20){
      name <- paste0("GeneModule_", z, "_Cluster", i)
      
  p <-  FeaturePlot(subset(combinedPools, ident = 6),
              features = paste0("GeneModule",z,"_1"), label = TRUE, repel = TRUE, pt.size = 2, split.by = "teatmentbycluster", cols = c("Blue", "Red"))
pdf(file = paste0("plots/geneModulesSC/", name, ".pdf"), height = 8, width = 16)
print(p)
dev.off()
    }
}
pdf("~/Documents/Module19.pdf", height = 20, width = 20)
grid.arrange(p, ncol = 6, nrow = 3)
dev.off()

MuteinSubset$barcode <- colnames(MuteinSubset)
MuteinSubset$UMAP_1 <- MuteinSubset@reductions$UMAP@cell.embeddings[,1]
MuteinSubset$UMAP_2 <- MuteinSubset@reductions$UMAP@cell.embeddings[,2]
write.csv(MuteinSubset@meta.data, file='/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/metadataMutein.csv', quote=F, row.names=F)

WTSubset$barcode <- colnames(WTSubset)
WTSubset$UMAP_1 <- WTSubset@reductions$UMAP@cell.embeddings[,1]
WTSubset$UMAP_2 <- WTSubset@reductions$UMAP@cell.embeddings[,2]
write.csv(WTSubset@meta.data, file='/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/metadataWT.csv', quote=F, row.names=F)

library(Matrix)
counts_matrixMutein <- GetAssayData(MuteinSubset, assay='RNA', slot='counts')
writeMM(counts_matrixMutein, file='/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/countsMutein.mtx')

counts_matrixWT <- GetAssayData(WTSubset, assay='RNA', slot='counts')
writeMM(counts_matrixWT, file='/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/countsWT.mtx')

write.csv(MuteinSubset@reductions$pca_RNA@cell.embeddings, file='/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/pcaMutein.csv', quote=F, row.names=F)
write.csv(WTSubset@reductions$pca_RNA@cell.embeddings, file='/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/pcaWT.csv', quote=F, row.names=F)

write.table(
  data.frame('gene'=rownames(counts_matrixMutein)),file='/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/gene_names.csv',
  quote=F,row.names=F,col.names=F
)


MuteinVelocities <- read.csv("/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/MuteinVelocities.csv")

WTVelocities <- read.csv("/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/WTVelocities.csv")

PBSSubset$barcode <- colnames(PBSSubset)
PBSSubset$UMAP_1 <- PBSSubset@reductions$UMAP@cell.embeddings[,1]
PBSSubset$UMAP_2 <- PBSSubset@reductions$UMAP@cell.embeddings[,2]
write.csv(PBSSubset@meta.data, file='/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/metadataPBS.csv', quote=F, row.names=F)

library(Matrix)
counts_matrixPBS <- GetAssayData(PBSSubset, assay='RNA', slot='counts')
writeMM(counts_matrixPBS, file='/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/countsPBS.mtx')

write.csv(PBSSubset@reductions$pca_RNA@cell.embeddings, file='/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/pcaPBS.csv', quote=F, row.names=F)

read.csv("/Volumes/Bioinformatics/workspace/mlawrance/230124_P471RNAVelo/WTVelocities.csv")
```


```{r Venn Diagram}

combo <- c(A = 52, B = 1203, C = 1268, D = 661, "A&B" = 31, "A&C" = 5, "A&D" = 17, "B&C" = 175, "B&D" = 282, "C&D" = 33, "A&B&C" = 4, "A&C&D" = 1, "A&B&D" = 13, "B&C&D" = 19)

write.xlsx(SigWT, file="~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/tables/VennGenes.xlsx", sheetName="Fc.WT", row.names=TRUE)
write.xlsx(SigMut, file="~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/tables/VennGenes.xlsx", sheetName="Fc.Mut24", row.names=TRUE, append = TRUE)
write.xlsx(TCRMutvsPBS, file="~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/tables/VennGenes.xlsx", sheetName="TCR", row.names=TRUE, append = TRUE)
write.xlsx(Stat5MutvsPBS, file="~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/tables/VennGenes.xlsx", sheetName="Stat5", row.names=TRUE, append = TRUE)


```

```{r Figure}

PanelA <- <- ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS",],
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1][combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS"],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2][combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS"],
          color = seurat_clusters)) +
  ggrastr::rasterise(
    geom_point(size = .25, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(values = pal.clusters) +
  facet_wrap(~treatment, nrow = 1) +
  guides(color = "none") + theme_bw()

PanelA <- ggplot(
    data = combinedPools@meta.data,
    mapping =
      aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
          y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
          color = seurat_clusters)) +
  ggrastr::rasterise(
    geom_point(size = .25, alpha = 0.8),
    dpi = rasterResolutionDpi) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(values = pal.clusters) +
  facet_wrap(~treatment, nrow = 1) +
  guides(color = "none") + theme_bw()

        clusterCounts <- combinedPools@meta.data %>%
  dplyr::group_by(seurat_clusters, treatment) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(treatment) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(treatment)

PanelB <- clusterCounts %>%
  ggplot(aes(x = treatment,
             y = pctCells,
             fill = seurat_clusters)) +
  geom_col() +
  scale_fill_manual(labels = c("C1 (cTR)", "C2 (eTR)", "C3 (STAT5)", "C4 (cell cycle 1)", "C5 (cell cycle 2)", "C6 (activated)"), values = pal.clusters) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Treatment",
       y = "% of cells",
       fill = "Cluster")


topMarkersDotPlot <- clusterMarkers$seurat_clusters_RNA_0p3 %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n=100) %>%
  dplyr::slice_max(avg_log2FC, n = 2) %>%
  dplyr::mutate(fc_order = order(avg_log2FC))

PancelC <- DotPlot(combinedPools, 
        features = unique(topMarkersDotPlot$gene))+ 
  RotatedAxis() + labs(x = "Marker Genes", y = "Cluster")

selectedFeatures <- unique(topMarkersDotPlot$gene)

clusterAverages <- AverageExpression(combinedPools, 
                                    assays = "RNA",
                                    features = selectedFeatures,
                                    group.by = "seurat_clusters",
                                    return.seurat = T) 

avgExprMat <- clusterAverages[["RNA"]]@data %>% as.matrix()

scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- c("C1 (cTR)", "C2 (eTR)", "C3 (STAT5)", "C4 (cell cycle 1)", "C5 (cell cycle 2)", "C6 (activated)")

columnAnno <- columnAnnotation(Cluster = clusterAnno,
                              col = list(Cluster = clustercolors[-7]),
                              simple_anno_size = unit(0.25, "cm"), show_legend = TRUE)


PanelCAlt <-as.ggplot(Heatmap(scaledExprMat, 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE, column_names_rot = 0,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10), show_heatmap_legend = FALSE, show_column_names = FALSE))

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/PanelFigureVer1.pdf", height = 20, width = 20)
grid.arrange(
    grobs = list(PanelA, PanelB, PanelCAlt, PanelD),
    widths = c(2, 2, 1),
    layout_matrix = rbind(c(1, 1),
                          c(2, 3),
                          c(4, 4)
))
dev.off()
```



```{r GSEA}
treatmentMarkers <- FindMarkers(combinedPools, ident.1 = "Fc.Mut24", ident.2 = "PBS", logfc.threshold = .01)

topMarkers <- treatmentMarkers %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::filter(p_val_adj < .05) %>%
  dplyr::mutate(fc_order = order(avg_log2FC))

original_gene_list <- treatmentMarkers$avg_log2FC
names(original_gene_list) <- rownames(treatmentMarkers)
gene_list<-na.omit(original_gene_list)
gene_list = sort(gene_list, decreasing = TRUE)

gse <- gseGO(geneList=gene_list, 
             ont ="ALL", 
             keyType = "SYMBOL", 
             nPerm = 10000, 
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = org.Mm.eg.db, 
             pAdjustMethod = "none")


```

```{r TCR inclusion and analysis}
add_clonotype <- function(tcr_prefix, seurat_obj, type="t"){
    tcr <- read.csv(paste(tcr_prefix,"filtered_contig_annotations.csv", sep=""))
    
    # Remove the -1 at the end of each barcode.
    # Subsets so only the first line of each barcode is kept,
    # as each entry for given barcode will have same clonotype.
    tcr <- tcr[!duplicated(tcr$barcode), ]
    
    # Only keep the barcode and clonotype columns. 
    # We'll get additional clonotype info from the clonotype table.
    tcr <- tcr[,c("barcode", "raw_clonotype_id")]
    names(tcr)[names(tcr) == "raw_clonotype_id"] <- "clonotype_id"
    
    # Clonotype-centric info.
    clono <- read.csv(paste(tcr_prefix,"clonotypes.csv", sep=""))
    
    # Slap the AA sequences onto our original table by clonotype_id.
    tcr <- merge(tcr, clono[, c("clonotype_id", "cdr3s_aa")])
    names(tcr)[names(tcr) == "cdr3s_aa"] <- "cdr3s_aa"
    
    # Reorder so barcodes are first column and set them as rownames.
    tcr <- tcr[, c(2,1,3)]
    rownames(tcr) <- tcr[,1]
    tcr[,1] <- NULL
    colnames(tcr) <- paste(type, colnames(tcr), sep="_")
    # Add to the Seurat object's metadata.
    clono_seurat <- AddMetaData(object=seurat_obj, metadata=tcr)
    return(clono_seurat)
}

Pool1 <- add_clonotype("/Volumes/Bioinformatics/pipeline/Illumina/MultiFlowcellProjects/Project_P471-1Processed_CellRangerV611_220413/multi/Pool_1/outs/per_sample_outs/Pool_1/vdj_t/", Pool1)

Pool2 <- add_clonotype("/Volumes/Bioinformatics/pipeline/Illumina/MultiFlowcellProjects/Project_P471-1Processed_CellRangerV611_220413/multi/Pool_2/outs/per_sample_outs/Pool_2/vdj_t/", Pool2)

Pool3 <- add_clonotype("/Volumes/Bioinformatics/pipeline/Illumina/MultiFlowcellProjects/Project_P471-1Processed_CellRangerV611_220413/multi/Pool_3/outs/per_sample_outs/Pool_3/vdj_t/", Pool3)

Pool4<- add_clonotype("/Volumes/Bioinformatics/pipeline/Illumina/MultiFlowcellProjects/Project_P471-1Processed_CellRangerV611_220413/multi/Pool_4/outs/per_sample_outs/Pool_4/vdj_t/", Pool4)



CombinedPoolsTest <- add_clonotype("/Volumes/Bioinformatics/pipeline/Illumina/MultiFlowcellProjects/Project_P471-1Processed_CellRangerV611_220413/multi/Pool_2/outs/per_sample_outs/Pool_2/vdj_t/", combinedPools)

clusterCounts <- combinedPools@meta.data[combinedPools@meta.data$t_cdr3s_aa %in% searchGroup,] %>%
  dplyr::group_by(t_cdr3s_aa, seurat_clusters) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(seurat_clusters) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(t_cdr3s_aa) %>% arrange(desc(nCells))

gClusterCounts <-clusterCounts %>%
  ggplot(aes(x = reorder(t_cdr3s_aa, -nCells, sum),
             y = nCells,
             fill = seurat_clusters)) +
  geom_col() +
  scale_fill_manual(values = big_colorblind_pal(7)[c(2, 3, 4,5,7)]) +
  labs(x = "",
       y = "# of cells",
       fill = "Cluster") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + scale_x_discrete(labels=paste0("clonotype ", 1:20)) + NoLegend()

clusterCounts2 <- combinedPools@meta.data[combinedPools@meta.data$t_cdr3s_aa %in% searchGroup,] %>%
  dplyr::group_by(t_cdr3s_aa, treatment) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(treatment) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(t_cdr3s_aa) %>% arrange(desc(nCells))

gClusterCounts2 <- clusterCounts2 %>%
  ggplot(aes(x = reorder(t_cdr3s_aa, -nCells, sum),
             y = nCells,
             fill = treatment)) +
  geom_col() +
  scale_fill_manual(values = c("red", "grey", "black")) +
  labs(x = "",
       y = "# of cells",
       fill = "Treatment") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + scale_x_discrete(labels=paste0("clonotype ", 1:20))

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/TCRClonotypeBreakdowntop20.pdf", height = 6, width = 10)
gClusterCounts + gClusterCounts2
dev.off()

clusterCounts3 <- combinedPools@meta.data %>%
  dplyr::group_by(t_cdr3s_aa, treatment) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(treatment) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(t_cdr3s_aa) %>% arrange(desc(nCells))

ggplot(data = clusterCounts3, aes(x=nCells, fill=treatment)) + geom_histogram() + scale_y_break(c(190,37900))+ theme_bw() +   theme(axis.line = element_line(colour = "black"), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.border = element_blank(), panel.background = element_blank()) + scale_fill_manual(values = c("red", "gray", "black"))

ggplot(data = clusterCounts3, aes(x=nCells, fill=treatment)) + geom_histogram() + coord_trans(y = "log1p") + theme_bw() +   theme(axis.line = element_line(colour = "black"), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.border = element_blank(), panel.background = element_blank()) + scale_fill_manual(values = c("red", "gray", "black")) + labs(x = "Number of Cells", y = "Number of Clones") + guides(fill=guide_legend(title="Treatment"))

```

```{r GSEA}
countsPCnorm <- pseudobulktotal
rownames(countsPCnorm) <- toupper(rownames(countsPCnorm))
hallmarkSig <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/Endosome Recycling Genes.csv")

clusters <- str_sub(colnames(countsPCnorm),-1,-1)
design <- model.matrix(~ 0+factor(clusters))
colnames(design) <- 0:6

resultsEndo <- data.frame()
for(i in 1:length(colnames(my.contrasts))){
entry <-roastGOLvsPBO_Wk0_Hallmark <- roast(
    y=countsPCnorm,
    index=ids2indices(hallmarkSig,
                      identifiers=rownames(countsPCnorm)),
    design=design,
    contrast=my.contrasts[,i])
resultsEndo <- rbind(resultsEndo,entry)
}

resultsEndo <- dplyr::arrange(resultsEndo, FDR)

hallmarkSig <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/TCR-upregulated Treg genes.csv")

hallmarkSig$Gene.symbol <- toupper(hallmarkSig$Gene.symbol)

resultsTCR <- data.frame()
for(i in 1:length(colnames(my.contrasts))){
entry <-roastGOLvsPBO_Wk0_Hallmark <- roast(
    y=countsPCnorm,
    index=ids2indices(unique(hallmarkSig$Gene.symbol),
                      identifiers=rownames(countsPCnorm)),
    design=design,
    contrast=my.contrasts[,i])
resultsTCR <- rbind(resultsTCR,entry)
}

resultsTCR <- dplyr::arrange(resultsTCR, FDR)

hallmarkSig <- read.table("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/GSE84553_matrix.txt")

hallmarkSig <- dplyr::arrange(hallmarkSig, STAT5bCA_vs_Control_Treg_Padj) %>% filter(STAT5bCA_vs_Control_Treg_Padj < .05) %>% filter(STAT5bCA_vs_Control_Treg_Log2FC > 1) 

hallmarkSig <- toupper(hallmarkSig$Gene_Symbol)

resultsStat5 <- data.frame()
for(i in 1:length(colnames(my.contrasts))){
entry <- roast(
    y=countsPCnorm,
    index=ids2indices(hallmarkSig,
                      identifiers=rownames(countsPCnorm)),
    design=design,
    contrast=my.contrasts[,i])
resultsStat5 <- rbind(resultsStat5,entry)
}

rownames(resultsStat5) <- colnames(my.contrasts)
resultsStat5 <- dplyr::arrange(resultsStat5, FDR)

hallmarkSig <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/CtrGenes.csv")
hallmarkSig <- toupper(hallmarkSig$X)


resultsCTR <- data.frame()
for(i in 1:length(colnames(my.contrasts))){
entry <- roast(
    y=countsPCnorm,
    index=ids2indices(hallmarkSig,
                      identifiers=rownames(countsPCnorm)),
    design=design,
    contrast=my.contrasts[,i])
resultsCTR <- rbind(resultsCTR,entry)
}

rownames(resultsCTR) <- colnames(my.contrasts)
resultsCTR <- dplyr::arrange(resultsCTR, FDR)

hallmarkSig <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/etrgenes.csv")
hallmarkSig <- toupper(hallmarkSig$X)


resultsETR <- data.frame()
for(i in 1:length(colnames(my.contrasts))){
entry <- roast(
    y=countsPCnorm,
    index=ids2indices(hallmarkSig,
                      identifiers=rownames(countsPCnorm)),
    design=design,
    contrast=my.contrasts[,i])
resultsETR <- rbind(resultsETR,entry)
}

rownames(resultsETR) <- colnames(my.contrasts)
resultsETR <- dplyr::arrange(resultsETR, FDR)

write.xls

addWorksheet(wb = wbook, sheetName = "TCR")
addWorksheet(wb = wbook, sheetName = "STAT5")
addWorksheet(wb = wbook, sheetName = "ETR")
addWorksheet(wb = wbook, sheetName = "CTR")
addWorksheet(wb = wbook, sheetName = "Endo")

writeData(wb = wbook, sheet = "TCR", x = resultsTCR, rowNames = TRUE)
writeData(wb = wbook, sheet = "STAT5", x = resultsStat5, rowNames = TRUE)
writeData(wb = wbook, sheet = "ETR", x = resultsETR,rowNames = TRUE)
writeData(wb = wbook, sheet = "CTR", x = resultsCTR,rowNames = TRUE)
writeData(wb = wbook, sheet = "Endo", x = resultsEndo,rowNames = TRUE)
```

```{r redo as treatment}

hallmarkSig <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/Endosome Recycling Genes.csv")


treatments <- str_match(colnames(countsPCnorm), "_\\s*(.*?)\\s*_")
treatments <- treatments[,2]

treatments<- gsub("PBS", "0", treatments)

treatments<- gsub("Fc.WT", "1", treatments)

treatments<- gsub("Fc.Mut24", "2", treatments)


design <- model.matrix(~ 0+factor(treatments))
colnames(design) <- c("PBS", "Fc.WT", "Fc.Mut24")


my.contrasts <-makeContrasts(PBS-Fc.WT, Fc.WT-PBS, PBS-Fc.Mut24, Fc.Mut24-PBS, Fc.WT-Fc.Mut24, Fc.Mut24-Fc.WT, levels=design)


resultsEndo <- data.frame()
for(i in 1:length(colnames(my.contrasts))){
entry <-roastGOLvsPBO_Wk0_Hallmark <- roast(
    y=countsPCnorm,
    index=ids2indices(hallmarkSig,
                      identifiers=rownames(countsPCnorm)),
    design=design,
    contrast=my.contrasts[,i])
rownames(entry) <-  colnames(my.contrasts)[i]
resultsEndo <- rbind(resultsEndo,entry)
}

resultsEndo <- dplyr::arrange(resultsEndo, FDR)

hallmarkSig <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/TCR-upregulated Treg genes.csv")

hallmarkSig$Gene.symbol <- toupper(hallmarkSig$Gene.symbol)

resultsTCR <- data.frame()
for(i in 1:length(colnames(my.contrasts))){
entry <-roastGOLvsPBO_Wk0_Hallmark <- roast(
    y=countsPCnorm,
    index=ids2indices(unique(hallmarkSig$Gene.symbol),
                      identifiers=rownames(countsPCnorm)),
    design=design,
    contrast=my.contrasts[,i])
rownames(entry) <-  colnames(my.contrasts)[i]
resultsTCR <- rbind(resultsTCR,entry)
}

resultsTCR <- dplyr::arrange(resultsTCR, FDR)

hallmarkSig <- read.table("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/GSE84553_matrix.txt")

hallmarkSig <- dplyr::arrange(hallmarkSig, STAT5bCA_vs_Control_Treg_Padj) %>% filter(STAT5bCA_vs_Control_Treg_Padj < .05) %>% filter(STAT5bCA_vs_Control_Treg_Log2FC > 1) 

hallmarkSig <- toupper(hallmarkSig$Gene_Symbol)

resultsStat5 <- data.frame()
for(i in 1:length(colnames(my.contrasts))){
entry <- roast(
    y=countsPCnorm,
    index=ids2indices(hallmarkSig,
                      identifiers=rownames(countsPCnorm)),
    design=design,
    contrast=my.contrasts[,i])
rownames(entry) <-  colnames(my.contrasts)[i]
resultsStat5 <- rbind(resultsStat5,entry)
}

rownames(resultsStat5) <- colnames(my.contrasts)
resultsStat5 <- dplyr::arrange(resultsStat5, FDR)

hallmarkSig <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/CtrGenes.csv")
hallmarkSig <- toupper(hallmarkSig$X)


resultsCTR <- data.frame()
for(i in 1:length(colnames(my.contrasts))){
entry <- roast(
    y=countsPCnorm,
    index=ids2indices(hallmarkSig,
                      identifiers=rownames(countsPCnorm)),
    design=design,
    contrast=my.contrasts[,i])
rownames(entry) <-  colnames(my.contrasts)[i]
resultsCTR <- rbind(resultsCTR,entry)
}

rownames(resultsCTR) <- colnames(my.contrasts)
resultsCTR <- dplyr::arrange(resultsCTR, FDR)

hallmarkSig <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/etrgenes.csv")
hallmarkSig <- toupper(hallmarkSig$X)


resultsETR <- data.frame()
for(i in 1:length(colnames(my.contrasts))){
entry <- roast(
    y=countsPCnorm,
    index=ids2indices(hallmarkSig,
                      identifiers=rownames(countsPCnorm)),
    nrot = 9999,
    design=design,
    contrast=my.contrasts[,i])
resultsETR[i] <- entry
}

rownames(resultsETR) <- colnames(my.contrasts)
resultsETR <- dplyr::arrange(resultsETR, FDR)

write.xls



wbook <- createWorkbook()

addWorksheet(wb = wbook, sheetName = "TCR")
addWorksheet(wb = wbook, sheetName = "STAT5")
addWorksheet(wb = wbook, sheetName = "ETR")
addWorksheet(wb = wbook, sheetName = "CTR")
addWorksheet(wb = wbook, sheetName = "Endo")

writeData(wb = wbook, sheet = "TCR", x = resultsTCR, rowNames = TRUE)
writeData(wb = wbook, sheet = "STAT5", x = resultsStat5, rowNames = TRUE)
writeData(wb = wbook, sheet = "ETR", x = resultsETR,rowNames = TRUE)
writeData(wb = wbook, sheet = "CTR", x = resultsCTR,rowNames = TRUE)
writeData(wb = wbook, sheet = "Endo", x = resultsEndo,rowNames = TRUE)

saveWorkbook(wb = wbook, file = "~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/tables/GSEAResultsTreatment.xlsx", overwrite = TRUE)
```

```{r Roast altrnative}
EndoGenes <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/Endosome Recycling Genes.csv")

TCRGenes <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/TCR-upregulated Treg genes.csv")

TCRGenes<- toupper(TCRGenes$Gene.symbol)

Stat5Genes <- read.table("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/GSE84553_matrix.txt")

Stat5Genes <- dplyr::arrange(Stat5Genes, STAT5bCA_vs_Control_Treg_Padj) %>% filter(STAT5bCA_vs_Control_Treg_Padj < .05) %>% filter(STAT5bCA_vs_Control_Treg_Log2FC > 1) 

Stat5Genes <- toupper(Stat5Genes$Gene_Symbol)


CTRgenes <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/CtrGenes.csv")
CTRgenes <- toupper(CTRgenes$X)

ETRgenes <- read.csv("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/etrgenes.csv")
ETRgenes <- toupper(ETRgenes$X)

index1 <- ids2indices(EndoGenes, identifiers = rownames(countsPCnorm))
index2 <-ids2indices(TCRGenes, identifiers = rownames(countsPCnorm))
index3 <-ids2indices(Stat5Genes, identifiers = rownames(countsPCnorm))
index4 <-ids2indices(CTRgenes, identifiers = rownames(countsPCnorm))
index5 <-ids2indices(ETRgenes, identifiers = rownames(countsPCnorm))

results <- list()
for(i in 1:length(colnames(my.contrasts))){
entry <- mroast(
    y=countsPCnorm,
    index=list(Endo=index1,TCR=index2, STAT5=index3, CTR=index4, ETR=index5),
    nrot = 9999,
    design=design,
    contrast=my.contrasts[,i])
results[[colnames(my.contrasts)[i]]] <- entry
}

for(i in 1:length(names(results))){
  results[[i]]$contrast <- names(results[i]
  results[[i]]$indent1 <- names(results[i]) %>% gsub(" -.*", "", .)
  results[[i]]$indent2 <- names(results[i]) %>% gsub(".*- ", "", .)
}

for(i in 1:length(names(results))){
  results[[i]]$colorValue <- results[[i]]$FDR*1
  results[[i]]$colorValue[results[[i]]$Direction == "Down"] <- results[[i]][results[[i]]$Direction == "Down"]$FDR*-1
}

figureFrame <- rbind(results$`Fc.Mut24 - PBS`, results$`Fc.Mut24 - Fc.WT`, results$`Fc.WT - PBS`)


ggplot(figureFrame, aes(x = contrast, y = set)) +
    geom_point(aes(size = figureFrame$NGenes, color = figureFrame$colorValue)) + scale_color_gradient(low ="red", high = "gray") + scale_x_discrete(labels=c("Fc.Mut24 - Fc.WT" = "Fc.WT", "Fc.Mut24 - PBS" = "PBS", "Fc.WT - PBS" = "PBS")) + theme(plot.margin = unit(c(1, 0.5, 0.5, 0.5), "inches")) + NoLegend()

MutLabel <- grobTree(textGrob("Fc.Mut24", x=0.1,  y=0.95, hjust=0,
  gp=gpar(col="red", fontsize=13, fontface="bold")))

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/dotPlotFirstPass.pdf", width = 3, height = 6)
ggplot(figureFrame, aes(x = contrast, y = set)) +
    geom_point(aes(size = figureFrame$NGenes, color = figureFrame$colorValue)) + scale_color_gradient(low ="red", high = "gray") + scale_x_discrete(labels=c("Fc.Mut24 - Fc.WT" = "Fc.WT", "Fc.Mut24 - PBS" = "PBS", "Fc.WT - PBS" = "PBS")) + theme_bw() + theme(plot.margin = unit(c(1, 0.5, 0.5, 0.5), "inches"), axis.text.x = element_text(angle =45, vjust = .5)) + NoLegend() +coord_cartesian(xlim = c(1, 3), clip = "off") + annotation_custom(MutLabel) + annotation_custom(FCLabel) + labs(x = "Comparison", y = "Reference GeneSet")
dev.off()
```

```{r heatmaps of pathway genes}

Idents(combinedPools) <- "seurat_clusters"

EndoAverages <- AverageExpression(combinedPools, 
                                    assays = "RNA",
                                    features = tolower(EndoGenes$Symbol) %>% str_to_title(),
                                    group.by = "treatment",
                                    return.seurat = T) 

avgExprMat <- EndoAverages[["RNA"]]@data %>% as.matrix()



scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- rownames(EndoAverages@meta.data)

columnAnno <- columnAnnotation(Treatment = clusterAnno,
                              col = list(Treatment = treatmentcolors),
                              simple_anno_size = unit(0.25, "cm"), show_legend = FALSE)

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/hallmarkHeatmaps/Endo.pdf",
    height = 8,
    width = 4)

Heatmap(na.omit(scaledExprMat),
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10),
        column_title = "Endosomal Recycling Gene Expression")

invisible(dev.off())

TCRAverages <- AverageExpression(combinedPools, 
                                    assays = "RNA",
                                    features = tolower(TCRGenes) %>% str_to_title(),
                                    group.by = "treatment",
                                    return.seurat = T) 

avgExprMat <- TCRAverages[["RNA"]]@data %>% as.matrix()



scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- rownames(TCRAverages@meta.data)

columnAnno <- columnAnnotation(Treatment = clusterAnno,
                              col = list(Treatment = treatmentcolors),
                              simple_anno_size = unit(0.25, "cm"), show_legend = FALSE)

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/hallmarkHeatmaps/TCR.pdf",
    height = 8,
    width = 4)

Heatmap(na.omit(scaledExprMat),
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10),
        column_title = "TCR Gene Expression")

invisible(dev.off())

Stat5Averages <- AverageExpression(combinedPools, 
                                    assays = "RNA",
                                    features = tolower(Stat5Genes) %>% str_to_title(),
                                    group.by = "treatment",
                                    return.seurat = T) 

avgExprMat <- Stat5Averages[["RNA"]]@data %>% as.matrix()



scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- rownames(Stat5Averages@meta.data)

columnAnno <- columnAnnotation(Treatment = clusterAnno,
                              col = list(Treatment = treatmentcolors),
                              simple_anno_size = unit(0.25, "cm"), show_legend = FALSE)

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/hallmarkHeatmaps/Stat5.pdf",
    height = 8,
    width = 4)

Heatmap(na.omit(scaledExprMat),
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10),
        column_title = "Stat5 Gene Expression")

invisible(dev.off())

ETRAverages <- AverageExpression(combinedPools, 
                                    assays = "RNA",
                                    features = tolower(ETRgenes) %>% str_to_title(),
                                    group.by = "treatment",
                                    return.seurat = T) 

avgExprMat <- ETRAverages[["RNA"]]@data %>% as.matrix()



scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- rownames(ETRAverages@meta.data)

columnAnno <- columnAnnotation(Treatment = clusterAnno,
                              col = list(Treatment = treatmentcolors),
                              simple_anno_size = unit(0.25, "cm"), show_legend = FALSE)

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/hallmarkHeatmaps/ETR.pdf",
    height = 8,
    width = 4)

Heatmap(na.omit(scaledExprMat),
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10),
        column_title = "ETR Gene Expression")

invisible(dev.off())


CTRAverages <- AverageExpression(combinedPools, 
                                    assays = "RNA",
                                    features = tolower(CTRgenes) %>% str_to_title(),
                                    group.by = "treatment",
                                    return.seurat = T) 

avgExprMat <- CTRAverages[["RNA"]]@data %>% as.matrix()



scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- rownames(CTRAverages@meta.data)

columnAnno <- columnAnnotation(Treatment = clusterAnno,
                              col = list(Treatment = treatmentcolors),
                              simple_anno_size = unit(0.25, "cm"), show_legend = FALSE)

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/hallmarkHeatmaps/CTR.pdf",
    height = 8,
    width = 4)

Heatmap(na.omit(scaledExprMat),
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10),
        column_title = "CTR Gene Expression")

invisible(dev.off())
```

```{r}

combinedPools <- AddModuleScore(combinedPools, assay = "RNA", features = list(EndoGenes, TCRGenes, Stat5Genes, CTRgenes, ETRgenes), name = c("Endosomal", "TCR", "Stat5", "CTR", "ETR"))

plot1 <- VlnPlot(combinedPools, "Endosomal1", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,.32)) + scale_fill_manual(values = c("black", "grey", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = .2) + stat_compare_means(comparisons = list(c(2,3)), label = "p.signif", label.y = .24) + stat_compare_means(comparisons = list(c(1,3)), label = "p.signif", label.y = .28) + NoLegend() + labs(x = "Treatment", y = "Gene Module Score", title = "Endosome Related Genes")

plot2 <- VlnPlot(combinedPools, "TCR2", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,.32)) + scale_fill_manual(values = c("black", "grey", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = .2) + stat_compare_means(comparisons = list(c(2,3)), label = "p.signif", label.y = .24) + stat_compare_means(comparisons = list(c(1,3)), label = "p.signif", label.y = .28) + NoLegend() + labs(x = "Treatment", y = "Gene Module Score", title = "TCR Dependent Genes")

plot3 <- VlnPlot(combinedPools, "Stat53", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,.32)) + scale_fill_manual(values = c("black", "grey", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = .2) + stat_compare_means(comparisons = list(c(2,3)), label = "p.signif", label.y = .24) + stat_compare_means(comparisons = list(c(1,3)), label = "p.signif", label.y = .28) + NoLegend() + labs(x = "Treatment", y = "Gene Module Score", title = "Stat5 Dependent Genes")

plot4 <- VlnPlot(combinedPools, "CTR4", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,.32)) + scale_fill_manual(values = c("black", "grey", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = .2) + stat_compare_means(comparisons = list(c(2,3)), label = "p.signif", label.y = .24) + stat_compare_means(comparisons = list(c(1,3)), label = "p.signif", label.y = .28) + NoLegend() + labs(x = "Treatment", y = "Gene Module Score", title = "CTR Genes")

plot5 <- VlnPlot(combinedPools, "ETR5", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,.32)) + scale_fill_manual(values = c("black", "grey", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = .2) + stat_compare_means(comparisons = list(c(2,3)), label = "p.signif", label.y = .24) + stat_compare_means(comparisons = list(c(1,3)), label = "p.signif", label.y = .28) + NoLegend() + labs(x = "Treatment", y = "Gene Module Score", title = "ETR Genes")

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/geneModulesViolin.pdf", height = 12, width = 8)
grid.arrange(plot1 + geom_boxplot(width = .05, aes(color = "white")) + scale_color_manual(values = c("darkgreen", "gray","gray")), plot2, plot3, plot4, plot5)
dev.off()
```

```{r}
ggplot(
    data = combinedPools@meta.data,
    mapping =
        aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
            y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
            color = (TCR2 - Stat53))) +
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.8),
        dpi = rasterResolutionDpi) +
    labs(x = "UMAP 1", y = "UMAP 2") +
    scale_color_gradient2(low = "blue", mid = "gray", high = "red") +
    facet_wrap(~treatment, nrow = 1) +
    guides(color = "none") + theme_bw()

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/UMAPGeneScore.pdf", height = 8, width = 10)
ggplot(
    data = combinedPools@meta.data,
    mapping =
        aes(x = combinedPools@reductions$umap_RNA@cell.embeddings[,1],
            y = combinedPools@reductions$umap_RNA@cell.embeddings[,2],
            colour = combinedPools@meta.data$TCRStat5blend)) +
 scale_color_identity() +  
    ggrastr::rasterise(
        geom_point(size = 1, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "UMAP 1", y = "UMAP 2")+
    facet_wrap(~treatment, nrow = 1) +
    guides(color = "none") + theme_bw()
dev.off()

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/biaxialGeneScore.pdf", height = 5, width = 8)
ggplot(
    data = combinedPools@meta.data,
    mapping =
        aes(x = combinedPools@meta.data$TCR2,
            y = combinedPools@meta.data$Stat53,
            colour = combinedPools@meta.data$TCRStat5blend)) +
 scale_color_identity() +  
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score")+
    facet_wrap(~treatment, nrow = 1) +
    guides(color = "none") + theme_bw()
dev.off()
```


```{r}
# CTLA4, ENTPD1, NT5E, TGFB1, TNFRSF9, TNFRSF18, ICOS

pdf("MuteinBiaxial.pdf", width = 8, height = 8)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.Mut24"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.Mut24"],
            colour = "red")) +
 scale_color_identity() +      ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.Mut24")+
    guides(color = "none") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(hjust = 0.5)) + stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "gray") + lims(y = c(-.15, .2), x = c(-.1,.18))
dev.off()

pdf("PBSBiaxial.pdf", width = 8, height = 8)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "PBS",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "PBS"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "PBS"],
            colour = "black")) + scale_color_identity() + 
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "PBS")+
    guides(color = "none") +theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(hjust = 0.5)) + stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "gray") + lims(y = c(-.15, .2), x = c(-.1,.18))
dev.off()


pdf("WTBiaxial.pdf", width = 8, height = 8)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.WT",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.WT"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.WT"],
            colour = "gray")) +
 scale_color_identity() +  
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.WT")+
    guides(color = "none") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(hjust = 0.5)) + stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "black") + lims(y = c(-.15, .2), x = c(-.1,.18))

dev.off()


genesOfInterest <- c("Ctla4", "Entpd1","Nt5e","Tgfb1", "Tnfrsf9", "Tnfrs18", "Icos")


for(gene in genesOfInterest){
object <- paste0("Mut24", "_", gene)
assign(object ,ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24",],
    mapping =
        aes_(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.Mut24"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.Mut24"],
            colour = combinedPools@assays$RNA@data[gene,][combinedPools@meta.data$treatment == "Fc.Mut24"])) + 
    ggrastr::rasterise(
        geom_point(size = .75, alpha = 1),
        dpi = rasterResolutionDpi) +
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = paste("Fc.Mut24", gene, "Expression"))+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_gradient(high = "red", low = "gray"))

}

for(gene in genesOfInterest){
object <- paste0("WT", "_", gene)
assign(object ,ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.WT",],
    mapping =
        aes_(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.WT"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.WT"],
            colour = combinedPools@assays$RNA@data[gene,][combinedPools@meta.data$treatment == "Fc.WT"])) + 
    ggrastr::rasterise(
        geom_point(size = .75, alpha = 1),
        dpi = rasterResolutionDpi) +
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = paste("Fc.WT", gene, "Expression"))+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_gradient(high = "red", low = "gray"))

}


for(i in 1:length(genesOfInterest)){
gene <- genesOfInterest[i]
print(gene)
assign(paste0("PBS", "_", gene) ,ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "PBS",],
    mapping =
        aes_(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "PBS"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "PBS"],
            colour = combinedPools@assays$RNA@data[gene,][combinedPools@meta.data$treatment == "PBS"])) + 
    ggrastr::rasterise(
        geom_point(size = .75, alpha = 1),
        dpi = rasterResolutionDpi) +
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = paste("PBS", gene, "Expression"))+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_gradient(high = "red", low = "gray"))
}

for(i in pngs){
png(filename = paste(i, ".png"), width = 5, height = 5, units = "in", res = 300)
  print(eval(parse(text=i)))
dev.off()
}


ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS"],
            colour = combinedPools@meta.data$treatment[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS"])) +      ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.Mut24 + Fc.WT")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_manual(values = c("black", "red"))


png("PBSMutBiaxial.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS"],
            colour = combinedPools@meta.data$treatment[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "PBS"])) +      ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.Mut24 + PBS")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_manual(values = c("black", "red"))
dev.off()

png("WTMutBiaxial.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "Fc.WT",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "Fc.WT"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "Fc.WT"],
            colour = combinedPools@meta.data$treatment[combinedPools@meta.data$treatment == "Fc.Mut24" | combinedPools@meta.data$treatment == "Fc.WT"])) +      ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.Mut24 + Fc.WT")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_manual(values = c("gray", "red"))
dev.off()

png("PBSWTBiaxial.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.WT" | combinedPools@meta.data$treatment == "PBS",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.WT" | combinedPools@meta.data$treatment == "PBS"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.WT" | combinedPools@meta.data$treatment == "PBS"],
            colour = combinedPools@meta.data$treatment[combinedPools@meta.data$treatment == "Fc.WT" | combinedPools@meta.data$treatment == "PBS"])) +      ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.Mut24 + PBS")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_manual(values = c("black", "gray"))
dev.off()
```


```{r}
zobs <- (FisherZ(-.44)-FisherZ(-.034))/sqrt((1/(13665-3))+(1/(14979-3)))
2 * pnorm(-abs(zobs))

zobs <- (FisherZ(-.1)-FisherZ(-.034))/sqrt((1/(14054-3))+(1/(14979-3)))
2 * pnorm(-abs(zobs))

zobs <- (FisherZ(-.44)-FisherZ(-.1))/sqrt((1/(13665-3))+(1/(14054-3)))
2 * pnorm(-abs(zobs))

png("MuteinBiaxialEvsCTR.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24",],
    mapping =
        aes(x = combinedPools@meta.data$ETR5[combinedPools@meta.data$treatment == "Fc.Mut24"],
            y = combinedPools@meta.data$CTR4[combinedPools@meta.data$treatment == "Fc.Mut24"],
            colour = "red")) +
 scale_color_identity() +      ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "ETR Gene Score", y = "CTR Gene Score", title = "Fc.Mut24")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "black")
dev.off()

png("PBSBiaxialEvsCTR.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "PBS",],
    mapping =
        aes(x = combinedPools@meta.data$ETR5[combinedPools@meta.data$treatment == "PBS"],
            y = combinedPools@meta.data$CTR4[combinedPools@meta.data$treatment == "PBS"],
            colour = "black")) + scale_color_identity() + 
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "ETR Gene Score", y = "CTR Gene Score", title = "PBS")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "gray")
dev.off()


png("WTBiaxialEvsCTR.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.WT",],
    mapping =
        aes(x = combinedPools@meta.data$ETR5[combinedPools@meta.data$treatment == "Fc.WT"],
            y = combinedPools@meta.data$CTR4[combinedPools@meta.data$treatment == "Fc.WT"],
            colour = "gray")) +
 scale_color_identity() +  
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "ETR Gene Score", y = "CTR Gene Score", title = "Fc.WT")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) +stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "black")

dev.off()


png("MuteinBiaxialClustersTCRStat5_facet.png", width = 8, height = 8, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.Mut24"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.Mut24"],
            colour = combinedPools@meta.data$seurat_clusters[combinedPools@meta.data$treatment == "Fc.Mut24"])) +
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.Mut24")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "black") + scale_color_manual(values = big_colorblind_pal(6)) + facet_wrap(~seurat_clusters)
dev.off()

png("PBSBiaxialClustersTCRStat5_facet.png", width = 8, height = 8, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "PBS",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "PBS"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "PBS"],
            colour = combinedPools@meta.data$seurat_clusters[combinedPools@meta.data$treatment == "PBS"])) + 
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "PBS")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "gray") + scale_color_manual(values = big_colorblind_pal(6)) + facet_wrap(~seurat_clusters)
dev.off()


png("WTBiaxialClustersTCRStat5_facet.png", width = 8, height = 8, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.WT",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.WT"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.WT"],
            colour = combinedPools@meta.data$seurat_clusters[combinedPools@meta.data$treatment == "Fc.WT"])) +
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.WT")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) +stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "black") + scale_color_manual(values = big_colorblind_pal(6)) + facet_wrap(~seurat_clusters)

dev.off()



png("MuteinEndoTCRStat5Bixial.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.Mut24"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.Mut24"],
            colour = combinedPools@meta.data$Endosomal1[combinedPools@meta.data$treatment == "Fc.Mut24"])) +
    ggrastr::rasterise(
        geom_point(size = .75, alpha = 1),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.Mut24")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_gradient(high = "red", low = "gray")
dev.off()

png("PBSEndoTCRStat5Bixial.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "PBS",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "PBS"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "PBS"],
            colour = combinedPools@meta.data$Endosomal1[combinedPools@meta.data$treatment == "PBS"])) + 
    ggrastr::rasterise(
        geom_point(size = .75, alpha = 1),
        dpi = rasterResolutionDpi) +
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "PBS")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_gradient(high = "red", low = "gray")
dev.off()


png("WTEndoTCRStat5Bixial.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.WT",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.WT"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.WT"],
            colour = combinedPools@meta.data$Endosomal1[combinedPools@meta.data$treatment == "Fc.WT"])) +
    ggrastr::rasterise(
        geom_point(size = .75, alpha = 1),
        dpi = rasterResolutionDpi) +
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.WT")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_gradient(high = "red", low = "gray")

dev.off()
```


```{r Braxton Talk plot}
selectedFeatures <- c("Sell", "Ccr7", "Ly6c1", "Bcl2", "Cd44", "Icos", "Tigit", "Lag3", "Pdcd1", "Ctla4", "Itgb1", "Cxcr3", "Izumo1r", "Tubb5", "Socs2", "Il2ra", "Ifitm2", "Ifitm3", "Gzmb", "Vim", "Cish", "Rad51", "Hells", "Tubb4b", "Cenpf", "Mki67", "Hist1h2ae", "Hist1h1b", "Top2a", "Pclaf", "Tuba1b", "Tnfrsf9", "Tnfrsf4", "Orai1", "Zap70")

clusterAverages <- AverageExpression(combinedPools, 
                                    assays = "RNA",
                                    features = selectedFeatures,
                                    group.by = "seurat_clusters",
                                    return.seurat = T) 

avgExprMat <- clusterAverages[["RNA"]]@data %>% as.matrix()

scaledExprMat <- t(scale(t(avgExprMat)))

columnAnno <- columnAnnotation(Cluster = clusterAnno,
                              col = list(Cluster = clustercolors[-7]),
                              simple_anno_size = unit(0.25, "cm"), show_legend = TRUE)


geneHeatmapBraxton <-as.ggplot(Heatmap(scaledExprMat, 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = FALSE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = FALSE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE, column_names_rot = 0,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10), show_heatmap_legend = TRUE, show_column_names = FALSE))


selectedFeatures <- c("Satb1", "Lef1", "Tcf7", "Hmgn2", "Bach2", "Maf", "Batf", "Hif1a", "Ikzf2", "Nfatc1", "Tox", "Foxp3", "Myc", "E2f1", "Hmgb2", "Tfdp1", "E2f2", "Mybl2", "Nr4a1", "Nr4a3", "Egr2", "Egr2", "Rorc", "Tbx21", "Gata3", "Irf4")

clusterAverages <- AverageExpression(combinedPools, 
                                    assays = "RNA",
                                    features = selectedFeatures,
                                    group.by = "seurat_clusters",
                                    return.seurat = T) 

avgExprMat <- clusterAverages[["RNA"]]@data %>% as.matrix()

scaledExprMat <- t(scale(t(avgExprMat)))

columnAnno <- columnAnnotation(Cluster = clusterAnno,
                              col = list(Cluster = clustercolors[-7]),
                              simple_anno_size = unit(0.25, "cm"), show_legend = TRUE)


tfHeatmapBraxton <-as.ggplot(Heatmap(scaledExprMat, 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = FALSE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = FALSE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE, column_names_rot = 0,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10), show_heatmap_legend = TRUE, show_column_names = FALSE))

pdf("transcriptionFactorHeatmapYpdate.pdf", height = 10, width =8)
tfHeatmapBraxton
dev.off()

pdf("markerGeneHeatmapUpdated.pdf", height = 10, width =8)
geneHeatmapBraxton
dev.off()

pdf("fullHeatmap.pdf", height = 10, width =8)
PanelCAlt
dev.off()

pdf("correctedBargraph.pdf", height = 6, width =6)
PanelB
dev.off()

```

```{r New BraxtonFigures}

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/TCRClonotypeBreakdowntop20.pdf", height = 6, width = 14)
gClusterCounts + gClusterCounts2
dev.off()

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/UMAPnoGridFacet.pdf", height = 6, width = 10)
PanelA
dev.off()


pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/UMAPnoGridTotal.pdf", height = 6, width = 6)
PanelAcombined
dev.off()


EndoMutvsPBS <- topMutvsPBSTotal[rownames(topMutvsPBSTotal) %in% EndoGenes,]

EndoLabel <- ggplot(data = topMutvsPBSTotal, aes(x=logFC,
                                                y=-log10(adj.P.Val),
                                                color = logFC>0)) +
    geom_point(alpha=0.7,
               size=3,
               shape = 19) +
    geom_point(data = EndoMutvsPBS,
               aes(x=logFC, y= -log10(adj.P.Val)),
               color = "darkgreen", size=3,
               shape = 19, alpha = 1) +
  geom_text_repel(data=head(EndoMutvsPBS[order(EndoMutvsPBS$adj.P.Val),], 15),
                  label = rownames(head(EndoMutvsPBS[order(EndoMutvsPBS$adj.P.Val),], 15)),
                  max.overlaps = Inf, min.segment.length = unit(0, 'lines'),
                  size = 5,
                  col = "black",
                  nudge_y = .6,
                  nudge_x = .6) +
      scale_color_manual(values = c("darkcyan", "red"),
                         name = "",
                         label = NULL) +
        labs(x = "Log Fold-change",
         y = "-log10 adj.p.val",
         title = "Mut.24 vs PBS, Endothelial Genes\n79 Up | 89 Down") +
    theme(plot.title = element_text(hjust = .5)) +
    geom_hline(yintercept= -log10(.05),
               color="black",linetype="dotted",size=1.0) +
    NoLegend()

CombinedPools

Rab11a <- VlnPlot(subset(x = combinedPools, idents = c("Fc.Mut24", "PBS")), "Rab11a", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,3.7)) + scale_fill_manual(values = c("black", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = 3.4)+ NoLegend() + labs(x = "Treatment", y = "Expression", title = "Rab11a")

Def6 <- VlnPlot(subset(x = combinedPools, idents = c("Fc.Mut24", "PBS")), "Def6", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,3.7)) + scale_fill_manual(values = c("black", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = 3.4)+ NoLegend() + labs(x = "Treatment", y = "Expression", title = "Def6")


Rabggtb <- VlnPlot(subset(x = combinedPools, idents = c("Fc.Mut24", "PBS")), "Rabggtb", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,3.7)) + scale_fill_manual(values = c("black", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = 3.4)+ NoLegend() + labs(x = "Treatment", y = "Expression", title = "Rabggtb")

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/EndoVolcano.pdf", height = 6, width = 6)
EndoLabel
dev.off()

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/Rab11a.pdf", height = 6, width = 6)
Rab11a
dev.off()

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/Def6.pdf", height = 6, width = 6)
Def6
dev.off()

pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/Rabggtb.pdf", height = 6, width = 6)
Rabggtb
dev.off()

 Rab11b <- VlnPlot(subset(x = combinedPools, idents = c("Fc.Mut24", "PBS")), "Rab11b", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,3.7)) + scale_fill_manual(values = c("black", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = 3.4)+ NoLegend() + labs(x = "Treatment", y = "Expression", title = "Rab11b")
 
 pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/Rab11b.pdf", height = 6, width = 6)
Rab11b
dev.off()
 
  Rab8a <- VlnPlot(subset(x = combinedPools, idents = c("Fc.Mut24", "PBS")), "Rab8a", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,3.7)) + scale_fill_manual(values = c("black", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = 3.4)+ NoLegend() + labs(x = "Treatment", y = "Expression", title = "Rab8a")
  
  pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/Rab8a.pdf", height = 6, width = 6)
Rab8a
dev.off()

  Trat1 <- VlnPlot(subset(x = combinedPools, idents = c("Fc.Mut24", "PBS")), "Trat1", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,3.7)) + scale_fill_manual(values = c("black", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = 3.4)+ NoLegend() + labs(x = "Treatment", y = "Expression", title = "Trat1")
  
  pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/Trat1.pdf", height = 6, width = 6)
Trat1
dev.off()

  Lax1 <- VlnPlot(subset(x = combinedPools, idents = c("Fc.Mut24", "PBS")), "Lax1", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,3.7)) + scale_fill_manual(values = c("black", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = 3.4)+ NoLegend() + labs(x = "Treatment", y = "Expression", title = "Lax1")
  
  pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/Lax1.pdf", height = 6, width = 6)
Lax1
dev.off()

  Arf1 <- VlnPlot(subset(x = combinedPools, idents = c("Fc.Mut24", "PBS")), "Arf1", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,3.7)) + scale_fill_manual(values = c("black", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = 3.4)+ NoLegend() + labs(x = "Treatment", y = "Expression", title = "Arf1")
  
  pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/Arf1.pdf", height = 6, width = 6)
Arf1
dev.off()

  Rab8b <- VlnPlot(subset(x = combinedPools, idents = c("Fc.Mut24", "PBS")), "Rab8b", group.by = "treatment", pt.size = 0) + scale_y_continuous(limits = c(-.2,3.7)) + scale_fill_manual(values = c("black", "red")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = 3.4)+ NoLegend() + labs(x = "Treatment", y = "Expression", title = "Rab8b")
  
  pdf("~/Library/CloudStorage/Box-Box/Projects/P471-1_10x_analysis/plots/10102023/Rab8b.pdf", height = 6, width = 6)
Rab8b
dev.off()
```

```{r}

png("WT_TRGenes_Clusters_facet.png", width = 8, height = 8, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.WT",],
    mapping =
        aes(x = combinedPools@meta.data$ETR5[combinedPools@meta.data$treatment == "Fc.WT"],
            y = combinedPools@meta.data$CTR4[combinedPools@meta.data$treatment == "Fc.WT"],
            colour = combinedPools@meta.data$seurat_clusters[combinedPools@meta.data$treatment == "Fc.WT"])) +
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "eTR Gene Score", y = "cTR Gene Score", title = "Fc.WT")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) +stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "black") + scale_color_manual(values = big_colorblind_pal(6)) + facet_wrap(~seurat_clusters)

dev.off()

png("PBS_TRGenes_Clusters_facet.png", width = 8, height = 8, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "PBS",],
    mapping =
        aes(x = combinedPools@meta.data$ETR5[combinedPools@meta.data$treatment == "PBS"],
            y = combinedPools@meta.data$CTR4[combinedPools@meta.data$treatment == "PBS"],
            colour = combinedPools@meta.data$seurat_clusters[combinedPools@meta.data$treatment == "PBS"])) +
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "eTR Gene Score", y = "cTR Gene Score", title = "PBS")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) +stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "black") + scale_color_manual(values = big_colorblind_pal(6))+ facet_wrap(~seurat_clusters)

dev.off()

png("Mut_TRGenes_Clusters_facet.png", width = 8, height = 8, units = 'in', res = 300)
ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24",],
    mapping =
        aes(x = combinedPools@meta.data$ETR5[combinedPools@meta.data$treatment == "Fc.Mut24"],
            y = combinedPools@meta.data$CTR4[combinedPools@meta.data$treatment == "Fc.Mut24"],
            colour = combinedPools@meta.data$seurat_clusters[combinedPools@meta.data$treatment == "Fc.Mut24"])) +
    ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi) +
    labs(x = "eTR Gene Score", y = "cTR Gene Score", title = "Fc.WT")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) +stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "black") + scale_color_manual(values = big_colorblind_pal(6))+ facet_wrap(~seurat_clusters)

dev.off()

combinedPools@meta.data$gate <- "Low"
combinedPools@meta.data$gate[scale(combinedPools@meta.data$TCR2) + (.75*scale(combinedPools@meta.data$Stat53)) > .95] <- "High"

gateMut24 <- ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.Mut24",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.Mut24"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.Mut24"],
            colour = combinedPools@meta.data$seurat_clusters[combinedPools@meta.data$treatment == "Fc.Mut24"])) +
    scale_color_identity() +      ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.Mut24")+
guides(color = "none") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(hjust = 0.5))  + lims(y = c(-.15, .2), x = c(-.1,.18)) + scale_color_manual(values = big_colorblind_pal(7)) +stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "gray")

gateWT <- ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "Fc.WT",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "Fc.WT"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "Fc.WT"],
            colour = combinedPools@meta.data$seurat_clusters[combinedPools@meta.data$treatment == "Fc.WT"])) +
    scale_color_identity() +      ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "Fc.WT")+
    guides(color = "none") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(hjust = 0.5))  + lims(y = c(-.15, .2), x = c(-.1,.18)) + scale_color_manual(values = big_colorblind_pal(7)) + stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "gray")

gatePBS <- ggplot(
    data = combinedPools@meta.data[combinedPools@meta.data$treatment == "PBS",],
    mapping =
        aes(x = combinedPools@meta.data$TCR2[combinedPools@meta.data$treatment == "PBS"],
            y = combinedPools@meta.data$Stat53[combinedPools@meta.data$treatment == "PBS"],
            colour = combinedPools@meta.data$seurat_clusters[combinedPools@meta.data$treatment == "PBS"])) +
    scale_color_identity() +      ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "PBS")+
    guides(color = "none") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(hjust = 0.5))  + lims(y = c(-.15, .2), x = c(-.1,.18)) + scale_color_manual(values = big_colorblind_pal(7)) + stat_cor(color = "black") +   stat_smooth(method = "lm",
              formula = y ~ x,
              geom = "smooth", color = "gray")

gateTotal <- ggplot(
    data = combinedPools@meta.data,
    mapping =
        aes(x = combinedPools@meta.data$TCR2,
            y = combinedPools@meta.data$Stat53,
            colour = combinedPools@meta.data$seurat_clusters)) +
    scale_color_identity() +      ggrastr::rasterise(
        geom_point(size = .25, alpha = 0.75),
        dpi = rasterResolutionDpi)+
    labs(x = "TCR Gene Score", y = "Stat5 Gene Score", title = "All Treatments, high-low Gate cutoff")+
    guides(color = "none") + theme_bw() +theme(plot.title = element_text(hjust = 0.5)) + scale_color_manual(values = big_colorblind_pal(7)) 

Idents(combinedPools) <- "treatment"

VlnPlot(subset(combinedPools, idents = "Fc.Mut24"), group.by = "gate", features = "TCR2")

png("ViolinCTLA4.png", height = 4, width =8, units = "in", res = 300)
grid.arrange(PBSCTLA4Gate, MuteinCTLA4Gate, nrow = 1)
dev.off()

Idents(combinedPools) <- "gate"
VlnPlot(subset(combinedPools, idents = "High"), group.by = "treatment", features = "Ctla4")

png("ViolinCTLA4_2.png", height = 4, width =8, units = "in", res = 300)
VlnPlot(subset(combinedPools, idents = "High"), group.by = "treatment", features = "Ctla4", cols = c("black", "gray", "red")) + labs(title = "High gate cells only, CTLA4")
dev.off()

png("gateBiaxial", height = 4, width =8, units = "in", res = 300)
gateTotal
dev.off()

Idents(MuteinOnly) <- "gate"
for(entry in c("Ctla4", "Cd44", "Icos", "Il2ra", "Tnfrsf9", "Tnfrsf18", "Ifitm2", "Ifitm3", "Gzmb", "Tgfb1")){
assign(paste0(entry, "Violin"),VlnPlot(subset(x = combinedPools, idents = c("High", "Low")), entry, group.by = "gate", pt.size = 0) + scale_y_continuous(limits = c(-.2,5)) + scale_fill_manual(values = c("coral", "darkcyan")) + stat_compare_means(comparisons = list(c(1,2)), label = "p.signif", label.y = 4.4)+ NoLegend() + labs(x = "Treatment", y = "Expression", title = entry))
}

pdf("ViolinsGatesAllTreatments.pdf", height = 4, width = 20)
grid.arrange(Ctla4Violin, Cd44Violin, IcosViolin, Il2raViolin, Tnfrsf9Violin, Tnfrsf18Violin, Ifitm2Violin, Ifitm3Violin, GzmbViolin, nrow = 1)
dev.off()

pdf("MuteinClusters.pdf", height = 8, width =8)
gateMut24
dev.off()

pdf("PBSClusters.pdf", height = 8, width =8)
gatePBS
dev.off()

pdf("WTClusters.pdf", height = 8, width =8)
gateWT
dev.off()

MedianVioSeu <- function() stat_summary(fun.y = median, geom='point', size = 15, colour = "darkred", shape = 95)

apply.MedianVioSeu.and.combine <- function(ls.ggplots=pl$A) {
  for (i in 1:length(ls.ggplots)) ls.ggplots[[i]] = ls.ggplots[[i]]+MedianVioSeu()
  return(CombinePlots(ls.ggplots, legend = 'none'))
}

pdf(" ViolinsGatesAllTreatmentsIQR.pdf", height = 4, width = 40)
grid.arrange(Ctla4Violin+ geom_boxplot(width=0.05), Cd44Violin+ geom_boxplot(width=0.05), IcosViolin+ geom_boxplot(width=0.05), Il2raViolin+ geom_boxplot(width=0.05), Tnfrsf9Violin+ geom_boxplot(width=0.05), Tnfrsf18Violin+ geom_boxplot(width=0.05), Ifitm2Violin+ geom_boxplot(width=0.05), Ifitm3Violin+ geom_boxplot(width=0.05), GzmbViolin+ geom_boxplot(width=0.05), Tgfb1Violin+ geom_boxplot(width=0.05), nrow = 1)
dev.off()

pdf("geneModulesViolin.pdf", height = 12, width = 8)
grid.arrange(plot1 + geom_boxplot(width = .05, aes(color = "white")) + scale_color_manual(values = c("darkgreen", "gray","gray")), plot2+ geom_boxplot(width = .05, aes(color = "white")) + scale_color_manual(values = c("darkgreen", "gray","gray")), plot3+ geom_boxplot(width = .05, aes(color = "white")) + scale_color_manual(values = c("darkgreen", "gray","gray")), plot4+ geom_boxplot(width = .05, aes(color = "white")) + scale_color_manual(values = c("darkgreen", "gray","gray")), plot5+ geom_boxplot(width = .05, aes(color = "white")) + scale_color_manual(values = c("darkgreen", "gray","gray")))
dev.off()
```